<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Intermediate R - biological data: Microbiome application</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png"><link rel="manifest" href="favicons/incubator/site.webmanifest"><link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/06-application-microbiome.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Intermediate R - biological data
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Intermediate R - biological data
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Intermediate R - biological data
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 67%" class="percentage">
    67%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 67%" aria-valuenow="67" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/06-application-microbiome.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-introduction.html">1. Introduction</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-bioconductor.html">2. Bioconductor Introduction &amp; Install</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-biological-sequences.html">3. Biological sequences</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-annotations.html">4. Working with annotations</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-s4.html">5. The S4 class</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        6. Microbiome application
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#install-packages">Install packages</a></li>
<li><a href="#example-datasets">Example datasets</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-reproducible-analysis.html">7. Communicating &amp; Reproducing Analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="05-s4.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="07-reproducible-analysis.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="05-s4.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: The S4 class
        </a>
        <a class="chapter-link float-end" href="07-reproducible-analysis.html" rel="next">
          Next: Communicating &amp;...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Microbiome application</h1>
        <p>Last updated on 2025-11-11 |

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/episodes/06-application-microbiome.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<p>In this example we will explore some bioconductor packages for the
analysis of microbiome data. We will focus on managing and storing data
in a structures and reproducible way and then fine tune some
visualizations. We will follow on with a similar set on analysis in the
next session for <a href="">communicating &amp; reproducing
analysis</a></p>
<p>During this we will be making use of two main packages from the
bioconductor library:</p>
<ul><li><a href="https://www.bioconductor.org/packages/release/bioc/html/microbiome.html" class="external-link"><em>Microbiome</em></a></li>
<li><a href="https://www.bioconductor.org/packages/release/bioc/html/phyloseq.html" class="external-link"><em>Phyloseq</em></a></li>
</ul><p>We will also introduce a new data class - <code>phyloseq-class</code>
from the package <a href="https://joey711.github.io/phyloseq/" class="external-link"><em>Phyloseq</em></a> that
uses the S4 class system.</p>
<blockquote>
<p>The <a href="https://www.bioconductor.org/packages/release/bioc/html/phyloseq.html" class="external-link"><em>Phyloseq</em></a>
package is designed to store and manage phylogenetic sequencing data,
including abundance data, phylogenetic trees, taxonomic assignments, and
sample metadata, within a single experiment-level S4 object. This S4
object-oriented approach provides a structured and organized way to
handle complex microbiome data, facilitating reproducible analyses and
data sharing.</p>
</blockquote>
<p>At its core is the <code>phyloseq-class</code> object, which
integrates multiple data types: OTU/ASV tables, taxonomic annotations,
sample metadata, and phylogenetic trees. While originally developed for
16S rRNA microbiome studies, phyloseq-class objects are also widely used
in:</p>
<ul><li>eDNA metabarcoding studies (e.g., biodiversity monitoring in aquatic
or terrestrial environments)</li>
<li>Fungal ITS sequencing projects</li>
<li>Metagenomic and metatranscriptomic datasets with annotated taxonomic
profiles</li>
<li>Environmental genomics and ecological community profiling</li>
</ul><section><h2 class="section-heading" id="install-packages">Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a></h2>
<hr class="half-width"><p>Install our <a href="https://www.bioconductor.org/packages/release/bioc/html/microbiome.html" class="external-link"><em>Microbiome</em></a>
and <a href="https://www.bioconductor.org/packages/release/bioc/html/phyloseq.html" class="external-link"><em>Phyloseq</em></a>
packages using <code>BiocManager::install()</code></p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw">require</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The following initializes usage of Bioc devel</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span>version<span class="op">=</span><span class="st">'devel'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"phyloseq"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"microbiome"</span><span class="op">)</span></span></code></pre>
</div>
<p>Now load in the packages we need for this analysis</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">microbiome</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">phyloseq</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">knitr</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="example-datasets">Example datasets<a class="anchor" aria-label="anchor" href="#example-datasets"></a></h2>
<hr class="half-width"><p>We will explore some example datasets that come with the packages <a href="https://microbiome.github.io/tutorials/Data.html" class="external-link">Microbiome*</a>
and <a href="https://joey711.github.io/phyloseq/Example-Data" class="external-link"><em>Phyloseq</em></a>.
We will take some time to explore what information they contain, how we
can view the datasets and how they differ.</p>
<p><strong>Explore dataset 1 - <code>dietswap</code></strong></p>
<p>The first dataset we will load is from <a href="https://dx.doi.org/10.1038/ncomms7342" class="external-link">O’Keefe S et al. Nature
Communications, 2015</a> made available by <a href="https://microbiome.github.io/tutorials/Data.html" class="external-link"><em>Microbiome</em></a>
package. This study contains microbiome data from a study with African
and African American groups undergoing a two-week diet swap. This data
set is based on the Human Intestinal Tract (HIT)Chip phylogenetic 16S
microarray (<a href="https://doi.org/10.1111/j.1462-2920.2009.01900.x" class="external-link">Rajilic-Stojanovic
et al. 2009</a>). This profiling technology differs from the more widely
used 16S rRNA amplicon sequencing. Column metadata includes the subject
identifier, sex, nationality, group information, sample identifier, time
point information, time point information within group and BMI group.
Row metadata contains taxonomic information on the Phylum, Family and
Genus level.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">data</span><span class="op">(</span><span class="va">dietswap</span><span class="op">)</span></span></code></pre>
</div>
<p><strong>Explore dataset 2 - <code>peerj32</code></strong></p>
<p>The second dataset we will load is the <code>peerj32</code> data from
<a href="https://doi.org/10.7717/peerj.32" class="external-link">Lahti et al. PeerJ 1:e32,
2013</a> made available by <a href="https://microbiome.github.io/tutorials/Data.html" class="external-link"><em>Microbiome</em></a>
package. This characterizes associations between human intestinal
microbiota and blood serum lipids. Note that this data set contains an
additional data matrix of lipid species.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">data</span><span class="op">(</span><span class="va">peerj32</span><span class="op">)</span></span></code></pre>
</div>
<p><strong>Explore dataset 3 - <code>GlobalPatterns</code></strong></p>
<p>The third dataset we will load is the <code>GlobalPatterns</code>
data from <a href="https://doi.org/10.1073/pnas.1000080107" class="external-link">Caporaso et
al. PNAS, 108, 4516-4522, 2011</a> made available by <a href="https://joey711.github.io/phyloseq/Example-Data" class="external-link"><em>Phyloseq</em></a>.
This work compared the microbial communities from 25 environmental
samples and three known “mock communities” – a total of 9 sample types –
at a depth averaging 3.1 million reads per sample. Authors were able to
reproduce diversity patterns seen in many other published studies, while
also investigating technical issues/bias by applying the same techniques
to simulated microbial communities of known composition.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">data</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span></span></code></pre>
</div>
<div id="exploring-data" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exploring-data" class="callout-inner">
<h3 class="callout-title">Exploring data</h3>
<div class="callout-content">
<p>Explore the datasets <code>dietswap</code>, <code>peerj32</code> and
<code>GlobalPatterns</code> that you just loaded. You can do this by
navigating to your “Environment” pane, or in the console try the
following <code>dietswap@</code> and see what options are available to
you. Can you view the metadata for this data set? Why does
<code>peerj32@</code> not work in the same way?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>To see the full data within the <code>peerj32</code> object try
typing <code>peerj32$</code> into the console.</p>
</div>
</div>
</div>
</div>
<div id="missing-information" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="missing-information" class="callout-inner">
<h3 class="callout-title">Missing information</h3>
<div class="callout-content">
<p>Can you identify what information is missing in each of the three
<code>phyloseq-class</code> objects for the datasets above?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>Check the contents of each of the three datasets separately and you
can see that both <code>dietswap</code> and
<code>peerj32$phyloseq</code> objects only contain the minimum data
required - <code>otu_table()</code>, <code>sample_data()</code> and
<code>tax_table()</code>. The <code>GlobalPatterns</code> data also
contained additional taxonomic ranks and <code>phy_tree()</code> which
is a phylogenetic tree that we can use when performing analysis.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="sc">&gt;</span> dietswap</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>phyloseq<span class="sc">-</span>class experiment<span class="sc">-</span>level object</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">otu_table</span>()   OTU Table<span class="sc">:</span>         [ <span class="dv">130</span> taxa and <span class="dv">222</span> samples ]</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="fu">sample_data</span>() Sample Data<span class="sc">:</span>       [ <span class="dv">222</span> samples by <span class="dv">8</span> sample variables ]</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="fu">tax_table</span>()   Taxonomy Table<span class="sc">:</span>    [ <span class="dv">130</span> taxa by <span class="dv">3</span> taxonomic ranks ]</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="sc">&gt;</span> peerj32<span class="sc">$</span>phyloseq</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>phyloseq<span class="sc">-</span>class experiment<span class="sc">-</span>level object</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="fu">otu_table</span>()   OTU Table<span class="sc">:</span>         [ <span class="dv">130</span> taxa and <span class="dv">44</span> samples ]</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="fu">sample_data</span>() Sample Data<span class="sc">:</span>       [ <span class="dv">44</span> samples by <span class="dv">5</span> sample variables ]</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="fu">tax_table</span>()   Taxonomy Table<span class="sc">:</span>    [ <span class="dv">130</span> taxa by <span class="dv">3</span> taxonomic ranks ]</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="sc">&gt;</span> GlobalPatterns</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>phyloseq<span class="sc">-</span>class experiment<span class="sc">-</span>level object</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="fu">otu_table</span>()   OTU Table<span class="sc">:</span>         [ <span class="dv">19216</span> taxa and <span class="dv">26</span> samples ]</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="fu">sample_data</span>() Sample Data<span class="sc">:</span>       [ <span class="dv">26</span> samples by <span class="dv">7</span> sample variables ]</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="fu">tax_table</span>()   Taxonomy Table<span class="sc">:</span>    [ <span class="dv">19216</span> taxa by <span class="dv">7</span> taxonomic ranks ]</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="fu">phy_tree</span>()    Phylogenetic Tree<span class="sc">:</span> [ <span class="dv">19216</span> tips and <span class="dv">19215</span> internal nodes ]</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><div class="section level1">
<h1 id="working-with-data">Working with data<a class="anchor" aria-label="anchor" href="#working-with-data"></a></h1>
<p>Now we will have a go at working through some data
<strong>visualizations</strong> and improving aesthetics ready for
publications or sharing with colleagues.</p>
<p>For this exercise we will work through using the
<code>GlobalPatterns</code> data.</p>
<blockquote>
<p>The <code>GlobalPatterns</code> data is from <a href="https://doi.org/10.1073/pnas.1000080107" class="external-link">Caporaso et al. PNAS,
108, 4516-4522, 2011</a> made available by <a href="https://joey711.github.io/phyloseq/Example-Data" class="external-link"><em>Phyloseq</em></a>.
This work compared the microbial communities from 25 environmental
samples and three known “mock communities”</p>
</blockquote>
<p>First a re-fresh of what the sample data contains for this
dataset</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#show just the first five rows</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">@</span><span class="va">sam_data</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Sample Data:        [6 samples by 7 sample variables]:
        X.SampleID  Primer Final_Barcode Barcode_truncated_plus_T
CL3            CL3 ILBC_01        AACGCA                   TGCGTT
CC1            CC1 ILBC_02        AACTCG                   CGAGTT
SV1            SV1 ILBC_03        AACTGT                   ACAGTT
M31Fcsw    M31Fcsw ILBC_04        AAGAGA                   TCTCTT
M11Fcsw    M11Fcsw ILBC_05        AAGCTG                   CAGCTT
M31Plmr    M31Plmr ILBC_07        AATCGT                   ACGATT
        Barcode_full_length SampleType
CL3             CTAGCGTGCGT       Soil
CC1             CATCGACGAGT       Soil
SV1             GTACGCACAGT       Soil
M31Fcsw         TCGACATCTCT      Feces
M11Fcsw         CGACTGCAGCT      Feces
M31Plmr         CGAGTCACGAT       Skin
                                       Description
CL3       Calhoun South Carolina Pine soil, pH 4.9
CC1       Cedar Creek Minnesota, grassland, pH 6.1
SV1     Sevilleta new Mexico, desert scrub, pH 8.3
M31Fcsw    M3, Day 1, fecal swab, whole body study
M11Fcsw   M1, Day 1, fecal swab, whole body study
M31Plmr    M3, Day 1, right palm, whole body study</code></pre>
</div>
<p>Next we often need to do some data pruning or subsetting. In this
case we are making sure we remove taxa that doesn’t have any sequences
present.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GP</span> <span class="op">&lt;-</span> <span class="fu">prune_species</span><span class="op">(</span><span class="fu">speciesSums</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">GlobalPatterns</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in prune_species(speciesSums(GlobalPatterns) &gt; 0, GlobalPatterns): 'prune_species' is deprecated.
Use 'prune_taxa' instead.
See help("Deprecated") and help("phyloseq-deprecated").</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in speciesSums(GlobalPatterns): 'speciesSums' is deprecated.
Use 'taxa_sums' instead.
See help("Deprecated") and help("phyloseq-deprecated").</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compare the two datasets</span></span>
<span><span class="va">GlobalPatterns</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GP</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 18988 taxa and 26 samples ]
sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
tax_table()   Taxonomy Table:    [ 18988 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 18988 tips and 18987 internal nodes ]</code></pre>
</div>
<div class="section level2">
<h2 id="alpha-diversity-plots">Alpha diversity plots<a class="anchor" aria-label="anchor" href="#alpha-diversity-plots"></a></h2>
<p>For microbiome (or community analysis) a common strategy is to
visualize the alpha diversity of samples. Essentially this is how many
unique taxa were identified in each sample.</p>
<p>At the simplest level we can plot all the alpha diversity measures
from the function <code>plot_richness()</code> from the <a href="https://www.bioconductor.org/packages/release/bioc/html/phyloseq.html" class="external-link"><em>Phyloseq</em></a>
package.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_richness</span><span class="op">(</span><span class="va">GP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: `aes_string()` was deprecated in ggplot2 3.0.0.
ℹ Please use tidy evaluation idioms with `aes()`.
ℹ See also `vignette("ggplot2-in-packages")` for more information.
ℹ The deprecated feature was likely used in the phyloseq package.
  Please report the issue at &lt;https://github.com/joey711/phyloseq/issues&gt;.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
generated.</code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-8-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Note that in this case, the Fisher calculation results in a warning
(but still plots).</p>
<p>There is a lot going on here so let’s try just look at two different
alpha diversity measures.</p>
<p>We can avoid this by specifying a measures argument to plot_richness,
which will include just the alpha-diversity measures that we want.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_richness</span><span class="op">(</span><span class="va">GP</span>, measures<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"Chao1"</span>, <span class="st">"Shannon"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-9-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="getting-help-on-functions" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="getting-help-on-functions" class="callout-inner">
<h3 class="callout-title">Getting help on functions</h3>
<div class="callout-content">
<p>You can find more information on the funciton by typing
<code>?plot_richness</code> into the console. Have a look and see what
measures are available using this function.</p>
</div>
</div>
</div>
<p>We can specify a sample variable on which to group/organize samples
along the horizontal (<code>x</code>) axis. An experimentally meaningful
categorical variable is usually a good choice – in this case, the
“<code>SampleType</code>” variable works much better than attempting to
interpret the sample names directly (as in the previous plot):</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_richness</span><span class="op">(</span><span class="va">GP</span>, x<span class="op">=</span><span class="st">"SampleType"</span>, measures<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"Chao1"</span>, <span class="st">"Shannon"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-10-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Now suppose we wanted to use an external variable in the plot that
isn’t in the <code>GP</code> dataset already – for example, a logical
that indicated whether or not the samples are human-associated. First,
define this new variable, <code>human</code>, as a factor (other vectors
could also work; or other data you might have describing the
samples).</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sampleData</span><span class="op">(</span><span class="va">GP</span><span class="op">)</span><span class="op">$</span><span class="va">human</span> <span class="op">&lt;-</span> <span class="fu">getVariable</span><span class="op">(</span><span class="va">GP</span>, <span class="st">"SampleType"</span><span class="op">)</span> <span class="op">%in%</span></span>
<span>  <span class="fu">c</span><span class="op">(</span><span class="st">"Feces"</span>, <span class="st">"Mock"</span>, <span class="st">"Skin"</span>, <span class="st">"Tongue"</span><span class="op">)</span></span></code></pre>
</div>
<p>Now tell <code>plot_richness</code> to map the new human variable on
the horizontal axis, and shade the points in different color groups,
according to which “<code>SampleType</code>” they belong.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_richness</span><span class="op">(</span></span>
<span>  <span class="va">GP</span>,</span>
<span>  x <span class="op">=</span> <span class="st">"human"</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"SampleType"</span>,</span>
<span>  measures <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Chao1"</span>, <span class="st">"Shannon"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-12-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can merge samples that are from the environment
(<code>SampleType</code>), and make the points bigger with a ggplot2
layer. First, merge the samples.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GPst</span> <span class="op">&lt;-</span> <span class="fu">merge_samples</span><span class="op">(</span><span class="va">GP</span>, <span class="st">"SampleType"</span><span class="op">)</span></span>
<span><span class="co"># repair variables that were damaged during merge (coerced to numeric)</span></span>
<span><span class="fu">sample_data</span><span class="op">(</span><span class="va">GPst</span><span class="op">)</span><span class="op">$</span><span class="va">SampleType</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="fu">sample_names</span><span class="op">(</span><span class="va">GPst</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">sample_data</span><span class="op">(</span><span class="va">GPst</span><span class="op">)</span><span class="op">$</span><span class="va">human</span> <span class="op">&lt;-</span> <span class="fu">as.logical</span><span class="op">(</span><span class="fu">sample_data</span><span class="op">(</span><span class="va">GPst</span><span class="op">)</span><span class="op">$</span><span class="va">human</span><span class="op">)</span></span></code></pre>
</div>
<p>Now we can plot this environment-merged version of the data. First
store the default ggplot graphic as <code>p</code>, then add an
additional <code>geom_point</code> layer with a large size and slight
transparency.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">plot_richness</span><span class="op">(</span></span>
<span>  <span class="va">GPst</span>,</span>
<span>  x <span class="op">=</span> <span class="st">"human"</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"SampleType"</span>,</span>
<span>  measures <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Chao1"</span>, <span class="st">"Shannon"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-14-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p><strong>More details about ggplot2</strong></p>
<p>For those interested in why this works so concisely
<code>(p + geom_point(size=4, alpha=0.7))</code>, it is because the rest
of the aesthetic mapping and data are contained in the ggplot object,
<code>p</code>, and so is inherited in the call to the ggplot2 geometric
object layer function, <code>geom_point</code>, by default since we
didn’t specify alternative <code>aes</code> or <code>data</code>
arguments. Although we could have if we wanted to. This perhaps sounds
more confusing than it is, and I find it easier to understand by
inspecting the examples I’ve shown here.</p>
<p>You’ll also notice that the original smaller points are still on the
plot. This is because they were the first layer, and our larger points
are semi-transparent. I find this kind of distracting, and doesn’t add
any information or clarity. The good news is that layers can be removed
from a ggplot object with standard list notation (using the dollar sign
<code>$</code>).</p>
<p>First, check which lists are present in <code>p</code>.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">layers</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$geom_point
geom_point: na.rm = TRUE
stat_identity: na.rm = TRUE
position_identity

$geom_errorbar
mapping: ymax = ~value + se, ymin = ~value - se
geom_errorbar: na.rm = FALSE, orientation = NA, lineend = butt, width = 0.1
stat_identity: na.rm = FALSE
position_identity </code></pre>
</div>
<p>We can see that the first layer is the one specifying the original
points, which are small. We can use negative indexing to “pop” it out,
then add a new <code>geom_point</code> layer with larger point size (the
following two lines).</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">layers</span> <span class="op">&lt;-</span> <span class="va">p</span><span class="op">$</span><span class="va">layers</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<p>We can further improvde on the plots by setting the theme to
something simpler.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">+</span> <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-17-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
<div class="section level2">
<h2 id="updating-the-data-re-formatting">Updating the data &amp; re-formatting<a class="anchor" aria-label="anchor" href="#updating-the-data-re-formatting"></a></h2>
<p>Now we are going to get a bit more tedious with our data formatting
and figures. Lets say on the plot above we want to actually order the
<code>SampleType</code> on the legend by group the sample of human
origin together and those of non-human together.</p>
<p>To do this - we have to first extract out the <code>sam_data</code>
as its own data frame and add in a new column with the factor in this
order.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract the sample data as a data frame</span></span>
<span><span class="va">sample_data_df</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">@</span><span class="va">sam_data</span><span class="op">)</span></span>
<span><span class="co"># Give it a slightly different column name `SampleTypeOrigin`</span></span>
<span><span class="va">sample_data_df</span><span class="op">$</span><span class="va">SampleTypeOrigin</span> <span class="op">&lt;-</span> <span class="va">sample_data_df</span><span class="op">$</span><span class="va">SampleType</span></span>
<span></span>
<span><span class="co"># Set the order for this factor</span></span>
<span><span class="va">desired_order_levels</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span></span>
<span>  <span class="st">"Feces"</span>,</span>
<span>  <span class="st">"Skin"</span>,</span>
<span>  <span class="st">"Tongue"</span>,</span>
<span>  <span class="st">"Mock"</span>,</span>
<span>  <span class="st">"Freshwater"</span>,</span>
<span>  <span class="st">"Freshwater (creek)"</span>,</span>
<span>  <span class="st">"Ocean"</span>,</span>
<span>  <span class="st">"Sediment (estuary)"</span>,</span>
<span>  <span class="st">"Soil"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Now we set the factor and order</span></span>
<span><span class="va">sample_data_df</span><span class="op">$</span><span class="va">SampleTypeOrigin</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">sample_data_df</span><span class="op">$</span><span class="va">SampleTypeOrigin</span>, levels <span class="op">=</span> <span class="va">desired_order_levels</span><span class="op">)</span></span>
<span><span class="co"># Re-make the `GlobalPatterns` data object with this data frame and you will see we now have 9 sample variables</span></span>
<span><span class="fu">sample_data</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample_data_df</span></span>
<span><span class="va">GlobalPatterns</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
sample_data() Sample Data:       [ 26 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</code></pre>
</div>
<p>But because we remade the phyloseq object we now need to perform the
same subsetting we did to generate the <code>GP</code> data object
again. We also have to add in that extra <code>human</code> variable
again too.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remove taxa with 0 reads</span></span>
<span><span class="va">GP</span> <span class="op">&lt;-</span> <span class="fu">prune_species</span><span class="op">(</span><span class="fu">speciesSums</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">GlobalPatterns</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in prune_species(speciesSums(GlobalPatterns) &gt; 0, GlobalPatterns): 'prune_species' is deprecated.
Use 'prune_taxa' instead.
See help("Deprecated") and help("phyloseq-deprecated").</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in speciesSums(GlobalPatterns): 'speciesSums' is deprecated.
Use 'taxa_sums' instead.
See help("Deprecated") and help("phyloseq-deprecated").</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add in the human variable for SampleType</span></span>
<span><span class="fu">sampleData</span><span class="op">(</span><span class="va">GP</span><span class="op">)</span><span class="op">$</span><span class="va">human</span> <span class="op">&lt;-</span> <span class="fu">getVariable</span><span class="op">(</span><span class="va">GP</span>, <span class="st">"SampleType"</span><span class="op">)</span> <span class="op">%in%</span></span>
<span>  <span class="fu">c</span><span class="op">(</span><span class="st">"Feces"</span>, <span class="st">"Mock"</span>, <span class="st">"Skin"</span>, <span class="st">"Tongue"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in getVariable(GP, "SampleType"): 'getVariable' is deprecated.
Use 'get_variable' instead.
See help("Deprecated") and help("phyloseq-deprecated").</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in sampleData(`*tmp*`): 'sampleData' is deprecated.
Use 'sample_data' instead.
See help("Deprecated") and help("phyloseq-deprecated").</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in `sampleData&lt;-`(`*tmp*`, value = new("sample_data", .Data = list(: 'sampleData&lt;-' is deprecated.
Use 'sample_data&lt;-' instead.
See help("Deprecated") and help("phyloseq-deprecated").</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sample_data</span><span class="op">(</span><span class="va">GP</span><span class="op">)</span><span class="op">$</span><span class="va">human</span> <span class="op">&lt;-</span> <span class="fu">as.logical</span><span class="op">(</span><span class="fu">sample_data</span><span class="op">(</span><span class="va">GP</span><span class="op">)</span><span class="op">$</span><span class="va">human</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="remaking-the-alpha-diversity-plot">Remaking the alpha diversity plot<a class="anchor" aria-label="anchor" href="#remaking-the-alpha-diversity-plot"></a></h3>
<p>Only <em>now</em> we can we remake that alpha diversity plot from
before - now the change in order of the legend</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">plot_richness</span><span class="op">(</span></span>
<span>  <span class="va">GP</span>,</span>
<span>  x <span class="op">=</span> <span class="st">"human"</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"SampleTypeOrigin"</span>,</span>
<span>  measures <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Chao1"</span>, <span class="st">"Shannon"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-20-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can then apply a manual customise colour scale for our variables,
grouping by colour</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">+</span> <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>  <span class="st">"#9e0142"</span>,</span>
<span>  <span class="st">"#d53e4f"</span>,</span>
<span>  <span class="st">"#f46d43"</span>,</span>
<span>  <span class="st">"#252525"</span>,</span>
<span>  <span class="st">"#e6f598"</span>,</span>
<span>  <span class="st">"#abdda4"</span>,</span>
<span>  <span class="st">"#66c2a5"</span>,</span>
<span>  <span class="st">"#3288bd"</span>,</span>
<span>  <span class="st">"#5e4fa2"</span></span>
<span><span class="op">)</span> <span class="op">)</span> <span class="op">+</span> <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-21-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><blockquote>
<p>In collaborative research, it’s almost inevitable that you’ll need to
add extra variables or create new ways to subset and view your data.
Different team members—each with unique backgrounds and priorities—often
request specific visualizations to highlight patterns relevant to their
perspective. Rather than treating these requests as a burden, you can
make them efficient and reproducible by following good practices.</p>
</blockquote>
<p>When your supervisor asks for a new grouping or your collaborator
wants to see results by location instead of treatment, you can adapt
quickly—without rewriting your entire analysis. This approach saves
time, reduces errors, and keeps your workflow transparent.</p>
</div>
</div>
<div class="section level2">
<h2 id="phylogenetic-trees">Phylogenetic trees<a class="anchor" aria-label="anchor" href="#phylogenetic-trees"></a></h2>
<p>The next set of figures work through some visualization pf
phylogenetic trees and incorporations of sample data for annotation,
examples taken from the <a href="https://joey711.github.io/phyloseq/plot_tree-examples.html" class="external-link">phyloseq
tree tutorial</a>.</p>
<p>We want to plot trees, sometimes even bootstrap values, but notice
that the node labels in the <code>GlobalPatterns</code> dataset are
actually a bit strange. They look like they might be bootstrap values,
but they sometimes have two decimals.</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">phy_tree</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span><span class="op">$</span><span class="va">node.label</span>, <span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] ""          "0.858.4"   "1.000.154" "0.764.3"   "0.995.2"   "1.000.2"
 [7] "0.943.7"   "0.971.6"   "0.766"     "0.611"    </code></pre>
</div>
<p>Could systematically remove the second decimal, but why not just take
the first 4 characters instead?</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">phy_tree</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span><span class="op">$</span><span class="va">node.label</span> <span class="op">=</span> <span class="fu">substr</span><span class="op">(</span><span class="fu">phy_tree</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span><span class="op">$</span><span class="va">node.label</span>, <span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span></span></code></pre>
</div>
<p>Great, now that we’re more happy with the node labels at least
looking like bootstrap values, we can move on to using these along with
other information about data mapped onto the tree graphic.</p>
<p>The <code>GlobalPatterns</code> dataset has many OTUs, more than we
would want to try to fit on a tree graphic</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ntaxa</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 19216</code></pre>
</div>
<p>So, let’s arbitrarily prune to just the first 50 OTUs in
<code>GlobalPatterns</code>, and store this as <code>physeq</code>,
which also happens to be the name for most main data parameters of
function in the phyloseq package.</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">physeq</span> <span class="op">&lt;-</span> <span class="fu">prune_taxa</span><span class="op">(</span><span class="fu">taxa_names</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span>, <span class="va">GlobalPatterns</span><span class="op">)</span></span></code></pre>
</div>
<p>Now let’s look at what happens with the default
<code>plot_tree</code> settings.</p>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">physeq</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-26-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>By default, black dots are annotated next to tips (OTUs) in the tree,
one for each sample in which that OTU was observed. Some have more dots
than others. Also by default, the node labels that were stored in the
tree were added next to each node without any processing (although we
had trimmed their length to 4 characters in the previous step).</p>
<p>What if we want to just see the tree with no sample points next to
the tips?</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">physeq</span>, <span class="st">"treeonly"</span><span class="op">)</span></span></code></pre>
</div>
<p><img src="fig/06-application-microbiome-rendered-unnamed-chunk-27-1.png" style="display: block; margin: auto;" class="figure">
And what about without the node labels either?</p>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">physeq</span>, <span class="st">"treeonly"</span>, <span class="va">nodeplotblank</span><span class="op">)</span></span></code></pre>
</div>
<p><img src="fig/06-application-microbiome-rendered-unnamed-chunk-28-1.png" style="display: block; margin: auto;" class="figure">
We can adjust the way branches are rotated to make it look nicer using
the <code>ladderize</code> parameter.</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">physeq</span>, <span class="st">"treeonly"</span>, <span class="va">nodeplotblank</span>, ladderize<span class="op">=</span><span class="st">"left"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-29-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">physeq</span>, <span class="st">"treeonly"</span>, <span class="va">nodeplotblank</span>, ladderize<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<p><img src="fig/06-application-microbiome-rendered-unnamed-chunk-29-2.png" style="display: block; margin: auto;" class="figure">
And what if we want to add the OTU labels next to each tip?</p>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">physeq</span>, nodelabf<span class="op">=</span><span class="va">nodeplotblank</span>, label.tips<span class="op">=</span><span class="st">"taxa_names"</span>, ladderize<span class="op">=</span><span class="st">"left"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-30-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Any <code>method</code> parameter argument other than
<code>"sampledodge"</code> (the default) will not add dodged sample
points next to the tips.</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">physeq</span>, <span class="st">"anythingelse"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-31-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="section level3">
<h3 id="mapping-variables-in-data">Mapping Variables in Data<a class="anchor" aria-label="anchor" href="#mapping-variables-in-data"></a></h3>
<p>In the default argument to <code>method</code>,
<code>"sampledodge"</code>, a point is added next to each OTU tip in the
tree for every sample in which that OTU was observed. We can then map
certain aesthetic features of these points to variables in our data.</p>
<div class="section level4">
<h4 id="color">Color<a class="anchor" aria-label="anchor" href="#color"></a></h4>
<p>Color is one of the most useful aesthetics in tree graphics when they
are complicated. Color can be mapped to either taxonomic ranks or sample
covariates. For instance, we can map color to the type of sample
collected (environmental location).</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">physeq</span>, nodelabf<span class="op">=</span><span class="fu">nodeplotboot</span><span class="op">(</span><span class="op">)</span>, ladderize<span class="op">=</span><span class="st">"left"</span>, color<span class="op">=</span><span class="st">"SampleType"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-32-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p><strong>Keeping colours consistant!</strong></p>
<p>Now that you are intermediate R users you would be picking up on the
different colour scales used to map against the same variable. It is
best practice to keep these consistent across the different types of
figures. So let’s regenerate the figure mapping the
<code>SampleTypeOrigin</code> variable to the same set of colours we
used above for the alpha diversity plot</p>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">physeq</span>,</span>
<span>  nodelabf <span class="op">=</span> <span class="fu">nodeplotboot</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  ladderize <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"SampleTypeOrigin"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_color_manual</span><span class="op">(</span></span>
<span>    values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>      <span class="st">"#9e0142"</span>,</span>
<span>      <span class="st">"#d53e4f"</span>,</span>
<span>      <span class="st">"#f46d43"</span>,</span>
<span>      <span class="st">"#252525"</span>,</span>
<span>      <span class="st">"#e6f598"</span>,</span>
<span>      <span class="st">"#abdda4"</span>,</span>
<span>      <span class="st">"#66c2a5"</span>,</span>
<span>      <span class="st">"#3288bd"</span>,</span>
<span>      <span class="st">"#5e4fa2"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-33-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>You can make this even more efficent by including the colouring of a
variable in your global R chunk for the set up and then refer to it
using the value - in this case <code>SampleTypeOriginCols</code></p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SampleTypeOriginCols</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span></span>
<span>      <span class="st">"#9e0142"</span>,</span>
<span>      <span class="st">"#d53e4f"</span>,</span>
<span>      <span class="st">"#f46d43"</span>,</span>
<span>      <span class="st">"#252525"</span>,</span>
<span>      <span class="st">"#e6f598"</span>,</span>
<span>      <span class="st">"#abdda4"</span>,</span>
<span>      <span class="st">"#66c2a5"</span>,</span>
<span>      <span class="st">"#3288bd"</span>,</span>
<span>      <span class="st">"#5e4fa2"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">physeq</span>,</span>
<span>  nodelabf <span class="op">=</span> <span class="fu">nodeplotboot</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  ladderize <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"SampleTypeOrigin"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_color_manual</span><span class="op">(</span></span>
<span>    values <span class="op">=</span> <span class="va">SampleTypeOriginCols</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-34-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can also map color to taxonomic class.</p>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">physeq</span>, nodelabf<span class="op">=</span><span class="fu">nodeplotboot</span><span class="op">(</span><span class="op">)</span>, ladderize<span class="op">=</span><span class="st">"left"</span>, color<span class="op">=</span><span class="st">"Class"</span><span class="op">)</span></span></code></pre>
</div>
<p><img src="fig/06-application-microbiome-rendered-unnamed-chunk-35-1.png" style="display: block; margin: auto;" class="figure">
This time we don’t mind the differrent colours because we are mapping it
against taxonomy</p>
</div>
<div class="section level4">
<h4 id="shape">Shape<a class="anchor" aria-label="anchor" href="#shape"></a></h4>
<p>You can also map a variable to point shape if it has 6 or fewer
categories, and this can be done even when color is also mapped. Here we
map shape to taxonomic class so that we can still indicate it in the
graphic while also mapping <code>SampleType</code> to point color.</p>
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  nodelabf <span class="op">=</span> <span class="fu">nodeplotboot</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  ladderize <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"SampleTypeOrigin"</span>,</span>
<span>  shape <span class="op">=</span> <span class="st">"Class"</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">SampleTypeOriginCols</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-36-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
</div>
<div class="section level3">
<h3 id="node-labels">Node labels<a class="anchor" aria-label="anchor" href="#node-labels"></a></h3>
<p>One of the most common reasons to label nodes is to add confidence
measures, often a bootstrap value, to the nodes of the tree. The
following graphics show different ways of doing this (labels are added
by default if present in your tree).</p>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The default</span></span>
<span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">physeq</span>, color <span class="op">=</span> <span class="st">"SampleTypeOrigin"</span>, ladderize <span class="op">=</span> <span class="st">"left"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">SampleTypeOriginCols</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-37-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Special bootstrap label</span></span>
<span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">physeq</span>,</span>
<span>  nodelabf <span class="op">=</span> <span class="fu">nodeplotboot</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"SampleTypeOrigin"</span>,</span>
<span>  ladderize <span class="op">=</span> <span class="st">"left"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">SampleTypeOriginCols</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-37-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Special bootstrap label with alternative thresholds</span></span>
<span><span class="fu">plot_tree</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  nodelabf <span class="op">=</span> <span class="fu">nodeplotboot</span><span class="op">(</span><span class="fl">80</span>, <span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"SampleTypeOrigin"</span>,</span>
<span>  ladderize <span class="op">=</span> <span class="st">"left"</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">SampleTypeOriginCols</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-37-3.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
<div class="section level3">
<h3 id="tip-labels">Tip labels<a class="anchor" aria-label="anchor" href="#tip-labels"></a></h3>
<ul><li>
<strong>label.tips</strong> - The <code>label.tips</code> parameter
controls labeling of tree tips (AKA leaves). Default is NULL, indicating
that no tip labels will be printed. If <code>"taxa_names"</code> is a
special argument resulting in the OTU name (try <code>taxa_names</code>
function) being labelled next to the leaves or next to the set of points
that label the leaves. Alternatively, if your data object contains a
<code>tax_table</code>, then one of the rank names (from
<code>rank_names(physeq)</code>) can be provided, and the classification
of each OTU at that rank will be labeled instead.</li>
<li>
<strong>text.size</strong> - A positive numeric argument indicating
the ggplot2 size parameter for the taxa labels. Default is
<code>NULL</code>. If left as <code>NULL</code>, this function will
automatically calculate a (hopefully) optimal text size given the size
constraints posed by the tree itself (for a vertical tree). This
argument is included mainly in case the automatically-calculated size is
wrong and you want to change it. Note that this parameter is only
meaningful if <code>label.tips</code> is not <code>NULL</code>
</li>
</ul><div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  nodelabf <span class="op">=</span> <span class="fu">nodeplotboot</span><span class="op">(</span><span class="fl">80</span>, <span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"SampleTypeOrigin"</span>,</span>
<span>  label.tips <span class="op">=</span> <span class="st">"taxa_names"</span>,</span>
<span>  ladderize <span class="op">=</span> <span class="st">"left"</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">SampleTypeOriginCols</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-38-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
<div class="section level3">
<h3 id="radial-tree">Radial Tree<a class="anchor" aria-label="anchor" href="#radial-tree"></a></h3>
<p>Making a radial tree is easy with ggplot2, simply recognizing that
our vertically-oriented tree is a cartesian mapping of the data to a
graphic – and that a radial tree is the same mapping, but with polar
coordinates instead.</p>
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">data</span><span class="op">(</span><span class="va">esophagus</span><span class="op">)</span></span>
<span><span class="fu">plot_tree</span><span class="op">(</span><span class="va">esophagus</span>, color <span class="op">=</span> <span class="st">"Sample"</span>, ladderize <span class="op">=</span> <span class="st">"left"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">coord_polar</span><span class="op">(</span>theta <span class="op">=</span></span>
<span>    <span class="st">"y"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-39-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The <code>GlobalPatterns</code> dataset has additional data we can
map, so we will re-do some preliminary data loading/trimming to make
this radial-tree example self contained, and then show the same plot as
above.</p>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  nodelabf <span class="op">=</span> <span class="fu">nodeplotboot</span><span class="op">(</span><span class="fl">60</span>, <span class="fl">60</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"SampleTypeOrigin"</span>,</span>
<span>  shape <span class="op">=</span> <span class="st">"Class"</span>,</span>
<span>  ladderize <span class="op">=</span> <span class="st">"left"</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu">coord_polar</span><span class="op">(</span>theta <span class="op">=</span> <span class="st">"y"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">SampleTypeOriginCols</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-40-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="more-examples-with-the-global-patterns-dataset" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="more-examples-with-the-global-patterns-dataset" class="callout-inner">
<h3 class="callout-title">More Examples with the Global Patterns
dataset</h3>
<div class="callout-content">
<p>Try exploring more options for generating trees but subsetting
specific taxa. Starting the Archaea (<em>hint: Archaea is a
kindgom</em>).</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb66">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gpa</span> <span class="op">&lt;-</span> <span class="fu">subset_taxa</span><span class="op">(</span><span class="va">GlobalPatterns</span>, <span class="va">Kingdom</span><span class="op">==</span><span class="st">"Archaea"</span><span class="op">)</span></span></code></pre>
</div>
<p>Identify if the number of different Archaeal taxa is suitable for a
tree</p>
<div class="codewrapper sourceCode" id="cb67">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ntaxa</span><span class="op">(</span><span class="va">gpa</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 208</code></pre>
</div>
<p>Visualise tree</p>
<div class="codewrapper sourceCode" id="cb69">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_tree</span><span class="op">(</span></span>
<span>  <span class="va">gpa</span>,</span>
<span>  nodelabf <span class="op">=</span> <span class="fu">nodeplotboot</span><span class="op">(</span><span class="fl">80</span>, <span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"SampleTypeOrigin"</span>,</span>
<span>  label.tips <span class="op">=</span> <span class="st">"taxa_names"</span>,</span>
<span>  ladderize <span class="op">=</span> <span class="st">"left"</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">SampleTypeOriginCols</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-43-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
</div>
</div>
</div>
<hr><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>Bespoke often domain specific data classes are often built to serve
a specific set of data, however by utilising fundamental R principals
such as S4 class structures transferring formats between packages for
analysis is much easier.</li>
<li>Look for widely accepted data classes in your field before
formatting your data for a specific package early on.</li>
<li>Packages that generate <a href="https://ggplot2.tidyverse.org/" class="external-link"><em>ggplot2</em></a> compatible
figures allow for <em>easier</em> manipulations and customise using
familar syntax</li>
<li>Get feedback early on from preliminary analysis, this will help you
improve your data and reporting for generating downstream analysis and
figures.</li>
<li>Get comfortabe with the data lifecycle and
<strong>re</strong>-producing analysis and figures!</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</div>
</div>



      </div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="05-s4.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="07-reproducible-analysis.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="05-s4.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: The S4 class
        </a>
        <a class="chapter-link float-end" href="07-reproducible-analysis.html" rel="next">
          Next: Communicating &amp;...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/episodes/06-application-microbiome.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/" class="external-link">Source</a></p>
				<p><a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:siobhon.egan@murdoch.edu.au">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://siobhon-egan.github.io/2025-11-13-COMBINE-WA/06-application-microbiome.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Microbiome application",
  "creativeWorkStatus": "active",
  "url": "https://siobhon-egan.github.io/2025-11-13-COMBINE-WA/06-application-microbiome.html",
  "identifier": "https://siobhon-egan.github.io/2025-11-13-COMBINE-WA/06-application-microbiome.html",
  "dateCreated": "2025-11-13",
  "dateModified": "2025-11-11",
  "datePublished": "2025-11-11"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

