<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Intermediate R - biological data: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png">
<link rel="manifest" href="favicons/incubator/site.webmanifest">
<link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/aio.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Intermediate R - biological data
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Intermediate R - biological data
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Intermediate R - biological data
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/aio.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-introduction.html">1. Introduction</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-bioconductor.html">2. Bioconductor Introduction &amp; Install</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-biological-sequences.html">3. Biological sequences</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-annotations.html">4. Working with annotations</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-s4.html">5. The S4 class</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-application-microbiome.html">6. Microbiome application</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-reproducible-analysis.html">7. Communicating &amp; Reproducing Analysis</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-01-introduction"><p>Content from <a href="01-introduction.html">Introduction</a></p>
<hr>
<p>Last updated on 2025-11-11 |

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/episodes/01-introduction.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Am I using the correct version of R for this lesson?</li>
<li>Why does my version of R matter?</li>
<li>How do I obtain the files that are used in this lesson?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Ensure that participants are using the correct version of R to
reproduce exactly the contents of this lesson.</li>
<li>Download the example files for this lesson.</li>
<li>Provide overview of lesson and structure</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="version-of-r">Version of R<a class="anchor" aria-label="anchor" href="#version-of-r"></a>
</h2>
<hr class="half-width">
<p>This lesson was developed and tested with R version 4.5.2
(2025-10-31).</p>
<p>Take a moment to launch RStudio and verify that you are using R
version 4.5.x, with <code>x</code> being any patch version,
e.g. 4.5.2.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">R.version.string</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "R version 4.5.2 (2025-10-31)"</code></pre>
</div>
<p>This is important because Bioconductor uses the version of R running
in the current session to determine the version of Bioconductor packages
that can be installed in the R library associated with the current R
session. Using a different version of R while following this lesson may
lead to unexpected results.</p>
</section><section><h2 class="section-heading" id="download-files">Download files<a class="anchor" aria-label="anchor" href="#download-files"></a>
</h2>
<hr class="half-width">
<p>Several episodes in this lesson rely on example files that
participants need to download.</p>
<p>Run the code below programmatically create a folder called
<code>data</code> in the current working directory, and download the
lesson files in that folder.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dir.create</span><span class="op">(</span><span class="st">"data"</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Bioconductor/bioconductor-teaching/master/data/TrimmomaticAdapters/TruSeq3-PE-2.fa"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/TruSeq3-PE-2.fa"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Bioconductor/bioconductor-teaching/master/data/ActbGtf/actb.gtf"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/actb.gtf"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Bioconductor/bioconductor-teaching/master/data/ActbOrf/actb_orfs.fasta"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/actb_orfs.fasta"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>  url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Bioconductor/bioconductor-teaching/devel/data/SummarizedExperiment/counts.csv"</span>,</span>
<span>  destfile <span class="op">=</span> <span class="st">"data/counts.csv"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>  url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Bioconductor/bioconductor-teaching/devel/data/SummarizedExperiment/gene_metadata.csv"</span>,</span>
<span>  destfile <span class="op">=</span> <span class="st">"data/gene_metadata.csv"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>  url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Bioconductor/bioconductor-teaching/devel/data/SummarizedExperiment/sample_metadata.csv"</span>,</span>
<span>  destfile <span class="op">=</span> <span class="st">"data/sample_metadata.csv"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div id="note" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="note" class="callout-inner">
<h3 class="callout-title">Note</h3>
<div class="callout-content">
<p>Ideally, participants might want to create a new <a href="https://support.rstudio.com/hc/en-us/articles/200526207-Using-Projects" class="external-link">RStudio
project</a> and download the lesson files in a sub-directory of that
project.</p>
<p>Using an RStudio project sets the working directory to the root
directory of that project. As a consequence, code is executed relative
to that root directory, often avoiding the need for using absolute file
paths to import/export data from/to files.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="structure-of-lesson">Structure of lesson<a class="anchor" aria-label="anchor" href="#structure-of-lesson"></a>
</h2>
<hr class="half-width">
<p>The content in this lesson has been heavily influenced from the
material in the Carpentries Incubator lesson <a href="https://carpentries-incubator.github.io/bioc-project/" class="external-link">The
Bioconductor project</a>.</p>
<p>The first three lessons have been restructured from episodes with
additional material added to provide some practice with storing
biological data sets, visualization, reproducible workflow and
communicating results.</p>
<ul>
<li>
<strong>Part A - Bioconductor episodes</strong> - a revised and
reduced content from the Carpentries Incubator lesson <a href="https://carpentries-incubator.github.io/bioc-project/" class="external-link">The
Bioconductor project</a>.
<ul>
<li>Introduce, install and getting help with bioconductor</li>
<li>Working with biological sequences</li>
<li>Reproducible annotations</li>
<li>Storing and organsing data using the s4 data-class and an example
with <code>summarizedexperiment</code>
</li>
</ul>
</li>
<li>
<strong>Part B - application with biological data</strong> - work
through an example data set and packages from bioconductor for the
storage, structure, and visualization of datasets</li>
<li>
<strong>Part C - reproducibility &amp; communication</strong> - a
workflow using principals described in previous lessons to work through
some practices for reproducible workflows and strategies for
communicating and sharing results.</li>
</ul>
<p>Material in the first three lessons has been trimmed down, and we
direct you to the full lesson material available in <a href="https://carpentries-incubator.github.io/bioc-project/" class="external-link">The
Bioconductor project</a>. Note that the <a href="https://carpentries-incubator.github.io/bioc-project/07-genomic-ranges.html" class="external-link">Genomic
ranges</a> episode has been removed from content presented here.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Participants will only be able to install the version of
Bioconductor packages described in this lesson and reproduce their exact
outputs if they use the correct version of R.</li>
<li>The files used in this lesson should be downloaded in a local path
that is easily accessible from an R session.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-02-bioconductor"><p>Content from <a href="02-bioconductor.html">Bioconductor Introduction &amp; Install</a></p>
<hr>
<p>Last updated on 2025-11-11 |

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/episodes/02-bioconductor.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What does the Bioconductor project comprise?</li>
<li>How does the Bioconductor project relate to the CRAN
repository?</li>
<li>How can I learn to use Bioconductor packages effectively?</li>
<li>How do I join and communicate with the Bioconductor community?</li>
<li>How do I install Bioconductor packages?</li>
<li>How do I check if newer versions of my installed packages are
available?</li>
<li>How do I update Bioconductor packages?</li>
<li>How do I find out the name of packages available from the
Bioconductor repositories?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe the Bioconductor project globally.</li>
<li>Gain a global view of the Bioconductor project in the R
ecosystem.</li>
<li>Identify sources of information to watch for future updates about
the Bioconductor project.</li>
<li>Install BiocManager.</li>
<li>Install Bioconductor packages.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="what-is-bioconductor">What is Bioconductor?<a class="anchor" aria-label="anchor" href="#what-is-bioconductor"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="a-brief-history-of-bioconductor">A brief history of Bioconductor<a class="anchor" aria-label="anchor" href="#a-brief-history-of-bioconductor"></a>
</h3>
<p>The Bioconductor project was started in the Fall of 2001, as an
initiative for the collaborative creation of extensible software for
computational biology and bioinformatics (Gentleman, Carey, Bates,
Bolstad, Dettling, Dudoit, Ellis, Gautier, Ge, Gentry, Hornik, Hothorn,
Huber, Iacus, Irizarry, Leisch, Li, Maechler, Rossini, Sawitzki, Smith,
Smyth, Tierney, Yang, and Zhang, 2004). From the very start, the stated
mission of the project was to develop tools for the statistical analysis
and comprehension of large datasets and technological artifacts in
rigorously and robustly designed experiments. Beyond statistical
analyses, the interpretation of statistical results is supported by
packages providing biological context, visualization, and
reproducibility.</p>
<p>Over the years, software packages contributed to the Bioconductor
project have reflected the evolution and emergence of several
high-throughput technologies, from microarrays to single-cell genomics,
through many variations of sequencing experiments (e.g., RNA-seq,
ChIP-seq, DNA-seq), analyses (e.g., sequence variation, copy number
variation, single nucleotide polymorphisms), and data modalities (e.g.,
flow cytometry, proteomics, microscopy and image analysis).</p>
<p>Crucially, the project has not only released software packages
implementing novel statistical tests and methodologies, but also
produced a diverse range of packages types granting access to databases
of molecular annotations and experimental datasets.</p>
<p>The Bioconductor project culminates at an annual conference in North
America in the summer, while regional conferences offer great
opportunities for networking in Europe, Asia, and North America. The
project is committed to promote a diverse and inclusive community,
including a <a href="https://www.bioconductor.org/about/code-of-conduct/" class="external-link">Code of
Conduct</a> enforced by a Code of Conduct committee.</p>
<figure><img src="fig/bioc-timeline.svg" alt="Timeline of major Bioconductor milestones alongside technological advancements." class="figure mx-auto d-block"></figure><p><strong>Timeline of major Bioconductor milestones alongside
technological advancements.</strong> Above the timeline, the figure
marks the first occurence of major events. Within the timeline, the name
of packages providing core infrastructure indicate the release date.
Below the timeline, major technological advancements contextualise the
evolution of the Bioconductor project over time.</p>
</div>
<div class="section level3">
<h3 id="a-scientific-project">A scientific project<a class="anchor" aria-label="anchor" href="#a-scientific-project"></a>
</h3>
<p>The original publication describes the aims and methods of the
project at its inception Gentleman, Carey, Bates et al. (2004).</p>
<p>Huber, Carey, Gentleman, Anders, Carlson, Carvalho, Bravo, Davis,
Gatto, Girke, Gottardo, Hahne, Hansen, Irizarry, Lawrence, Love,
MacDonald, Obenchain, Oles, Pages, Reyes, Shannon, Smyth, Tenenbaum,
Waldron, and Morgan (2015) illustrates the progression of the project,
including descriptions of core infrastructure and case studies, from the
perspective of both users and developers.</p>
<p>Amezquita, Lun, Becht, Carey, Carpp, Geistlinger, Marini,
Rue-Albrecht, Risso, Soneson, Waldron, Pages, Smith, Huber, Morgan,
Gottardo, and Hicks (2020) reviewed further developments of the project
in the wake of single-cell genomics technologies.</p>
<p>Many more publications and book chapters cite the Bioconductor
project, with recent example listed on the <a href="https://www.bioconductor.org/help/publications/" class="external-link">Bioconductor
website</a>.</p>
</div>
<div class="section level3">
<h3 id="a-package-repository">A package repository<a class="anchor" aria-label="anchor" href="#a-package-repository"></a>
</h3>
<div class="section level4">
<h4 id="overview-and-relationship-to-cran">Overview and relationship to CRAN<a class="anchor" aria-label="anchor" href="#overview-and-relationship-to-cran"></a>
</h4>
<p>Undoubtedly, software packages are the best-known aspect of the
Bioconductor project. Since its inception in 2001, the repository has
grown over time to host thousands of packages.</p>
<p>The Bioconductor project has extended the preexisting CRAN repository
– much larger and general-purpose in scope – to comprise R packages
primarily catering for bioinformatics and computational biology
analyses.</p>
<div id="going-further" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="going-further" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>The <a href="discuss.html#bioconductor-and-CRAN">Discussion</a>
article of this lesson includes a section discussing the relationship of
Bioconductor and CRAN in further details.</p>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="the-bioconductor-release-cycle">The Bioconductor release cycle<a class="anchor" aria-label="anchor" href="#the-bioconductor-release-cycle"></a>
</h4>
<p>The Bioconductor project also extended the packaging infrastructure
of the CRAN repository to better support the deployment and management
of packages at the user level (Gentleman, Carey, Bates et al., 2004). In
particular, the Bioconductor projects features a 6-month release cycle
(typically around April and October), which sees a snapshot of the
current version of all packages in the Bioconductor repository earmarked
for a specific version of R. R itself is released on an annual basis
(typically around April), meaning that for each release of R, two
compatible releases of Bioconductor packages are available.</p>
<p>As such, Bioconductor package developers are required to always use
the version of R that will be associated with the next release of the
Bioconductor project. This means using the development version of R
between October and April, and the release version of R between April
and October.</p>
<p>Crucially, the strict Bioconductor release cycle prevents users from
installing temporally distant versions of packages that were very likely
never tested together. This practice reflects the development cycle of
packages of both CRAN and Bioconductor, where contemporaneous packages
are regularly tested by automated systems to ensure that the latest
software updates in package dependencies do not break downstream
packages, or prompts those package maintainers to update their own
software as a consequence.</p>
<p>Prior to each Bioconductor release, packages that do not pass the
requires suites of automated tests are deprecated and subsequently
removed from the repository. This ensures that each Bioconductor release
provides a suite of packages that are mutually compatible, traceable,
and guaranteed to function for the associated version of R.</p>
<figure><img src="fig/bioc-release-cycle.svg" alt="Timeline of release dates for selected Bioconductor and R versions." class="figure mx-auto d-block"></figure><p><strong>Timeline of release dates for selected Bioconductor and R
versions.</strong> The upper section of the timeline indicates versions
and approximate release dates for the R project. The lower section of
the timeline indicates versions and release dates for the Bioconductor
project. Source: <a href="https://bioconductor.org/about/release-announcements/" class="external-link">Bioconductor</a>.</p>
</div>
<div class="section level4">
<h4 id="package-types">Package types<a class="anchor" aria-label="anchor" href="#package-types"></a>
</h4>
<p>Packages are broadly divided in four major categories:</p>
<ul>
<li>software</li>
<li>annotation data</li>
<li>experiment data</li>
<li>workflows</li>
</ul>
<p><a href="reference.html#software-package">Software packages</a>
themselves can be subdivided into packages that provide infrastructure
(i.e., classes) to store and access data, and packages that provide
methodological tools to process data stored in those data structures.
This separation of structure and analysis is at the core of the
Bioconductor project, encouraging developers of new methodological
software packages to thoughtfully re-use existing data containers where
possible, and reducing the cognitive burden imposed on users who can
more easily experiment with alternative workflows without the need to
learn and convert between different data structures.</p>
<p><a href="reference.html#annotationdata-package">Annotation data
packages</a> provide self-contained databases of diverse genomic
annotations (e.g., gene identifiers, biological pathways). Different
collections of annotation packages can be found in the Bioconductor
project. They are identifiable by their respective naming pattern, and
the information that they contain. For instance, the so-called
<code>OrgDb</code> packages (e.g., the <em><a href="https://bioconductor.org/packages/3.22/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>
package) provide information mapping different types of gene identifiers
and pathway databases; the so-called <code>EnsDb</code> (e.g., <em><a href="https://bioconductor.org/packages/3.22/EnsDb.Hsapiens.v86" class="external-link">EnsDb.Hsapiens.v86</a></em>)
packages encapsulate individual versions of the Ensembl annotations in
Bioconductor packages; and the so-called <code>TxDb</code> packages
(e.g., <em><a href="https://bioconductor.org/packages/3.22/TxDb.Hsapiens.UCSC.hg38.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg38.knownGene</a></em>)
encapsulate individual versions UCSC gene annotation tables.</p>
<p><a href="reference.html#experimentdata-package">Experiment data
packages</a> provide self-contained datasets that are often used by
software package developers to demonstrate the use of their package on
well-known standard datasets in their <a href="reference.html#vignette">package vignettes</a>.</p>
<p>Finally, <a href="reference.html#workflow-package">workflow
packages</a> exclusively provide collections of vignettes that
demonstrate the combined usage of several other packages as a coherent
workflow, but do not provide any new source code or functionality
themselves.</p>
<div id="challenge-the-bioconductor-website" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-the-bioconductor-website" class="callout-inner">
<h3 class="callout-title">Challenge: The Bioconductor website</h3>
<div class="callout-content">
<p>The Bioconductor website is accessible at <a href="https://bioconductor.org/" class="external-link">https://bioconductor.org/</a>.</p>
<p>Browse the website to find information answering the following
questions:</p>
<ol style="list-style-type: decimal">
<li>How many packages does the current release of the Bioconductor
project include?</li>
<li>How many packages of each type does this number include?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>The following solution includes numbers that were valid at the time
of writing (Bioconductor release 3.13); numbers will inevitably be
different for future releases of the Bioconductor project.</p>
<ol style="list-style-type: decimal">
<li>On the page <a href="https://bioconductor.org/" class="external-link">https://bioconductor.org/</a>, in the
section “Install”, we can read:</li>
</ol>
<blockquote>
<p>Discover 2042 software packages available in Bioconductor release
3.13.</p>
</blockquote>
<ol start="2" style="list-style-type: decimal">
<li>On the page <a href="https://bioconductor.org/" class="external-link">https://bioconductor.org/</a>, in the
section “News”, click on the link that reads “Bioconductor Bioc
<code>X.Y</code> Released” (<code>X.Y</code> being the version of the
current Bioconductor release when you go through this exercise
yourself). On the linked page, we can read:</li>
</ol>
<blockquote>
<p>We are pleased to announce Bioconductor 3.13, consisting of 2042
software packages, 406 experiment data packages, 965 annotation
packages, and 29 workflows.</p>
<p>There are 133 new software packages, 22 new data experiment packages,
7 new annotation packages, 1 new workflow, no new books, and many
updates and improvements to existing packages; Bioconductor 3.13 is
compatible with R 4.1.0, and is supported on Linux, 32- and 64-bit
Windows, and macOS 10.14.6 Mojave or higher. This release will include
an updated Bioconductor Docker containers.</p>
</blockquote>
</div>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="package-classification-using-biocviews">Package classification using biocViews<a class="anchor" aria-label="anchor" href="#package-classification-using-biocviews"></a>
</h4>
<p>The Bioconductor project uses <a href="reference.html#biocviews">biocViews</a>, a set of terms from a
controlled vocabulary, to classify Bioconductor packages and facilitate
their discovery by thematic search on the <a href="https://www.bioconductor.org/packages/release/BiocViews.html" class="external-link">Bioconductor
website</a>.</p>
<p>Each Bioconductor package is tagged with a small set of terms chosen
from the available controlled vocabulary, to describe the type and
functionality of the package. Terms are initially selected by the
package authors, and subsequently refined during package review or
updates to the controlled vocabulary.</p>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Visit the listing of all packages on the Bioconductor <a href="https://www.bioconductor.org/packages/release/BiocViews.html" class="external-link">biocViews</a>
web page. Use the “Autocomplete biocViews search” box in the upper left
to filter packages by category and explore the graph of software
packages by expanding and contracting individual terms.</p>
<ol style="list-style-type: decimal">
<li>What biocView terms can be used to identify packages that have been
tagged for RNA sequencing analysis? ChIP-seq? Epigenetics? Variant
annotation? Proteomics? Single-cell genomics?</li>
<li>In the <code>RNASeq</code> category, two very popular packages are
<em><a href="https://bioconductor.org/packages/3.22/DESeq2" class="external-link">DESeq2</a></em> and
<em><a href="https://bioconductor.org/packages/3.22/edgeR" class="external-link">edgeR</a></em>.
Which one is more popular in terms of download statistics (i.e., lower
rank)?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<ol style="list-style-type: decimal">
<li>
<code>RNAseq</code>, <code>ChIPSeq</code>, <code>Epigenetics</code>,
<code>VariantAnnotation</code>, <code>Proteomics</code>,
<code>SingleCell</code>.</li>
<li>For Bioconductor release <code>3.14</code>, <em><a href="https://bioconductor.org/packages/3.22/DESeq2" class="external-link">DESeq2</a></em> and
<em><a href="https://bioconductor.org/packages/3.22/edgeR" class="external-link">edgeR</a></em> are
listed at ranks 23 and 28 respectively. In other words, the two packages
are among the most frequently downloaded packages in the Bioconductor
project, in this instance with a small advantage in favour of <em><a href="https://bioconductor.org/packages/3.22/edgeR" class="external-link">edgeR</a></em>.</li>
</ol>
</div>
</div>
</div>
</div>
<div id="going-further-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="going-further-1" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>The Bioconductor package <em><a href="https://bioconductor.org/packages/3.22/biocViews" class="external-link">biocViews</a></em>
is used to support and manage the infrastructure of the controlled
vocabulary. It can also be used to programmatically inspect and subset
the list of terms available using their relationship as a graph.</p>
<p>Furthermore, the <em><a href="https://bioconductor.org/packages/3.22/BiocPkgTools" class="external-link">BiocPkgTools</a></em>
package can be used to browse packages under different <a href="reference.html#biocviews">biocViews</a> (Su, Carey, Shepherd,
Ritchie, Morgan, and Davis, 2019).</p>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="packages-interoperability">Packages interoperability<a class="anchor" aria-label="anchor" href="#packages-interoperability"></a>
</h4>
<p>At the core of the Bioconductor philosophy is the notion of
interoperability. Namely, the capacity of packages to operate on the
same data structures. Importantly, interoperability benefits both users
and developers.</p>
<p>Users can more easily write arbitrarily complex workflows that
combine multiple packages. With packages operating on the same data
structure, users can maximize their attention to the practical steps of
their workflow, and minimize time spent in often complex and error-prone
conversions between different data structures specific to each package.
Comparative benchmarks are also easier to implement and can be evaluated
more fairly when competing software packages offering similar
functionality operate on input and outputs stored in the same data
structures.</p>
<p>Similarly, developers of new packages can focus on the implementation
of novel functionality, borrowing existing data structures that offer
robust and trusted infrastructure for storage, verification, and
indexing of information.</p>
<p>Ultimately, the figure below illustrates how many different
Bioconductor packages - as well as base R packages - can be combined to
carry out a diverse range of analyses, from importing sequencing data
into an R session, to the annotation, integration and visualization of
data and results.</p>
<figure><img src="fig/bioc-sequencing-ecosystem.svg" alt="Sequencing Ecosystem." class="figure mx-auto d-block"></figure><p><strong>Sequencing Ecosystem</strong> Major data processing steps
(blue) and relevant software packages (pink) are listed in the context
of archetypal workflows for various types of genomics analyses. The
sequential relation of workflow steps and software package illustrates
the importance of interoperability between software package in order to
assemble complete end-to-end workflows.</p>
</div>
</div>
<div class="section level3">
<h3 id="conferences-courses-and-workshops">Conferences, courses and workshops<a class="anchor" aria-label="anchor" href="#conferences-courses-and-workshops"></a>
</h3>
<p>The Bioconductor community regularly organizes a number of events
throughout the year and across the world. For example:</p>
<ul>
<li>The annual BioC summer conference in North America</li>
<li>Regional conference in winter (e.g. BioC Europe, BioC Asia)</li>
<li>Summer schools (e.g., CSAMA)</li>
<li>Online meetings open to all community members (e.g., Bioconductor
Developers Forum)</li>
</ul>
<p>Course materials are regularly uploaded on the <a href="https://bioconductor.org/help/course-materials/" class="external-link">Bioconductor
website</a> following each of those events. In particular, <a href="https://www.bioconductor.org/books/release/" class="external-link">online books</a> are
being developed and maintained by community members.</p>
<p>The Bioconductor <a href="https://www.youtube.com/user/bioconductor" class="external-link">YouTube channel</a> is
used to publish video recordings of conference presentations including
talks and workshops, as well as editions of the regular Bioconductor
developers forum (link needed).</p>
</div>
<div class="section level3">
<h3 id="online-communication-channels">Online communication channels<a class="anchor" aria-label="anchor" href="#online-communication-channels"></a>
</h3>
<div class="section level4">
<h4 id="support-site">Support site<a class="anchor" aria-label="anchor" href="#support-site"></a>
</h4>
<p>The Bioconductor <a href="https://support.bioconductor.org/" class="external-link">support
site</a> provides a platform where users and developers can communicate
freely (following the Bioconductor <a href="https://www.bioconductor.org/about/code-of-conduct/" class="external-link">Code of
Conduct</a>) to discuss issues on a range of subjects, ranging from
packages to conceptual questions about best practices.</p>
</div>
<div class="section level4">
<h4 id="slack-workspace">Slack workspace<a class="anchor" aria-label="anchor" href="#slack-workspace"></a>
</h4>
<p>The Bioconductor <a href="https://bioc-community.herokuapp.com/" class="external-link">Slack workspace</a> is an
open space that all community members are welcome to join (for free) and
use for rapid interactions. Currently, the “Pro” pricing plan kindly
supported by core funding provides:</p>
<ul>
<li>Unlimited message archive</li>
<li>Unlimited apps</li>
<li>Group video calls with screen sharing</li>
<li>Work securely with other organizations using Slack Connect</li>
</ul>
<p>A wide range of channels have been created to discuss a range of
subjects, and community members can freely join the discussion on those
channels of create new ones to discuss new subjects.</p>
<p>Important announcements are posted on the <code>#general</code>
channel.</p>
<div id="note" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="note" class="callout-inner">
<h3 class="callout-title">Note</h3>
<div class="callout-content">
<p>Users are encouraged to use the Bioconductor <a href="https://support.bioconductor.org/" class="external-link">support site</a> to raise
issues that are relevant to the wider community. The Slack workspace is
often most useful for live discussions, and widely subscribed channels
(e.g. <code>#general</code>) should be used with moderation.</p>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="developer-mailing-list">Developer Mailing List<a class="anchor" aria-label="anchor" href="#developer-mailing-list"></a>
</h4>
<p>The <a href="mailto:bioc-devel@r-project.org">bioc-devel@r-project.org</a>
mailing list is used for communication between package developers, and
announcements from the Biocondutor core team.</p>
</div>
</div>
<div class="section level3">
<h3 id="a-scientific-and-technical-community">A scientific and technical community<a class="anchor" aria-label="anchor" href="#a-scientific-and-technical-community"></a>
</h3>
<ul>
<li>
<a href="https://bioconductor.org/about/scientific-advisory-board/" class="external-link">Scientific
Advisory Board (SAB)</a> Meet Annually, External and Internal leader in
the field who act as project advisors. No Term limits.</li>
<li>
<a href="https://bioconductor.org/about/technical-advisory-board/" class="external-link">Technical
Advisory Board (TAB)</a>. Meet monthly to consider technical aspects of
core infastructure and scientific direction of the project. 15 members,
3 year term. Annual open-to-all elections to rotate members. Current
officers are Vince Carey (chair), Levi Waldron (vice Chair) Charlotte
Soneson (Secretary).</li>
<li>
<a href="https://bioconductor.org/about/community-advisory-board/" class="external-link">Community
Advisory Board (CAB)</a> Meet monthly to consider community outreach,
events, education and training. 15 members, 3 year term. Annual
open-to-all elections to rotate members. Current officers are Aedin
Culhane (chair), Matt Ritchie (co Chair), Lori Kern (Secretary).</li>
<li><a href="https://www.bioconductor.org/about/code-of-conduct/" class="external-link">Code
of Conduct committee</a></li>
</ul>
<div id="note-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="note-1" class="callout-inner">
<h3 class="callout-title">Note</h3>
<div class="callout-content">
<p>At least 1 member of TAB/CAB sits on both to act at the liason to
ensure communication of the board.</p>
</div>
</div>
</div>
<hr>
</div>
</section><section><h2 class="section-heading" id="biocmanager">BiocManager<a class="anchor" aria-label="anchor" href="#biocmanager"></a>
</h2>
<hr class="half-width">
<p>The <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em>
package is the entry point into the Bioconductor package repository.
Technically, this is the only Bioconductor package distributed on the
CRAN repository.</p>
<p>It provides functions to safely install Bioconductor packages and
check for available updates.</p>
<p>Once the package is installed, the function
<code>BiocManager::install()</code> can be used to install packages from
the Bioconductor repository. The function is also capable of installing
packages from other repositories (e.g., CRAN), if those packages are not
found in the Bioconductor repository first.</p>
<figure><img src="fig/bioc-install.svg" alt="The package BiocManager is available from the CRAN repository and used to install packages from the Bioconductor repository." class="figure mx-auto d-block"></figure><p><strong>The package BiocManager is available from the CRAN repository
and used to install packages from the Bioconductor repository.</strong>
The function <code>install.packages()</code> from the base R package
<code>utils</code> can be used to install the <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em>
package distributed on the CRAN repository. In turn, the function
<code>BiocManager::install()</code> can be used to install packages
available on the Bioconductor repository. Notably, the
<code>BiocManager::install()</code> function will fall back on the CRAN
repository if a package cannot be found in the Bioconductor
repository.</p>
<p>Install the package using the code below.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre>
</div>
<div id="going-further-2" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="going-further-2" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>A number of packages that are not part of the base R installation
also provide functions to install packages from various repositories.
For instance:</p>
<ul>
<li><code>devtools::install()</code></li>
<li><code>remotes::install_bioc()</code></li>
<li><code>remotes::install_bitbucket()</code></li>
<li><code>remotes::install_cran()</code></li>
<li><code>remotes::install_dev()</code></li>
<li><code>remotes::install_github()</code></li>
<li><code>remotes::install_gitlab()</code></li>
<li><code>remotes::install_git()</code></li>
<li><code>remotes::install_local()</code></li>
<li><code>remotes::install_svn()</code></li>
<li><code>remotes::install_url()</code></li>
<li><code>renv::install()</code></li>
</ul>
<p>Those functions are beyond the scope of this lesson, and should be
used with caution and adequate knowledge of their specific behaviors.
The general recommendation is to use <code>BiocManager::install()</code>
over any other installation mechanism because it ensures proper
versioning of Bioconductor packages.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="bioconductor-releases-and-current-version">Bioconductor releases and current version<a class="anchor" aria-label="anchor" href="#bioconductor-releases-and-current-version"></a>
</h2>
<hr class="half-width">
<p>Once the <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em>
package is installed, the <code>BiocManager::version()</code> function
displays the version (i.e., release) of the Bioconductor project that is
currently active in the R session.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">version</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] '3.22'</code></pre>
</div>
<p>Using the correct version of R and Bioconductor packages is a key
aspect of reproducibility. The <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em>
packages uses the version of R running in the current session to
determine the version of Biocondutor packages that can be installed in
the current R library.</p>
<p>The Bioconductor project produces two releases each year, one around
April and another one around October. The April release of Bioconductor
coincides with the annual release of R. The October release of
Bioconductor continues to use the same version of R for that annual
cycle (i.e., until the next release, in April).</p>
<figure><img src="fig/bioc-release-cycle.svg" alt="Timeline of release dates for selected Bioconductor and R versions." class="figure mx-auto d-block"></figure><p><strong>Timeline of release dates for selected Bioconductor and R
versions.</strong> The upper section of the timeline indicates versions
and approximate release dates for the R project. The lower section of
the timeline indicates versions and release dates for the Bioconductor
project. Source: <a href="https://bioconductor.org/about/release-announcements/" class="external-link">Bioconductor</a>.</p>
<p>During each 6-month cycle of package development, Bioconductor tests
packages for compatibility with the version of R that will be available
for the next release cycle. Then, each time a new Bioconductor release
is produced, the version of every package in the Bioconductor repository
is incremented, including the package <em><a href="https://bioconductor.org/packages/3.22/BiocVersion" class="external-link">BiocVersion</a></em>
which determines the version of the Bioconductor project.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">packageVersion</span><span class="op">(</span><span class="st">"BiocVersion"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] '3.22.0'</code></pre>
</div>
<p>This is the case for every package, even those which have not been
updated at all since the previous release. That new version of each
package is earmarked for the corresponding version of R; in other words,
that version of the package can only be installed and accessed in an R
session that uses the correct version of R. This version increment is
essential to associate a each version of a Bioconductor package with a
unique release of the Bioconductor project.</p>
<p>Following the April release, this means that users must install the
new version of R to access the newly released versions of Bioconductor
packages.</p>
<p>Instead, in October, users can continue to use the same version of R
to access the newly released version of Bioconductor packages. However,
to update an R library from the April release to the October release of
Bioconductor, users need to call the function
<code>BiocManager::install()</code> specifying the correct version of
Bioconductor as the <code>version</code> option, for instance:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.14"</span><span class="op">)</span></span></code></pre>
</div>
<p>This needs to be done only once, as the <em><a href="https://bioconductor.org/packages/3.22/BiocVersion" class="external-link">BiocVersion</a></em>
package will be updated to the corresponding version, indicating the
version of Bioconductor in use in this R library.</p>
<div id="going-further-3" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="going-further-3" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>The <a href="discuss.html#the-bioconductor-release-cycle">Discussion</a>
article of this lesson includes a section discussing the release cycle
of the Bioconductor project.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="check-for-updates">Check for updates<a class="anchor" aria-label="anchor" href="#check-for-updates"></a>
</h2>
<hr class="half-width">
<p>The <code>BiocManager::valid()</code> function inspects the version
of packages currently installed in the user library, and checks whether
a new version is available for any of them on the Bioconductor
repository.</p>
<p>If everything is up-to-date, the function will simply print
<code>TRUE</code>.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">valid</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: 1 packages out-of-date; 0 packages too new</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
* sessionInfo()

R version 4.5.2 (2025-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] bibtex_0.5.1     RefManageR_1.4.0 BiocStyle_2.38.0

loaded via a namespace (and not attached):
 [1] httr_1.4.7          cli_3.6.5           knitr_1.50
 [4] rlang_1.1.6         xfun_0.54           stringi_1.8.7
 [7] renv_1.1.5          generics_0.1.4      jsonlite_2.0.0
[10] glue_1.8.0          backports_1.5.0     plyr_1.8.9
[13] htmltools_0.5.8.1   rmarkdown_2.29      evaluate_1.0.5
[16] fastmap_1.2.0       yaml_2.3.10         lifecycle_1.0.4
[19] BiocManager_1.30.26 stringr_1.6.0       compiler_4.5.2
[22] Rcpp_1.1.0          timechange_0.3.0    digest_0.6.37
[25] R6_2.6.1            magrittr_2.0.4      tools_4.5.2
[28] lubridate_1.9.4     xml2_1.4.1

Bioconductor version '3.22'

  * 1 packages out-of-date
  * 0 packages too new

create a valid installation with

  BiocManager::install("rmarkdown", update = TRUE, ask = FALSE, force = TRUE)

more details: BiocManager::valid()$too_new, BiocManager::valid()$out_of_date</code></pre>
</div>
<p>Conveniently, if any package can be updated, the function generates
and displays the command needed to update those packages. Users simply
need to copy-paste and run that command in their R console.</p>
<div id="example-of-out-of-date-package-library" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="example-of-out-of-date-package-library" class="callout-inner">
<h3 class="callout-title">Example of out-of-date package library</h3>
<div class="callout-content">
<p>In the example below, the <code>BiocManager::valid()</code> function
did not return <code>TRUE</code>. Instead, it includes information about
the active user session, and displays the exact call to
<code>BiocManager::install()</code> that the user should run to replace
all the outdated packages detected in the user library with the latest
version available in CRAN or Bioconductor.</p>
<pre><code>&gt; BiocManager::valid()

* sessionInfo()

R version 4.1.0 (2021-05-18)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 11.6

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base

loaded via a namespace (and not attached):
[1] BiocManager_1.30.16 compiler_4.1.0      tools_4.1.0         renv_0.14.0

Bioconductor version '3.13'

  * 18 packages out-of-date
  * 0 packages too new

create a valid installation with

  BiocManager::install(c(
    "cpp11", "data.table", "digest", "hms", "knitr", "lifecycle", "matrixStats", "mime", "pillar", "RCurl",
    "readr", "remotes", "S4Vectors", "shiny", "shinyWidgets", "tidyr", "tinytex", "XML"
  ), update = TRUE, ask = FALSE)

more details: BiocManager::valid()$too_new, BiocManager::valid()$out_of_date

Warning message:
18 packages out-of-date; 0 packages too new </code></pre>
<p>Specifically, in this example, the message tells the user to run the
following command to bring their installation up to date:</p>
<pre><code><span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"cpp11"</span>, <span class="st">"data.table"</span>, <span class="st">"digest"</span>, <span class="st">"hms"</span>, <span class="st">"knitr"</span>, <span class="st">"lifecycle"</span>, <span class="st">"matrixStats"</span>, <span class="st">"mime"</span>, <span class="st">"pillar"</span>, <span class="st">"RCurl"</span>,</span>
<span>    <span class="st">"readr"</span>, <span class="st">"remotes"</span>, <span class="st">"S4Vectors"</span>, <span class="st">"shiny"</span>, <span class="st">"shinyWidgets"</span>, <span class="st">"tidyr"</span>, <span class="st">"tinytex"</span>, <span class="st">"XML"</span></span>
<span>  <span class="op">)</span>, update <span class="op">=</span> <span class="cn">TRUE</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exploring-the-package-repository">Exploring the package repository<a class="anchor" aria-label="anchor" href="#exploring-the-package-repository"></a>
</h2>
<hr class="half-width">
<p>The Bioconductor <a href="reference.html#biocviews">biocViews</a>,
demonstrated in the earlier episode <a href="https://carpentries-incubator.github.io/bioc-project/02-introduction-to-bioconductor/index.html#package-classification-using-biocviews" class="external-link">Introduction
to Bioconductor</a>, are a great way to discover new packages by
thematically browsing the hierarchical classification of Bioconductor
packages.</p>
<p>In addition, the <code>BiocManager::available()</code> function
returns the complete list of package names that are can be installed
from the Bioconductor and CRAN repositories. For instance the total
number of numbers that could be installed using <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em></p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">available</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 26624</code></pre>
</div>
<p>Specifically, the union of current Bioconductor repositories and
other repositories on the search path can be displayed as follows.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">repositories</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span>                                                <span class="va">BioCsoft</span></span>
<span>           <span class="st">"https://bioconductor.org/packages/3.22/bioc"</span></span>
<span>                                                 <span class="va">BioCann</span></span>
<span><span class="st">"https://bioconductor.org/packages/3.22/data/annotation"</span></span>
<span>                                                 <span class="va">BioCexp</span></span>
<span><span class="st">"https://bioconductor.org/packages/3.22/data/experiment"</span></span>
<span>                                           <span class="va">BioCworkflows</span></span>
<span>      <span class="st">"https://bioconductor.org/packages/3.22/workflows"</span></span>
<span>                                               <span class="va">BioCbooks</span></span>
<span>          <span class="st">"https://bioconductor.org/packages/3.22/books"</span></span>
<span>                                             <span class="va">carpentries</span></span>
<span>                    <span class="st">"https://carpentries.r-universe.dev"</span></span>
<span>                                     <span class="va">carpentries_archive</span></span>
<span>                    <span class="st">"https://carpentries.github.io/drat"</span></span>
<span>                                                    <span class="va">CRAN</span></span>
<span>                              <span class="st">"https://cran.rstudio.com"</span> </span></code></pre>
</div>
<p>Each repository URL can be accessed in a web browser, displaying the
full list of packages available from that repository. For instance,
navigate to <a href="https://bioconductor.org/packages/3.14/bioc" class="external-link">https://bioconductor.org/packages/3.14/bioc</a>.</p>
<div id="going-further-4" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="going-further-4" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>The function <code>BiocManager::repositories()</code> can be combined
with the base function <code>available.packages()</code> to query
packages available specifically from any package repository, e.g. the
Bioconductor <a href="reference.html#software-package">software
package</a> repository.</p>
<pre><code>&gt; db = available.packages(repos = BiocManager::repositories()["BioCsoft"])
&gt; dim(db)
[1] 1948   17
&gt; head(rownames(db))
[1] "a4"          "a4Base"      "a4Classif"   "a4Core"      "a4Preproc"
[6] "a4Reporting"</code></pre>
</div>
</div>
</div>
<p>Conveniently, <code>BiocManager::available()</code> includes a
<code>pattern=</code> argument, particularly useful to navigate
annotation resources (the original use case motivating it). For
instance, a range of <a href="reference.html#annotationdata-package">Annotation data
packages</a> available for the mouse model organism can be listed as
follows.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">available</span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"*Mmusculus"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "BSgenome.Mmusculus.UCSC.mm10"        "BSgenome.Mmusculus.UCSC.mm10.masked"
 [3] "BSgenome.Mmusculus.UCSC.mm39"        "BSgenome.Mmusculus.UCSC.mm8"
 [5] "BSgenome.Mmusculus.UCSC.mm8.masked"  "BSgenome.Mmusculus.UCSC.mm9"
 [7] "BSgenome.Mmusculus.UCSC.mm9.masked"  "EnsDb.Mmusculus.v75"
 [9] "EnsDb.Mmusculus.v79"                 "PWMEnrich.Mmusculus.background"
[11] "TxDb.Mmusculus.UCSC.mm10.ensGene"    "TxDb.Mmusculus.UCSC.mm10.knownGene"
[13] "TxDb.Mmusculus.UCSC.mm39.knownGene"  "TxDb.Mmusculus.UCSC.mm39.refGene"
[15] "TxDb.Mmusculus.UCSC.mm9.knownGene"  </code></pre>
</div>
</section><section><h2 class="section-heading" id="installing-packages">Installing packages<a class="anchor" aria-label="anchor" href="#installing-packages"></a>
</h2>
<hr class="half-width">
<p>The <code>BiocManager::install()</code> function is used to install
or update packages.</p>
<p>The function takes a character vector of package names, and attempts
to install them from the Bioconductor repository.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"S4Vectors"</span>, <span class="st">"BiocGenerics"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>However, if any package cannot be found in the Bioconductor
repository, the function also searches for those packages in
repositories listed in the global option <code>repos</code>.</p>
<div id="contribute" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="contribute" class="callout-inner">
<h3 class="callout-title">Contribute !</h3>
<div class="callout-content">
<p>Add an example of non-Bioconductor package that can be installed
using BioManager. Preferably, a package that will be used later in this
lesson.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="uninstalling-packages">Uninstalling packages<a class="anchor" aria-label="anchor" href="#uninstalling-packages"></a>
</h2>
<hr class="half-width">
<p>Bioconductor packages can be removed from the R library like any
other R package, using the base R function
<code>remove.packages()</code>. In essence, this function simply removes
installed packages and updates index information as necessary. As a
result, it will not be possible to attach the package to a session or
browse the documentation of that package anymore.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remove.packages</span><span class="op">(</span><span class="st">"S4Vectors"</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="getting-help">Getting help<a class="anchor" aria-label="anchor" href="#getting-help"></a>
</h2>
<hr class="half-width">
<p>For time we will not go over the full episode on <a href="https://carpentries-incubator.github.io/bioc-project/04-getting-help.html" class="external-link">Getting
Help</a>, instead below is a some summary of key points and sources:</p>
<p><strong>The Bioconductor website</strong></p>
<p>The main [Bioconductor website][bioc-website] provides a host of
resources, all freely available without even the need to install R or
any Bioconductor package.</p>
<p>In particular, the <a href="https://www.bioconductor.org/packages/release/BiocViews.html" class="external-link">biocViews</a>
page is a great way to thematically explore the collection of packages
and identify Bioconductor packages providing a certain
functionality.</p>
<p>Furthermore, the website also collects materials from [courses and
conferences][bioc-courses-conferences-materials], including
presentations, video recordings, and teaching materials.</p>
<p><strong>Package landing pages</strong></p>
<p>Each package accepted in the Bioconductor project is granted a
landing page on the main Bioconductor website, e.g. <em><a href="https://bioconductor.org/packages/3.22/S4Vectors" class="external-link">S4Vectors</a></em>.</p>
<p>Package landing pages contains useful information that can be
consulted without the need to install the package itself. This is
particularly useful while browsing the Bioconductor repository is search
of packages suitable for a specific task.</p>
<p><strong>Package vignettes</strong></p>
<p>When Bioconductor packages are installed in the user’s R library, the
vignettes associated with that particular version of the package are
also installed on the user’s computer. Those locally installed vignettes
are the gold standard reference for the version of the package that is
currently installed in the R library and used in the R session. They can
be accessed using the function <code>browseVignettes()</code>, for
instance:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">browseVignettes</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre>
</div>
<p><strong>Package help pages</strong></p>
<p>Bioconductor requires every user-facing package function to be
documented in one of the package help pages (often referred to as “man
pages”, after the name of the package sub-directory where they are
stored).</p>
<p>Help pages can be accessed using the <code>help()</code> function or
the question mark symbol <code>?</code>.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">help</span><span class="op">(</span>topic <span class="op">=</span> <span class="st">"install"</span>, package <span class="op">=</span> <span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">?</span><span class="fu">BiocManager</span><span class="fu">::</span><span class="va">install</span></span></code></pre>
</div>
<p><strong>The Bioconductor support site</strong></p>
<p>The <a href="https://support.bioconductor.org/" class="external-link">Bioconductor support
site</a> provides a platform for the community of users and developers
to ask questions, and help each other through the doubts and challenges
related to Bioconductor packages and analytical workflows.</p>
<p><strong>Workflow packages</strong></p>
<p>Bioconductor workflow packages are special in the way that they are
only expected to contain vignettes, without any additional code or
functionality of their own. Instead, the vignettes of workflow packages
exclusively import functionality from other packages, and demonstrate
how to combine functions from those packages into an integrated workflow
that users are likely to face in their day-to-day work.</p>
<p>Workflow packages can be browsed in a dedicated section of the
[biocViews page][biocviews-workflow].</p>
<p><strong>Slack workspace</strong></p>
<p>The <a href="https://bioc-community.herokuapp.com/" class="external-link">Bioconductor
Slack workspace</a> was created in 2016.</p>
<p>The workspace can be freely joined using this [Heroku
app][bioc-slack-heroku] to generate invitations for individual email
addresses.</p>
<div class="section level3">
<h3 id="how-to-efficiently-ask-for-help">How to efficiently ask for help<a class="anchor" aria-label="anchor" href="#how-to-efficiently-ask-for-help"></a>
</h3>
<p>Most community members provide help voluntarily, on their own spare
time and without any form of compensation. As such, it is important to
ask questions as clearly as possible, providing relevant information, to
help those volunteers identify the source of the issue as rapidly as
possible, thus saving time for both the helper and the original
poster.</p>
<p>Depending on the question, some key information includes:</p>
<ul>
<li>Operating system</li>
<li>Version of R</li>
<li>Version of Bioconductor</li>
<li>Version of individual packages installed in the R library</li>
<li>Version of individual packages attached to the current session</li>
<li>Third-party software, libraries, and compilers installed on the
user’s system</li>
<li>Source of packages installed in the R library (e.g., Bioconductor,
CRAN, GitHub)</li>
<li>Code executed in the R session leading up to the issue</li>
<li>Global options active in the session (accessible using
<code>options()</code>)</li>
</ul>
</div>
</section><section><h2 class="section-heading" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<hr class="half-width">
<p>[1] R. A. Amezquita, A. T. L. Lun, E. Becht, et al. “Orchestrating
single-cell analysis with Bioconductor”. In: <em>Nat Methods</em> 17.2
(2020), pp. 137-145. ISSN: 1548-7105 (Electronic) 1548-7091 (Linking).
DOI: 10.1038/s41592-019-0654-x. <a href="https://www.ncbi.nlm.nih.gov/pubmed/31792435" class="external-link uri">https://www.ncbi.nlm.nih.gov/pubmed/31792435</a>.</p>
<p>[2] R. C. Gentleman, V. J. Carey, D. M. Bates, et al. “Bioconductor:
open software development for computational biology and bioinformatics”.
In: <em>Genome Biol</em> 5.10 (2004), p. R80. ISSN: 1474-760X
(Electronic) 1474-7596 (Linking). DOI: 10.1186/gb-2004-5-10-r80. <a href="https://www.ncbi.nlm.nih.gov/pubmed/15461798" class="external-link uri">https://www.ncbi.nlm.nih.gov/pubmed/15461798</a>.</p>
<p>[3] W. Huber, V. J. Carey, R. Gentleman, et al. “Orchestrating
high-throughput genomic analysis with Bioconductor”. In: <em>Nat
Methods</em> 12.2 (2015), pp. 115-21. ISSN: 1548-7105 (Electronic)
1548-7091 (Linking). DOI: 10.1038/nmeth.3252. <a href="https://www.ncbi.nlm.nih.gov/pubmed/25633503" class="external-link uri">https://www.ncbi.nlm.nih.gov/pubmed/25633503</a>.</p>
<p>[4] S. Su, V. Carey, L. Shepherd, et al. “BiocPkgTools: Toolkit for
mining the Bioconductor package ecosystem [version 1; peer review: 2
approved, 1 approved with reservations]”. In: <em>F1000Research</em>
8.752 (2019). DOI: 10.12688/f1000research.19410.1.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>R packages are but one aspect of the Bioconductor project.</li>
<li>The Bioconductor project extends and complements the CRAN
repository.</li>
<li>Different types of packages provide not only software, but also
annotations, experimental data, and demonstrate the use of multiple
packages in integrated workflows.</li>
<li>Interoperability beteen Bioconductor packages facilitates the
writing of integrated workflows and minimizes the cognitive burden on
users.</li>
<li>Educational materials from courses and conferences are archived and
accessible on the Bioconductor website and YouTube channel.</li>
<li>Different channels of communication enable community members to
converse and help each other, both as users and package developers.</li>
<li>The Bioconductor project is governed by scientific, technical, and
advisory boards, as well as a Code of Conduct committee.</li>
<li>The BiocManager package is available from the CRAN repository.</li>
<li>
<code>BiocManager::install()</code> is used to install and update
Bioconductor packages (but also from CRAN and GitHub).</li>
<li>
<code>BiocManager::valid()</code> is used to check for available
package updates.</li>
<li>
<code>BiocManager::version()</code> reports the version of
Bioconductor currently installed.</li>
<li>
<code>BiocManager::install()</code> can also be used to update an
entire R library to a specific version of Bioconductor.</li>
<li>The <code>browseVignettes()</code> function is recommended to access
the vignette(s) installed with each package.</li>
<li>The Bioconductor main website contains general information, package
documentation, and course materials.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-03-biological-sequences"><p>Content from <a href="03-biological-sequences.html">Biological sequences</a></p>
<hr>
<p>Last updated on 2025-11-11 |

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/episodes/03-biological-sequences.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is the recommended way to represent biological sequences in
Bioconductor?</li>
<li>What Bioconductor packages provides methods to efficiently process
biological sequences?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain how biological sequences are represented in the Bioconductor
project.</li>
<li>Identify Bioconductor packages and methods available to process
biological sequences.</li>
<li>Explain how gene annotations are managed in the Bioconductor
project.</li>
<li>Identify Bioconductor packages and methods available to fetch and
use gene annotations.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="install-packages">Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a>
</h2>
<hr class="half-width">
<p>Before we can proceed into the following sections, we install some
Bioconductor packages that we will need. First, we check that the <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em>
package is installed before trying to use it; otherwise we install it.
Then we use the <code>BiocManager::install()</code> function to install
the necessary packages.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">requireNamespace</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"Biostrings"</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="the-biostrings-package-and-classes">The Biostrings package and classes<a class="anchor" aria-label="anchor" href="#the-biostrings-package-and-classes"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="why-do-we-need-classes-for-biological-sequences">Why do we need classes for biological sequences?<a class="anchor" aria-label="anchor" href="#why-do-we-need-classes-for-biological-sequences"></a>
</h3>
<p>Biological sequences are arguably some of the simplest biological
entities to represent computationally. Examples include nucleic acid
sequences (e.g., DNA, RNA) and protein sequences composed of amino
acids.</p>
<p>That is because alphabets have been designed and agreed upon to
represent individual monomers using character symbols.</p>
<p>For instance, using the alphabet for amino acids, the reference
protein sequence for the <a href="https://www.uniprot.org/uniprot/P68133#sequences" class="external-link">Actin, alpha
skeletal muscle protein sequence</a> is represented as follows.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "MCDEDETTALVCDNGSGLVKAGFAGDDAPRAVFPSIVGRPRHQGVMVGMGQKDSYVGDEAQSKRGILTLKYPIEHGIITNWDDMEKIWHHTFYNELRVAPEEHPTLLTEAPLNPKANREKMTQIMFETFNVPAMYVAIQAVLSLYASGRTTGIVLDSGDGVTHNVPIYEGYALPHAIMRLDLAGRDLTDYLMKILTERGYSFVTTAEREIVRDIKEKLCYVALDFENEMATAASSSSLEKSYELPDGQVITIGNERFRCPETLFQPSFIGMESAGIHETTYNSIMKCDIDIRKDLYANNVMSGGTTMYPGIADRMQKEITALAPSTMKIKIIAPPERKYSVWIGGSILASLSTFQQMWITKQEYDEAGPSIVHRKCF"</code></pre>
</div>
<p>However, a major limitation of regular character vectors is that they
do not check the validity of the sequences that they contain.
Practically, it is possible to store meaningless sequences of symbols in
character strings, including symbols that are not part of the official
alphabet for the relevant type of polymer. In those cases, the burden of
checking the validity of sequences falls on the programs that process
them, or causing those programs to run into errors when they
unexpectedly encounter invalid symbols in a sequence.</p>
<p>Instead, <a href="reference.html#s4-class">S4 classes</a> –
demonstrated in the earlier episode <a href="05-s4.html">The S4 class
system</a> – provide a way to label objects as distinct “DNA”, “RNA”, or
“protein” varieties of biological sequences. This label is an extremely
powerful way to inform programs on the set of character symbols they can
expect in the sequence, but also the range of computational operations
that can be applied to those sequences. For instance, a function
designed to translate nucleic acid sequences into the corresponding
amino acid sequence should only be allowed to run on sequences that
represent nucleic acids.</p>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Can you tell whether this character string is a valid DNA
sequence?</p>
<pre><code><span><span class="va">AATTGGCCRGGCCAATT</span></span></code></pre>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Yes, this is a valid DNA sequence using ambiguity codes defined in
the <a href="https://en.wikipedia.org/wiki/Nucleic_acid_notation#IUPAC_notation" class="external-link">IUPAC</a>
notation. In this case, <code>A</code>, <code>T</code>, <code>C</code>,
and <code>G</code> represents the four standard types of nucleotides,
while the <code>R</code> symbol acts as a regular expression
representing either of the two purine nucleotide bases, <code>A</code>
and <code>G</code>.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="the-biostrings-package">The Biostrings package<a class="anchor" aria-label="anchor" href="#the-biostrings-package"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>In the Bioconductor project, the <em><a href="https://bioconductor.org/packages/3.22/Biostrings" class="external-link">Biostrings</a></em>
package implements S4 classes to represent biological sequences as S4
objects, e.g. <code>DNAString</code> for sequences of nucleotides in
deoxyribonucleic acid polymers, and <code>AAString</code> for sequences
of amino acids in protein polymers. Those S4 classes provide
memory-efficient containers for character strings, automatic
validity-checking functionality for each class of biological molecules,
and methods implementing various string matching algorithms and other
utilities for fast manipulation and processing of large biological
sequences or sets of sequences.</p>
<p>A short presentation of the basic classes defined in the <em><a href="https://bioconductor.org/packages/3.22/Biostrings" class="external-link">Biostrings</a></em>
package is available in one of the package vignettes, accessible as
<code>vignette("Biostrings2Classes")</code>, while more detailed
information is provided in the other package vignettes, accessible as
<code>browseVignettes("Biostrings")</code>.</p>
</div>
<div class="section level3">
<h3 id="first-steps">First steps<a class="anchor" aria-label="anchor" href="#first-steps"></a>
</h3>
<p>To get started, we load the package.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">Biostrings</span><span class="op">)</span></span></code></pre>
</div>
<p>With the package loaded and attached to the session, we have access
to the package functions. Those include functions that let us create new
objects of the classes defined in the package. For instance, we can
create an object that represents a DNA sequence, using the
<code>DNAString()</code> constructor function. Without assigning the
output to an object, we let the resulting object be printed in the
console.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DNAString</span><span class="op">(</span><span class="st">"ATCG"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>4-letter DNAString object
seq: ATCG</code></pre>
</div>
<p>Notably, DNA sequences may only contain the symbols <code>A</code>,
<code>T</code>, <code>C</code>, and <code>G</code>, to represent the
four DNA nucleotide bases, the symbol <code>N</code> as a placeholder
for an unknown or unspecified base, and a restricted set of additional
symbols with special meaning defined in the <a href="https://www.bioinformatics.org/sms/iupac.html" class="external-link">IUPAC Extended
Genetic Alphabet</a>. Notice that the constructor function does not let
us create objects that contain invalid characters,
e.g. <code>Z</code>.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DNAString</span><span class="op">(</span><span class="st">"ATCGZ"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in .Call2("new_XString_from_CHARACTER", class(x0), string, start, : key 90 (char 'Z') not in lookup table</code></pre>
</div>
<p>Specifically, the <a href="https://www.bioinformatics.org/sms/iupac.html" class="external-link">IUPAC Extended
Genetic Alphabet</a> defines ambiguity codes that represent sets of
nucleotides, in a way similar to regular expressions. The
<code>IUPAC_CODE_MAP</code> named character vector contains the mapping
from the IUPAC nucleotide ambiguity codes to their meaning.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">IUPAC_CODE_MAP</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     A      C      G      T      M      R      W      S      Y      K      V
   "A"    "C"    "G"    "T"   "AC"   "AG"   "AT"   "CG"   "CT"   "GT"  "ACG"
     H      D      B      N
 "ACT"  "AGT"  "CGT" "ACGT" </code></pre>
</div>
<p>Any of those nucleotide codes are allowed in the sequence of a
<code>DNAString</code> object. For instance, the symbol <code>M</code>
represents either of the two nucleotides <code>A</code> or
<code>C</code> at a given position in a nucleic acid sequence.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DNAString</span><span class="op">(</span><span class="st">"ATCGM"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>5-letter DNAString object
seq: ATCGM</code></pre>
</div>
<p>In particular, pattern matching methods implemented in the <em><a href="https://bioconductor.org/packages/3.22/Biostrings" class="external-link">Biostrings</a></em>
package recognize the meaning of ambiguity codes for each class of
biological sequence, allowing them to efficiently match motifs queried
by users without the need to design elaborate regular expressions. For
instance, the method <code>matchPattern()</code> takes a
<code>pattern=</code> and a <code>subject=</code> argument, and returns
a <code>Views</code> object that reports and displays any match of the
<code>pattern</code> expression at any position in the
<code>subject</code> sequence.</p>
<p>Note that the default option <code>fixed = TRUE</code> instructs the
method to match the query exactly – i.e., ignore ambiguity codes – which
in this case does not report any exact match.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dna1</span> <span class="op">&lt;-</span> <span class="fu">DNAString</span><span class="op">(</span><span class="st">"ATCGCTTTGA"</span><span class="op">)</span></span>
<span><span class="fu">matchPattern</span><span class="op">(</span><span class="st">"GM"</span>, <span class="va">dna1</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Views on a 10-letter DNAString subject
subject: ATCGCTTTGA
views: NONE</code></pre>
</div>
<p>Instead, to indicate that the pattern includes some ambiguity code,
the argument <code>fixed</code> must be set to <code>FALSE</code>.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">matchPattern</span><span class="op">(</span><span class="st">"GM"</span>, <span class="va">dna1</span>, fixed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Views on a 10-letter DNAString subject
subject: ATCGCTTTGA
views:
      start end width
  [1]     4   5     2 [GC]
  [2]     9  10     2 [GA]</code></pre>
</div>
<p>In this particular example, two views describe matches of the pattern
in the subject sequence. Specifically, the pattern <code>GM</code> first
matched the sequence <code>GC</code> spanning positions 4 to 5 in the
subject sequence, and then also matched the sequence <code>GA</code>
from positions 9 to 10.</p>
<p>Similarly to the method <code>matchPattern()</code>, the method
<code>countPattern()</code> can be applied to simply count the number of
matches of the <code>pattern</code> in the <code>subject</code>
sequence. And again, the option <code>fixed</code> controls whether to
respect ambiguity codes, or match them exactly.</p>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>How many hits does the following code return? Why?</p>
<pre><code><span><span class="va">dna2</span> <span class="op">&lt;-</span> <span class="fu">DNAString</span><span class="op">(</span><span class="st">"TGATTGCTTGGTTGMTT"</span><span class="op">)</span></span>
<span><span class="fu">countPattern</span><span class="op">(</span><span class="st">"GM"</span>, <span class="va">dna2</span>, fixed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>The method <code>countPattern()</code> reports 3 hits, because the
option <code>fixed = FALSE</code> allows the pattern <code>GM</code> to
match <code>GA</code>, <code>GC</code>, and <code>GM</code> sequences,
due to the use of the ambiguity code <code>M</code> in the
<code>pattern</code>.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="importing-biological-strings-from-files">Importing biological strings from files<a class="anchor" aria-label="anchor" href="#importing-biological-strings-from-files"></a>
</h3>
<p>In practice, users rarely type the strings representing biological
sequences themselves. Most of the time, biological strings are imported
from files, either downloaded from public repositories or generated
locally using bioinformatics programs.</p>
<p>For instance, we can load the set of adapter sequences for the <a href="https://emea.illumina.com/products/by-type/sequencing-kits/library-prep-kits.html" class="external-link">TruSeq™
DNA PCR-Free whole-genome sequencing library preparation</a> kit from a
file that we downloaded during the lesson setup. Since adapter sequences
are nucleic acid sequences, we must use the function
<code>readDNAStringSet()</code>.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">truseq_adapters</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span>filepath <span class="op">=</span> <span class="st">"data/TruSeq3-PE-2.fa"</span><span class="op">)</span></span>
<span><span class="va">truseq_adapters</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DNAStringSet object of length 6:
    width seq                                               names
[1]    34 TACACTCTTTCCCTACACGACGCTCTTCCGATCT                PrefixPE/1
[2]    34 GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT                PrefixPE/2
[3]    34 TACACTCTTTCCCTACACGACGCTCTTCCGATCT                PE1
[4]    34 AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA                PE1_rc
[5]    34 GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT                PE2
[6]    34 AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC                PE2_rc</code></pre>
</div>
<div id="going-further" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="going-further" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>The help page of the function <code>readDNAStringSet()</code> –
accessible using <code>help(readDNAStringSet)</code> – documents related
functions designed to import other types of biological sequences, e.g
<code>readRNAStringSet()</code>, <code>readAAStringSet()</code>.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="operations-on-biological-strings">Operations on biological strings<a class="anchor" aria-label="anchor" href="#operations-on-biological-strings"></a>
</h3>
<div class="section level4">
<h4 id="computing-the-frequency-of-symbols">Computing the frequency of symbols<a class="anchor" aria-label="anchor" href="#computing-the-frequency-of-symbols"></a>
</h4>
<p>The <em><a href="https://bioconductor.org/packages/3.22/Biostrings" class="external-link">Biostrings</a></em>
package provides several functions to process and manipulate classes of
biological strings. For example, we have come across
<code>matchPattern()</code> and <code>countPattern()</code> earlier in
this episode.</p>
<p>Another example of a method that can be applied to biological strings
is <code>letterFrequency()</code>, to compute the frequency of letters
in a biological sequence.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">letterFrequency</span><span class="op">(</span><span class="va">truseq_adapters</span>, letters <span class="op">=</span> <span class="va">DNA_ALPHABET</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>      A  C  G  T M R W S Y K V H D B N - + .
[1,]  6 14  3 11 0 0 0 0 0 0 0 0 0 0 0 0 0 0
[2,]  5  8 10 11 0 0 0 0 0 0 0 0 0 0 0 0 0 0
[3,]  6 14  3 11 0 0 0 0 0 0 0 0 0 0 0 0 0 0
[4,] 11  3 14  6 0 0 0 0 0 0 0 0 0 0 0 0 0 0
[5,]  5  8 10 11 0 0 0 0 0 0 0 0 0 0 0 0 0 0
[6,] 11 10  8  5 0 0 0 0 0 0 0 0 0 0 0 0 0 0</code></pre>
</div>
<p>The output is a matrix with one row for each sequence in the
<code>DNAStringSet</code> object, and one column for each symbol in the
alphabet of deoxyribonucleic acids, provided by the <em><a href="https://bioconductor.org/packages/3.22/Biostrings" class="external-link">Biostrings</a></em>
package in a built-in object called <code>DNA_ALPHABET</code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="amino-acid-sequences">Amino acid sequences<a class="anchor" aria-label="anchor" href="#amino-acid-sequences"></a>
</h3>
<p>Similarly to the <code>DNAString</code> and <code>DNAStringSet</code>
classes, the classes <code>AAString</code> and <code>AAStringSet</code>
allow efficient storage and manipulation of a long amino acid sequence,
or a set thereof.</p>
<p>Similarly to built-in objects for the DNA alphabet, the built-in
objects <code>AA_ALPHABET</code>, <code>AA_STANDARD</code> and
<code>AA_PROTEINOGENIC</code> describe different subsets of the alphabet
of valid symbols for amino acid sequences.</p>
<p>For instance, the <code>AA_ALPHABET</code> object describes the set
of symbols in the full amino acid alphabet.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">AA_ALPHABET</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "A" "R" "N" "D" "C" "Q" "E" "G" "H" "I" "L" "K" "M" "F" "P" "S" "T" "W" "Y"
[20] "V" "U" "O" "B" "J" "Z" "X" "*" "-" "+" "."</code></pre>
</div>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Use base R code to identify the two symbols present in the
<code>AA_PROTEINOGENIC</code> alphabet object that are absent from the
<code>AA_STANDARD</code> alphabet object. What do those two symbols
represent?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<pre><code>&gt; setdiff(AA_PROTEINOGENIC, AA_STANDARD)
[1] "U" "O"</code></pre>
<p>The symbols <code>U</code> and <code>O</code> represent
selenocysteine and pyrrolysine, respectively. Those two amino acids are
in some species coded for by codons that are usually interpreted as stop
codons. As such, they are not included in the alphabet of “standard”
amino acids, and an alphabet of “proteinogenic” amino acids was defined
to acknowledge the special biology of those amino acids. Either of those
alphabets may be used to determine the validity of an amino acid
sequence, depending on its biological nature.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="translating-nucleotide-sequences">Translating nucleotide sequences<a class="anchor" aria-label="anchor" href="#translating-nucleotide-sequences"></a>
</h3>
<p>One of the key motivations for the use of <a href="reference.html#s4-class">S4 classes</a> and the object-oriented
programming (OOP) model relies on the infrastructure of S4 generics and
methods. As described in the earlier episode <a href="05-s4.html">The S4
class system</a>, generics provide a mechanism for defining and applying
distinct implementations of the same generic function name, according to
the nature of the input object(s) provided to the function call.</p>
<p>For instance, the <em><a href="https://bioconductor.org/packages/3.22/Biostrings" class="external-link">Biostrings</a></em>
package provides multiple implementations of a generic called
<code>translate()</code>, for translating DNA or RNA sequences into
amino acid sequences. The set of input objects supported by the generic
<code>translate()</code> can be listed using the function
<code>showMethods()</code>, from the CRAN package <em><a href="https://CRAN.R-project.org/package=methods" class="external-link">methods</a></em>.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">showMethods</span><span class="op">(</span><span class="st">"translate"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Function: translate (package Biostrings)
x="DNAString"
x="DNAStringSet"
x="MaskedDNAString"
x="MaskedRNAString"
x="RNAString"
x="RNAStringSet"</code></pre>
</div>
<p>In the output above, we see that that the generic function
<code>translate()</code> includes methods capable of handling objects
representing DNA and RNA sequences in the <code>DNAString</code> and
<code>RNAString</code> classes, respectively; but also lists of DNA and
RNA sequences in objects of class <code>DNAStringSet</code> and
<code>RNAStringSet</code>, as well as other classes capable of storing
DNA and RNA sequences.</p>
<p>To demonstrate the use of the <code>translate()</code> method, we
first load a set of open reading frames (ORFs) identified by the <a href="https://www.ncbi.nlm.nih.gov/orffinder/" class="external-link">NIH Open Reading Frame
Finder</a> for the <em>Homo sapiens</em> actin beta (ACTB) mRNA (RefSeq:
NM_001101), using the standard genetic code, a minimal ORF length of 75
nucleotides, and starting with the “ATG” start codon only.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">actb_orf_nih</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span><span class="st">"data/actb_orfs.fasta"</span><span class="op">)</span></span>
<span><span class="va">actb_orf_nih</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DNAStringSet object of length 13:
     width seq                                              names
 [1]   222 ATGCCCACCATCACGCCCTGGTG...CGGGGCGGACGCGGTCTCGGCG gi|1519311456|ref...
 [2]  1128 ATGGATGATGATATCGCCGCGCT...CGTCCACCGCAAATGCTTCTAG gi|1519311456|ref...
 [3]   126 ATGATGATATCGCCGCGCTCGTC...CGCCCCAGGCACCAGGGCGTGA gi|1519311456|ref...
 [4]    90 ATGTCGTCCCAGTTGGTGACGAT...CTGGGCCTCGTCGCCCACATAG gi|1519311456|ref...
 [5]   225 ATGGGCACAGTGTGGGTGACCCC...AGCCACACGCAGCTCATTGTAG gi|1519311456|ref...
 ...   ... ...
 [9]   342 ATGAGATTGGCATGGCTTTATTT...ATGTAATGCAAAATTTTTTTAA gi|1519311456|ref...
[10]   168 ATGGCTTTATTTGTTTTTTTTGT...TTGCACATTGTTGTTTTTTTAA gi|1519311456|ref...
[11]   111 ATGACTATTAAAAAAACAACAAT...CCTTCACCGTTCCAGTTTTTAA gi|1519311456|ref...
[12]   105 ATGCAAAATTTTTTTAATCTTCG...CCTTTTTTGTCCCCCAACTTGA gi|1519311456|ref...
[13]   135 ATGATGAGCCTTCGTGCCCCCCC...TGACTTGAGACCAGTTGAATAA gi|1519311456|ref...</code></pre>
</div>
<p>Having imported the nucleotide sequences as a
<code>DNAStringSet</code> object, we can apply the
<code>translate()</code> method to that object to produce the amino acid
sequence that results from the translation process for each nucleotide
sequence.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">actb_aa</span> <span class="op">&lt;-</span> <span class="fu">translate</span><span class="op">(</span><span class="va">actb_orf_nih</span><span class="op">)</span></span>
<span><span class="va">actb_aa</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>AAStringSet object of length 13:
     width seq                                              names
 [1]    74 MPTITPWCLGRPTMEGKTARGAS...VWTGGGSAKARLCARGADAVSA gi|1519311456|ref...
 [2]   376 MDDDIAALVVDNGSGMCKAGFAG...MWISKQEYDESGPSIVHRKCF* gi|1519311456|ref...
 [3]    42 MMISPRSSSTTAPACARPASRATMPPGPSSPPSWGAPGTRA*       gi|1519311456|ref...
 [4]    30 MSSQLVTMPCSMGYFRVRMPLLLWASSPT*                   gi|1519311456|ref...
 [5]    75 MGTVWVTPSPESITMPVVRPEAY...GFRGASVSSTGCSSGATRSSL* gi|1519311456|ref...
 ...   ... ...
 [9]   114 MRLAWLYLFFLFCFGFFFFFGLT...QVHTGEVIALLSCKLCNAKFF* gi|1519311456|ref...
[10]    56 MALFVFFVLFWFFFFFWLDSGFK...ERASPKVHNVAEDFDCTLLFF* gi|1519311456|ref...
[11]    37 MTIKKTTMCNQSPRPHCELWGMLAPTDCCHLHRSSF*            gi|1519311456|ref...
[12]    35 MQNFFNLRLNTFLFCFILNDEPSCPPFPLFCPPT*              gi|1519311456|ref...
[13]    45 MMSLRAPPSPFFVPQLEMYEGFWSPWEWVEAARAYLYTDLRPVE*    gi|1519311456|ref...</code></pre>
</div>
<p>In the example above, all amino acid sequences visible start with the
typical methionin amino acid encoded by the “ATG” start codon. We also
see that all but one of the amino acid sequences visible end with the
<code>*</code> symbol, which indicates that the translation process
ended on a stop codon. In contrast, the first open reading frame above
reached the end of the nucleotide sequence without encoutering a stop
codon.</p>
<p>Conveniently, the number of amino acids in each sequence is stated
under the header <code>width</code>.</p>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Extract the length of each amino acid sequence above as an integer
vector. What is the length of the longest amino acid sequence translated
from any of those open reading frames?</p>
<p>Compare your result with the sequence information on the UniPro page
for ACTB (<a href="https://www.uniprot.org/uniprot/P60709#sequences" class="external-link">https://www.uniprot.org/uniprot/P60709#sequences</a>).</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<pre><code><span><span class="fu">width</span><span class="op">(</span><span class="va">actb_aa</span><span class="op">)</span></span>
<span><span class="co"># or</span></span>
<span><span class="fu">max</span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">actb_aa</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>The longest translated sequence contains 376 amino acids.</p>
<p>The Uniprot page reports a sequence of 375 amino acids. However, the
UniProt amino acid sequence does not comprise any symbol to represent
the stop codon. That difference aside, the UniPro amino acid sequence is
identical to the sequence that was produced by the
<code>translate()</code> method.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="the-bsgenome-package">The BSgenome package<a class="anchor" aria-label="anchor" href="#the-bsgenome-package"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="overview-1">Overview<a class="anchor" aria-label="anchor" href="#overview-1"></a>
</h3>
<p>In the Bioconductor project, the <em><a href="https://bioconductor.org/packages/3.22/BSgenome" class="external-link">BSgenome</a></em>
package provides software infrastructure for efficient representation of
full genome and their single-nucleotide polymorphisms.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.22/BSgenome" class="external-link">BSgenome</a></em>
package itself does not contain any genome sequence itself, but provides
functionality to access genome sequences available in other Bioconductor
packages, as we demonstrate in the next section.</p>
</div>
<div class="section level3">
<h3 id="first-steps-1">First steps<a class="anchor" aria-label="anchor" href="#first-steps-1"></a>
</h3>
<p>To get started, we load the package.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">BSgenome</span><span class="op">)</span></span></code></pre>
</div>
<p>With the package loaded and attached to the session, we have access
to the package functions.</p>
<p>In particular, the function
<code>BSgenome::available.genomes()</code> can be used to display the
names of Bioconductor packages that contain genome sequences.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">available.genomes</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'getOption("repos")' replaces Bioconductor standard repositories, see
'help("repositories", package = "BiocManager")' for details.
Replacement repositories:
    BioCsoft: https://bioconductor.org/packages/3.22/bioc
    BioCann: https://bioconductor.org/packages/3.22/data/annotation
    BioCexp: https://bioconductor.org/packages/3.22/data/experiment
    BioCworkflows: https://bioconductor.org/packages/3.22/workflows
    BioCbooks: https://bioconductor.org/packages/3.22/books
    CRAN: https://cran.rstudio.com</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  [1] "BSgenome.Alyrata.JGI.v1"
  [2] "BSgenome.Amellifera.BeeBase.assembly4"
  [3] "BSgenome.Amellifera.NCBI.AmelHAv3.1"
  [4] "BSgenome.Amellifera.UCSC.apiMel2"
  [5] "BSgenome.Amellifera.UCSC.apiMel2.masked"
  [6] "BSgenome.Aofficinalis.NCBI.V1"
  [7] "BSgenome.Athaliana.TAIR.04232008"
  [8] "BSgenome.Athaliana.TAIR.TAIR9"
  [9] "BSgenome.Btaurus.UCSC.bosTau3"
 [10] "BSgenome.Btaurus.UCSC.bosTau3.masked"
 [11] "BSgenome.Btaurus.UCSC.bosTau4"
 [12] "BSgenome.Btaurus.UCSC.bosTau4.masked"
 [13] "BSgenome.Btaurus.UCSC.bosTau6"
 [14] "BSgenome.Btaurus.UCSC.bosTau6.masked"
 [15] "BSgenome.Btaurus.UCSC.bosTau8"
 [16] "BSgenome.Btaurus.UCSC.bosTau9"
 [17] "BSgenome.Btaurus.UCSC.bosTau9.masked"
 [18] "BSgenome.Carietinum.NCBI.v1"
 [19] "BSgenome.Celegans.UCSC.ce10"
 [20] "BSgenome.Celegans.UCSC.ce11"
 [21] "BSgenome.Celegans.UCSC.ce2"
 [22] "BSgenome.Celegans.UCSC.ce6"
 [23] "BSgenome.Cfamiliaris.UCSC.canFam2"
 [24] "BSgenome.Cfamiliaris.UCSC.canFam2.masked"
 [25] "BSgenome.Cfamiliaris.UCSC.canFam3"
 [26] "BSgenome.Cfamiliaris.UCSC.canFam3.masked"
 [27] "BSgenome.Cjacchus.UCSC.calJac3"
 [28] "BSgenome.Cjacchus.UCSC.calJac4"
 [29] "BSgenome.CneoformansVarGrubiiKN99.NCBI.ASM221672v1"
 [30] "BSgenome.Creinhardtii.JGI.v5.6"
 [31] "BSgenome.Dmelanogaster.UCSC.dm2"
 [32] "BSgenome.Dmelanogaster.UCSC.dm2.masked"
 [33] "BSgenome.Dmelanogaster.UCSC.dm3"
 [34] "BSgenome.Dmelanogaster.UCSC.dm3.masked"
 [35] "BSgenome.Dmelanogaster.UCSC.dm6"
 [36] "BSgenome.Drerio.UCSC.danRer10"
 [37] "BSgenome.Drerio.UCSC.danRer11"
 [38] "BSgenome.Drerio.UCSC.danRer5"
 [39] "BSgenome.Drerio.UCSC.danRer5.masked"
 [40] "BSgenome.Drerio.UCSC.danRer6"
 [41] "BSgenome.Drerio.UCSC.danRer6.masked"
 [42] "BSgenome.Drerio.UCSC.danRer7"
 [43] "BSgenome.Drerio.UCSC.danRer7.masked"
 [44] "BSgenome.Dvirilis.Ensembl.dvircaf1"
 [45] "BSgenome.Ecoli.NCBI.20080805"
 [46] "BSgenome.Gaculeatus.UCSC.gasAcu1"
 [47] "BSgenome.Gaculeatus.UCSC.gasAcu1.masked"
 [48] "BSgenome.Ggallus.UCSC.galGal3"
 [49] "BSgenome.Ggallus.UCSC.galGal3.masked"
 [50] "BSgenome.Ggallus.UCSC.galGal4"
 [51] "BSgenome.Ggallus.UCSC.galGal4.masked"
 [52] "BSgenome.Ggallus.UCSC.galGal5"
 [53] "BSgenome.Ggallus.UCSC.galGal6"
 [54] "BSgenome.Gmax.NCBI.Gmv40"
 [55] "BSgenome.Hsapiens.1000genomes.hs37d5"
 [56] "BSgenome.Hsapiens.NCBI.GRCh38"
 [57] "BSgenome.Hsapiens.NCBI.T2T.CHM13v2.0"
 [58] "BSgenome.Hsapiens.UCSC.hg17"
 [59] "BSgenome.Hsapiens.UCSC.hg17.masked"
 [60] "BSgenome.Hsapiens.UCSC.hg18"
 [61] "BSgenome.Hsapiens.UCSC.hg18.masked"
 [62] "BSgenome.Hsapiens.UCSC.hg19"
 [63] "BSgenome.Hsapiens.UCSC.hg19.masked"
 [64] "BSgenome.Hsapiens.UCSC.hg38"
 [65] "BSgenome.Hsapiens.UCSC.hg38.dbSNP151.major"
 [66] "BSgenome.Hsapiens.UCSC.hg38.dbSNP151.minor"
 [67] "BSgenome.Hsapiens.UCSC.hg38.masked"
 [68] "BSgenome.Hsapiens.UCSC.hs1"
 [69] "BSgenome.Mdomestica.UCSC.monDom5"
 [70] "BSgenome.Mfascicularis.NCBI.5.0"
 [71] "BSgenome.Mfascicularis.NCBI.6.0"
 [72] "BSgenome.Mfuro.UCSC.musFur1"
 [73] "BSgenome.Mmulatta.UCSC.rheMac10"
 [74] "BSgenome.Mmulatta.UCSC.rheMac2"
 [75] "BSgenome.Mmulatta.UCSC.rheMac2.masked"
 [76] "BSgenome.Mmulatta.UCSC.rheMac3"
 [77] "BSgenome.Mmulatta.UCSC.rheMac3.masked"
 [78] "BSgenome.Mmulatta.UCSC.rheMac8"
 [79] "BSgenome.Mmusculus.UCSC.mm10"
 [80] "BSgenome.Mmusculus.UCSC.mm10.masked"
 [81] "BSgenome.Mmusculus.UCSC.mm39"
 [82] "BSgenome.Mmusculus.UCSC.mm8"
 [83] "BSgenome.Mmusculus.UCSC.mm8.masked"
 [84] "BSgenome.Mmusculus.UCSC.mm9"
 [85] "BSgenome.Mmusculus.UCSC.mm9.masked"
 [86] "BSgenome.Osativa.MSU.MSU7"
 [87] "BSgenome.Ppaniscus.UCSC.panPan1"
 [88] "BSgenome.Ppaniscus.UCSC.panPan2"
 [89] "BSgenome.Ptroglodytes.UCSC.panTro2"
 [90] "BSgenome.Ptroglodytes.UCSC.panTro2.masked"
 [91] "BSgenome.Ptroglodytes.UCSC.panTro3"
 [92] "BSgenome.Ptroglodytes.UCSC.panTro3.masked"
 [93] "BSgenome.Ptroglodytes.UCSC.panTro5"
 [94] "BSgenome.Ptroglodytes.UCSC.panTro6"
 [95] "BSgenome.Rnorvegicus.UCSC.rn4"
 [96] "BSgenome.Rnorvegicus.UCSC.rn4.masked"
 [97] "BSgenome.Rnorvegicus.UCSC.rn5"
 [98] "BSgenome.Rnorvegicus.UCSC.rn5.masked"
 [99] "BSgenome.Rnorvegicus.UCSC.rn6"
[100] "BSgenome.Rnorvegicus.UCSC.rn7"
[101] "BSgenome.Scerevisiae.UCSC.sacCer1"
[102] "BSgenome.Scerevisiae.UCSC.sacCer2"
[103] "BSgenome.Scerevisiae.UCSC.sacCer3"
[104] "BSgenome.Sscrofa.UCSC.susScr11"
[105] "BSgenome.Sscrofa.UCSC.susScr3"
[106] "BSgenome.Sscrofa.UCSC.susScr3.masked"
[107] "BSgenome.Tgondii.ToxoDB.7.0"
[108] "BSgenome.Tguttata.UCSC.taeGut1"
[109] "BSgenome.Tguttata.UCSC.taeGut1.masked"
[110] "BSgenome.Tguttata.UCSC.taeGut2"
[111] "BSgenome.Vvinifera.URGI.IGGP12Xv0"
[112] "BSgenome.Vvinifera.URGI.IGGP12Xv2"
[113] "BSgenome.Vvinifera.URGI.IGGP8X"                    </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="installing-bsgenome-packages">Installing BSgenome packages<a class="anchor" aria-label="anchor" href="#installing-bsgenome-packages"></a>
</h3>
<p>To use one of the available genomes, the corresponding package must
be installed first. For instance, the example below demonstrates how the
data package <em><a href="https://bioconductor.org/packages/3.22/BSgenome.Hsapiens.UCSC.hg38.masked" class="external-link">BSgenome.Hsapiens.UCSC.hg38.masked</a></em>
can be installed using the function <code>BiocManager::install()</code>
that we have seen before.</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg38.masked"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="using-bsgenome-packages">Using BSgenome packages<a class="anchor" aria-label="anchor" href="#using-bsgenome-packages"></a>
</h3>
<p>Once installed, BSgenome packages can be loaded like any other R
package, using the <code>library()</code> function.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38.masked</span><span class="op">)</span></span></code></pre>
</div>
<p>Each BSgenome package contains an object that is named identically to
the package and contains the genome sequence.</p>
<p>Having loaded the package <em><a href="https://bioconductor.org/packages/3.22/BSgenome.Hsapiens.UCSC.hg38.masked" class="external-link">BSgenome.Hsapiens.UCSC.hg38.masked</a></em>
above, we can display the BSgenome object as follows.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BSgenome.Hsapiens.UCSC.hg38.masked</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>| BSgenome object for Human
| - organism: Homo sapiens
| - provider: UCSC
| - genome: hg38
| - release date: 2023/01/31
| - 711 sequence(s):
|     chr1                    chr2                    chr3
|     chr4                    chr5                    chr6
|     chr7                    chr8                    chr9
|     chr10                   chr11                   chr12
|     chr13                   chr14                   chr15
|     ...                     ...                     ...
|     chr19_KV575256v1_alt    chr19_KV575257v1_alt    chr19_KV575258v1_alt
|     chr19_KV575259v1_alt    chr19_KV575260v1_alt    chr19_MU273387v1_alt
|     chr22_KN196485v1_alt    chr22_KN196486v1_alt    chr22_KQ458387v1_alt
|     chr22_KQ458388v1_alt    chr22_KQ759761v1_alt    chrX_KV766199v1_alt
|     chrX_MU273395v1_alt     chrX_MU273396v1_alt     chrX_MU273397v1_alt
|
| Tips: call 'seqnames()' on the object to get all the sequence names, call
| 'seqinfo()' to get the full sequence info, use the '$' or '[[' operator to
| access a given sequence, see '?BSgenome' for more information.</code></pre>
</div>
<p>Given the length and the complexity of the object name, it is common
practice to assign a copy of BSgenome objects to a new object simply
called <code>genome</code>.</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genome</span> <span class="op">&lt;-</span> <span class="va">BSgenome.Hsapiens.UCSC.hg38.masked</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="using-bsgenome-objects">Using BSgenome objects<a class="anchor" aria-label="anchor" href="#using-bsgenome-objects"></a>
</h3>
<p>When printing BSgenome objects in the console (see above), some
helpful tips are displayed under the object itself, hinting at functions
commonly used to access information in the object.</p>
<p>For instance, the function <code>seqnames()</code> can be used get
the list of sequence names (i.e., chromosomes and contigs) present in
the object.</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seqnames</span><span class="op">(</span><span class="va">genome</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  [1] "chr1"                    "chr2"
  [3] "chr3"                    "chr4"
  [5] "chr5"                    "chr6"
  [7] "chr7"                    "chr8"
  [9] "chr9"                    "chr10"
 [11] "chr11"                   "chr12"
 [13] "chr13"                   "chr14"
 [15] "chr15"                   "chr16"
 [17] "chr17"                   "chr18"
 [19] "chr19"                   "chr20"
 [21] "chr21"                   "chr22"
 [23] "chrX"                    "chrY"
 [25] "chrM"                    "chr1_GL383518v1_alt"
 [27] "chr1_GL383519v1_alt"     "chr1_GL383520v2_alt"
 [29] "chr1_KI270759v1_alt"     "chr1_KI270760v1_alt"
 [31] "chr1_KI270761v1_alt"     "chr1_KI270762v1_alt"
 [33] "chr1_KI270763v1_alt"     "chr1_KI270764v1_alt"
 [35] "chr1_KI270765v1_alt"     "chr1_KI270766v1_alt"
 [37] "chr1_KI270892v1_alt"     "chr2_GL383521v1_alt"
 [39] "chr2_GL383522v1_alt"     "chr2_GL582966v2_alt"
 [41] "chr2_KI270767v1_alt"     "chr2_KI270768v1_alt"
 [43] "chr2_KI270769v1_alt"     "chr2_KI270770v1_alt"
 [45] "chr2_KI270771v1_alt"     "chr2_KI270772v1_alt"
 [47] "chr2_KI270773v1_alt"     "chr2_KI270774v1_alt"
 [49] "chr2_KI270775v1_alt"     "chr2_KI270776v1_alt"
 [51] "chr2_KI270893v1_alt"     "chr2_KI270894v1_alt"
 [53] "chr3_GL383526v1_alt"     "chr3_JH636055v2_alt"
 [55] "chr3_KI270777v1_alt"     "chr3_KI270778v1_alt"
 [57] "chr3_KI270779v1_alt"     "chr3_KI270780v1_alt"
 [59] "chr3_KI270781v1_alt"     "chr3_KI270782v1_alt"
 [61] "chr3_KI270783v1_alt"     "chr3_KI270784v1_alt"
 [63] "chr3_KI270895v1_alt"     "chr3_KI270924v1_alt"
 [65] "chr3_KI270934v1_alt"     "chr3_KI270935v1_alt"
 [67] "chr3_KI270936v1_alt"     "chr3_KI270937v1_alt"
 [69] "chr4_GL000257v2_alt"     "chr4_GL383527v1_alt"
 [71] "chr4_GL383528v1_alt"     "chr4_KI270785v1_alt"
 [73] "chr4_KI270786v1_alt"     "chr4_KI270787v1_alt"
 [75] "chr4_KI270788v1_alt"     "chr4_KI270789v1_alt"
 [77] "chr4_KI270790v1_alt"     "chr4_KI270896v1_alt"
 [79] "chr4_KI270925v1_alt"     "chr5_GL339449v2_alt"
 [81] "chr5_GL383530v1_alt"     "chr5_GL383531v1_alt"
 [83] "chr5_GL383532v1_alt"     "chr5_GL949742v1_alt"
 [85] "chr5_KI270791v1_alt"     "chr5_KI270792v1_alt"
 [87] "chr5_KI270793v1_alt"     "chr5_KI270794v1_alt"
 [89] "chr5_KI270795v1_alt"     "chr5_KI270796v1_alt"
 [91] "chr5_KI270897v1_alt"     "chr5_KI270898v1_alt"
 [93] "chr6_GL000250v2_alt"     "chr6_GL000251v2_alt"
 [95] "chr6_GL000252v2_alt"     "chr6_GL000253v2_alt"
 [97] "chr6_GL000254v2_alt"     "chr6_GL000255v2_alt"
 [99] "chr6_GL000256v2_alt"     "chr6_GL383533v1_alt"
[101] "chr6_KB021644v2_alt"     "chr6_KI270758v1_alt"
[103] "chr6_KI270797v1_alt"     "chr6_KI270798v1_alt"
[105] "chr6_KI270799v1_alt"     "chr6_KI270800v1_alt"
[107] "chr6_KI270801v1_alt"     "chr6_KI270802v1_alt"
[109] "chr7_GL383534v2_alt"     "chr7_KI270803v1_alt"
[111] "chr7_KI270804v1_alt"     "chr7_KI270805v1_alt"
[113] "chr7_KI270806v1_alt"     "chr7_KI270807v1_alt"
[115] "chr7_KI270808v1_alt"     "chr7_KI270809v1_alt"
[117] "chr7_KI270899v1_alt"     "chr8_KI270810v1_alt"
[119] "chr8_KI270811v1_alt"     "chr8_KI270812v1_alt"
[121] "chr8_KI270813v1_alt"     "chr8_KI270814v1_alt"
[123] "chr8_KI270815v1_alt"     "chr8_KI270816v1_alt"
[125] "chr8_KI270817v1_alt"     "chr8_KI270818v1_alt"
[127] "chr8_KI270819v1_alt"     "chr8_KI270820v1_alt"
[129] "chr8_KI270821v1_alt"     "chr8_KI270822v1_alt"
[131] "chr8_KI270900v1_alt"     "chr8_KI270901v1_alt"
[133] "chr8_KI270926v1_alt"     "chr9_GL383539v1_alt"
[135] "chr9_GL383540v1_alt"     "chr9_GL383541v1_alt"
[137] "chr9_GL383542v1_alt"     "chr9_KI270823v1_alt"
[139] "chr10_GL383545v1_alt"    "chr10_GL383546v1_alt"
[141] "chr10_KI270824v1_alt"    "chr10_KI270825v1_alt"
[143] "chr11_GL383547v1_alt"    "chr11_JH159136v1_alt"
[145] "chr11_JH159137v1_alt"    "chr11_KI270826v1_alt"
[147] "chr11_KI270827v1_alt"    "chr11_KI270829v1_alt"
[149] "chr11_KI270830v1_alt"    "chr11_KI270831v1_alt"
[151] "chr11_KI270832v1_alt"    "chr11_KI270902v1_alt"
[153] "chr11_KI270903v1_alt"    "chr11_KI270927v1_alt"
[155] "chr12_GL383549v1_alt"    "chr12_GL383550v2_alt"
[157] "chr12_GL383551v1_alt"    "chr12_GL383552v1_alt"
[159] "chr12_GL383553v2_alt"    "chr12_GL877875v1_alt"
[161] "chr12_GL877876v1_alt"    "chr12_KI270833v1_alt"
[163] "chr12_KI270834v1_alt"    "chr12_KI270835v1_alt"
[165] "chr12_KI270836v1_alt"    "chr12_KI270837v1_alt"
[167] "chr12_KI270904v1_alt"    "chr13_KI270838v1_alt"
[169] "chr13_KI270839v1_alt"    "chr13_KI270840v1_alt"
[171] "chr13_KI270841v1_alt"    "chr13_KI270842v1_alt"
[173] "chr13_KI270843v1_alt"    "chr14_KI270844v1_alt"
[175] "chr14_KI270845v1_alt"    "chr14_KI270846v1_alt"
[177] "chr14_KI270847v1_alt"    "chr15_GL383554v1_alt"
[179] "chr15_GL383555v2_alt"    "chr15_KI270848v1_alt"
[181] "chr15_KI270849v1_alt"    "chr15_KI270850v1_alt"
[183] "chr15_KI270851v1_alt"    "chr15_KI270852v1_alt"
[185] "chr15_KI270905v1_alt"    "chr15_KI270906v1_alt"
[187] "chr16_GL383556v1_alt"    "chr16_GL383557v1_alt"
[189] "chr16_KI270853v1_alt"    "chr16_KI270854v1_alt"
[191] "chr16_KI270855v1_alt"    "chr16_KI270856v1_alt"
[193] "chr17_GL000258v2_alt"    "chr17_GL383563v3_alt"
[195] "chr17_GL383564v2_alt"    "chr17_GL383565v1_alt"
[197] "chr17_GL383566v1_alt"    "chr17_JH159146v1_alt"
[199] "chr17_JH159147v1_alt"    "chr17_JH159148v1_alt"
[201] "chr17_KI270857v1_alt"    "chr17_KI270858v1_alt"
[203] "chr17_KI270859v1_alt"    "chr17_KI270860v1_alt"
[205] "chr17_KI270861v1_alt"    "chr17_KI270862v1_alt"
[207] "chr17_KI270907v1_alt"    "chr17_KI270908v1_alt"
[209] "chr17_KI270909v1_alt"    "chr17_KI270910v1_alt"
[211] "chr18_GL383567v1_alt"    "chr18_GL383568v1_alt"
[213] "chr18_GL383569v1_alt"    "chr18_GL383570v1_alt"
[215] "chr18_GL383571v1_alt"    "chr18_GL383572v1_alt"
[217] "chr18_KI270863v1_alt"    "chr18_KI270864v1_alt"
[219] "chr18_KI270911v1_alt"    "chr18_KI270912v1_alt"
[221] "chr19_GL000209v2_alt"    "chr19_GL383573v1_alt"
[223] "chr19_GL383574v1_alt"    "chr19_GL383575v2_alt"
[225] "chr19_GL383576v1_alt"    "chr19_GL949746v1_alt"
[227] "chr19_GL949747v2_alt"    "chr19_GL949748v2_alt"
[229] "chr19_GL949749v2_alt"    "chr19_GL949750v2_alt"
[231] "chr19_GL949751v2_alt"    "chr19_GL949752v1_alt"
[233] "chr19_GL949753v2_alt"    "chr19_KI270865v1_alt"
[235] "chr19_KI270866v1_alt"    "chr19_KI270867v1_alt"
[237] "chr19_KI270868v1_alt"    "chr19_KI270882v1_alt"
[239] "chr19_KI270883v1_alt"    "chr19_KI270884v1_alt"
[241] "chr19_KI270885v1_alt"    "chr19_KI270886v1_alt"
[243] "chr19_KI270887v1_alt"    "chr19_KI270888v1_alt"
[245] "chr19_KI270889v1_alt"    "chr19_KI270890v1_alt"
[247] "chr19_KI270891v1_alt"    "chr19_KI270914v1_alt"
[249] "chr19_KI270915v1_alt"    "chr19_KI270916v1_alt"
[251] "chr19_KI270917v1_alt"    "chr19_KI270918v1_alt"
[253] "chr19_KI270919v1_alt"    "chr19_KI270920v1_alt"
[255] "chr19_KI270921v1_alt"    "chr19_KI270922v1_alt"
[257] "chr19_KI270923v1_alt"    "chr19_KI270929v1_alt"
[259] "chr19_KI270930v1_alt"    "chr19_KI270931v1_alt"
[261] "chr19_KI270932v1_alt"    "chr19_KI270933v1_alt"
[263] "chr19_KI270938v1_alt"    "chr20_GL383577v2_alt"
[265] "chr20_KI270869v1_alt"    "chr20_KI270870v1_alt"
[267] "chr20_KI270871v1_alt"    "chr21_GL383578v2_alt"
[269] "chr21_GL383579v2_alt"    "chr21_GL383580v2_alt"
[271] "chr21_GL383581v2_alt"    "chr21_KI270872v1_alt"
[273] "chr21_KI270873v1_alt"    "chr21_KI270874v1_alt"
[275] "chr22_GL383582v2_alt"    "chr22_GL383583v2_alt"
[277] "chr22_KB663609v1_alt"    "chr22_KI270875v1_alt"
[279] "chr22_KI270876v1_alt"    "chr22_KI270877v1_alt"
[281] "chr22_KI270878v1_alt"    "chr22_KI270879v1_alt"
[283] "chr22_KI270928v1_alt"    "chrX_KI270880v1_alt"
[285] "chrX_KI270881v1_alt"     "chrX_KI270913v1_alt"
[287] "chr1_KI270706v1_random"  "chr1_KI270707v1_random"
[289] "chr1_KI270708v1_random"  "chr1_KI270709v1_random"
[291] "chr1_KI270710v1_random"  "chr1_KI270711v1_random"
[293] "chr1_KI270712v1_random"  "chr1_KI270713v1_random"
[295] "chr1_KI270714v1_random"  "chr2_KI270715v1_random"
[297] "chr2_KI270716v1_random"  "chr3_GL000221v1_random"
[299] "chr4_GL000008v2_random"  "chr5_GL000208v1_random"
[301] "chr9_KI270717v1_random"  "chr9_KI270718v1_random"
[303] "chr9_KI270719v1_random"  "chr9_KI270720v1_random"
[305] "chr11_KI270721v1_random" "chr14_GL000009v2_random"
[307] "chr14_GL000194v1_random" "chr14_GL000225v1_random"
[309] "chr14_KI270722v1_random" "chr14_KI270723v1_random"
[311] "chr14_KI270724v1_random" "chr14_KI270725v1_random"
[313] "chr14_KI270726v1_random" "chr15_KI270727v1_random"
[315] "chr16_KI270728v1_random" "chr17_GL000205v2_random"
[317] "chr17_KI270729v1_random" "chr17_KI270730v1_random"
[319] "chr22_KI270731v1_random" "chr22_KI270732v1_random"
[321] "chr22_KI270733v1_random" "chr22_KI270734v1_random"
[323] "chr22_KI270735v1_random" "chr22_KI270736v1_random"
[325] "chr22_KI270737v1_random" "chr22_KI270738v1_random"
[327] "chr22_KI270739v1_random" "chrY_KI270740v1_random"
[329] "chrUn_GL000195v1"        "chrUn_GL000213v1"
[331] "chrUn_GL000214v1"        "chrUn_GL000216v2"
[333] "chrUn_GL000218v1"        "chrUn_GL000219v1"
[335] "chrUn_GL000220v1"        "chrUn_GL000224v1"
[337] "chrUn_GL000226v1"        "chrUn_KI270302v1"
[339] "chrUn_KI270303v1"        "chrUn_KI270304v1"
[341] "chrUn_KI270305v1"        "chrUn_KI270310v1"
[343] "chrUn_KI270311v1"        "chrUn_KI270312v1"
[345] "chrUn_KI270315v1"        "chrUn_KI270316v1"
[347] "chrUn_KI270317v1"        "chrUn_KI270320v1"
[349] "chrUn_KI270322v1"        "chrUn_KI270329v1"
[351] "chrUn_KI270330v1"        "chrUn_KI270333v1"
[353] "chrUn_KI270334v1"        "chrUn_KI270335v1"
[355] "chrUn_KI270336v1"        "chrUn_KI270337v1"
[357] "chrUn_KI270338v1"        "chrUn_KI270340v1"
[359] "chrUn_KI270362v1"        "chrUn_KI270363v1"
[361] "chrUn_KI270364v1"        "chrUn_KI270366v1"
[363] "chrUn_KI270371v1"        "chrUn_KI270372v1"
[365] "chrUn_KI270373v1"        "chrUn_KI270374v1"
[367] "chrUn_KI270375v1"        "chrUn_KI270376v1"
[369] "chrUn_KI270378v1"        "chrUn_KI270379v1"
[371] "chrUn_KI270381v1"        "chrUn_KI270382v1"
[373] "chrUn_KI270383v1"        "chrUn_KI270384v1"
[375] "chrUn_KI270385v1"        "chrUn_KI270386v1"
[377] "chrUn_KI270387v1"        "chrUn_KI270388v1"
[379] "chrUn_KI270389v1"        "chrUn_KI270390v1"
[381] "chrUn_KI270391v1"        "chrUn_KI270392v1"
[383] "chrUn_KI270393v1"        "chrUn_KI270394v1"
[385] "chrUn_KI270395v1"        "chrUn_KI270396v1"
[387] "chrUn_KI270411v1"        "chrUn_KI270412v1"
[389] "chrUn_KI270414v1"        "chrUn_KI270417v1"
[391] "chrUn_KI270418v1"        "chrUn_KI270419v1"
[393] "chrUn_KI270420v1"        "chrUn_KI270422v1"
[395] "chrUn_KI270423v1"        "chrUn_KI270424v1"
[397] "chrUn_KI270425v1"        "chrUn_KI270429v1"
[399] "chrUn_KI270435v1"        "chrUn_KI270438v1"
[401] "chrUn_KI270442v1"        "chrUn_KI270448v1"
[403] "chrUn_KI270465v1"        "chrUn_KI270466v1"
[405] "chrUn_KI270467v1"        "chrUn_KI270468v1"
[407] "chrUn_KI270507v1"        "chrUn_KI270508v1"
[409] "chrUn_KI270509v1"        "chrUn_KI270510v1"
[411] "chrUn_KI270511v1"        "chrUn_KI270512v1"
[413] "chrUn_KI270515v1"        "chrUn_KI270516v1"
[415] "chrUn_KI270517v1"        "chrUn_KI270518v1"
[417] "chrUn_KI270519v1"        "chrUn_KI270521v1"
[419] "chrUn_KI270522v1"        "chrUn_KI270528v1"
[421] "chrUn_KI270529v1"        "chrUn_KI270530v1"
[423] "chrUn_KI270538v1"        "chrUn_KI270539v1"
[425] "chrUn_KI270544v1"        "chrUn_KI270548v1"
[427] "chrUn_KI270579v1"        "chrUn_KI270580v1"
[429] "chrUn_KI270581v1"        "chrUn_KI270582v1"
[431] "chrUn_KI270583v1"        "chrUn_KI270584v1"
[433] "chrUn_KI270587v1"        "chrUn_KI270588v1"
[435] "chrUn_KI270589v1"        "chrUn_KI270590v1"
[437] "chrUn_KI270591v1"        "chrUn_KI270593v1"
[439] "chrUn_KI270741v1"        "chrUn_KI270742v1"
[441] "chrUn_KI270743v1"        "chrUn_KI270744v1"
[443] "chrUn_KI270745v1"        "chrUn_KI270746v1"
[445] "chrUn_KI270747v1"        "chrUn_KI270748v1"
[447] "chrUn_KI270749v1"        "chrUn_KI270750v1"
[449] "chrUn_KI270751v1"        "chrUn_KI270752v1"
[451] "chrUn_KI270753v1"        "chrUn_KI270754v1"
[453] "chrUn_KI270755v1"        "chrUn_KI270756v1"
[455] "chrUn_KI270757v1"        "chr1_KN196472v1_fix"
[457] "chr1_KN196473v1_fix"     "chr1_KN196474v1_fix"
[459] "chr1_KN538360v1_fix"     "chr1_KN538361v1_fix"
[461] "chr1_KQ031383v1_fix"     "chr1_KZ208906v1_fix"
[463] "chr1_KZ559100v1_fix"     "chr1_MU273333v1_fix"
[465] "chr1_MU273334v1_fix"     "chr1_MU273335v1_fix"
[467] "chr1_MU273336v1_fix"     "chr2_KN538362v1_fix"
[469] "chr2_KN538363v1_fix"     "chr2_KQ031384v1_fix"
[471] "chr2_ML143341v1_fix"     "chr2_ML143342v1_fix"
[473] "chr2_MU273341v1_fix"     "chr2_MU273342v1_fix"
[475] "chr2_MU273343v1_fix"     "chr2_MU273344v1_fix"
[477] "chr2_MU273345v1_fix"     "chr3_KN196475v1_fix"
[479] "chr3_KN196476v1_fix"     "chr3_KN538364v1_fix"
[481] "chr3_KQ031385v1_fix"     "chr3_KQ031386v1_fix"
[483] "chr3_KV766192v1_fix"     "chr3_KZ559104v1_fix"
[485] "chr3_MU273346v1_fix"     "chr3_MU273347v1_fix"
[487] "chr3_MU273348v1_fix"     "chr4_KQ983257v1_fix"
[489] "chr4_ML143344v1_fix"     "chr4_ML143345v1_fix"
[491] "chr4_ML143346v1_fix"     "chr4_ML143347v1_fix"
[493] "chr4_ML143348v1_fix"     "chr4_ML143349v1_fix"
[495] "chr4_MU273350v1_fix"     "chr4_MU273351v1_fix"
[497] "chr5_KV575244v1_fix"     "chr5_ML143350v1_fix"
[499] "chr5_MU273352v1_fix"     "chr5_MU273353v1_fix"
[501] "chr5_MU273354v1_fix"     "chr5_MU273355v1_fix"
[503] "chr6_KN196478v1_fix"     "chr6_KQ031387v1_fix"
[505] "chr6_KQ090016v1_fix"     "chr6_KV766194v1_fix"
[507] "chr6_KZ208911v1_fix"     "chr6_ML143351v1_fix"
[509] "chr7_KQ031388v1_fix"     "chr7_KV880764v1_fix"
[511] "chr7_KV880765v1_fix"     "chr7_KZ208912v1_fix"
[513] "chr7_ML143352v1_fix"     "chr8_KV880766v1_fix"
[515] "chr8_KV880767v1_fix"     "chr8_KZ208914v1_fix"
[517] "chr8_KZ208915v1_fix"     "chr8_MU273359v1_fix"
[519] "chr8_MU273360v1_fix"     "chr8_MU273361v1_fix"
[521] "chr8_MU273362v1_fix"     "chr8_MU273363v1_fix"
[523] "chr9_KN196479v1_fix"     "chr9_ML143353v1_fix"
[525] "chr9_MU273364v1_fix"     "chr9_MU273365v1_fix"
[527] "chr9_MU273366v1_fix"     "chr10_KN196480v1_fix"
[529] "chr10_KN538365v1_fix"    "chr10_KN538366v1_fix"
[531] "chr10_KN538367v1_fix"    "chr10_KQ090021v1_fix"
[533] "chr10_ML143354v1_fix"    "chr10_ML143355v1_fix"
[535] "chr10_MU273367v1_fix"    "chr11_KN196481v1_fix"
[537] "chr11_KQ090022v1_fix"    "chr11_KQ759759v1_fix"
[539] "chr11_KQ759759v2_fix"    "chr11_KV766195v1_fix"
[541] "chr11_KZ559108v1_fix"    "chr11_KZ559109v1_fix"
[543] "chr11_ML143356v1_fix"    "chr11_ML143357v1_fix"
[545] "chr11_ML143358v1_fix"    "chr11_ML143359v1_fix"
[547] "chr11_ML143360v1_fix"    "chr11_MU273369v1_fix"
[549] "chr11_MU273370v1_fix"    "chr11_MU273371v1_fix"
[551] "chr12_KN196482v1_fix"    "chr12_KN538369v1_fix"
[553] "chr12_KN538370v1_fix"    "chr12_KQ759760v1_fix"
[555] "chr12_KZ208916v1_fix"    "chr12_KZ208917v1_fix"
[557] "chr12_ML143361v1_fix"    "chr12_ML143362v1_fix"
[559] "chr12_MU273372v1_fix"    "chr13_KN196483v1_fix"
[561] "chr13_KN538371v1_fix"    "chr13_KN538372v1_fix"
[563] "chr13_KN538373v1_fix"    "chr13_ML143363v1_fix"
[565] "chr13_ML143364v1_fix"    "chr13_ML143365v1_fix"
[567] "chr13_ML143366v1_fix"    "chr14_KZ208920v1_fix"
[569] "chr14_ML143367v1_fix"    "chr14_MU273373v1_fix"
[571] "chr15_KN538374v1_fix"    "chr15_ML143369v1_fix"
[573] "chr15_ML143370v1_fix"    "chr15_ML143371v1_fix"
[575] "chr15_ML143372v1_fix"    "chr15_MU273374v1_fix"
[577] "chr16_KV880768v1_fix"    "chr16_KZ559113v1_fix"
[579] "chr16_ML143373v1_fix"    "chr16_MU273376v1_fix"
[581] "chr16_MU273377v1_fix"    "chr17_KV575245v1_fix"
[583] "chr17_KV766196v1_fix"    "chr17_ML143374v1_fix"
[585] "chr17_ML143375v1_fix"    "chr17_MU273379v1_fix"
[587] "chr17_MU273380v1_fix"    "chr17_MU273381v1_fix"
[589] "chr17_MU273382v1_fix"    "chr17_MU273383v1_fix"
[591] "chr18_KQ090028v1_fix"    "chr18_KZ208922v1_fix"
[593] "chr18_KZ559115v1_fix"    "chr19_KN196484v1_fix"
[595] "chr19_KQ458386v1_fix"    "chr19_ML143376v1_fix"
[597] "chr19_MU273384v1_fix"    "chr19_MU273385v1_fix"
[599] "chr19_MU273386v1_fix"    "chr20_MU273388v1_fix"
[601] "chr20_MU273389v1_fix"    "chr21_ML143377v1_fix"
[603] "chr21_MU273390v1_fix"    "chr21_MU273391v1_fix"
[605] "chr21_MU273392v1_fix"    "chr22_KQ759762v1_fix"
[607] "chr22_KQ759762v2_fix"    "chr22_ML143378v1_fix"
[609] "chr22_ML143379v1_fix"    "chr22_ML143380v1_fix"
[611] "chrX_ML143381v1_fix"     "chrX_ML143382v1_fix"
[613] "chrX_ML143383v1_fix"     "chrX_ML143384v1_fix"
[615] "chrX_ML143385v1_fix"     "chrX_MU273393v1_fix"
[617] "chrX_MU273394v1_fix"     "chrY_KN196487v1_fix"
[619] "chrY_KZ208923v1_fix"     "chrY_KZ208924v1_fix"
[621] "chrY_MU273398v1_fix"     "chr1_KQ458382v1_alt"
[623] "chr1_KQ458383v1_alt"     "chr1_KQ458384v1_alt"
[625] "chr1_KQ983255v1_alt"     "chr1_KV880763v1_alt"
[627] "chr1_KZ208904v1_alt"     "chr1_KZ208905v1_alt"
[629] "chr1_MU273330v1_alt"     "chr1_MU273331v1_alt"
[631] "chr1_MU273332v1_alt"     "chr2_KQ983256v1_alt"
[633] "chr2_KZ208907v1_alt"     "chr2_KZ208908v1_alt"
[635] "chr2_MU273337v1_alt"     "chr2_MU273338v1_alt"
[637] "chr2_MU273339v1_alt"     "chr2_MU273340v1_alt"
[639] "chr3_KZ208909v1_alt"     "chr3_KZ559101v1_alt"
[641] "chr3_KZ559102v1_alt"     "chr3_KZ559103v1_alt"
[643] "chr3_KZ559105v1_alt"     "chr3_ML143343v1_alt"
[645] "chr4_KQ090013v1_alt"     "chr4_KQ090014v1_alt"
[647] "chr4_KQ090015v1_alt"     "chr4_KQ983258v1_alt"
[649] "chr4_KV766193v1_alt"     "chr4_MU273349v1_alt"
[651] "chr5_KN196477v1_alt"     "chr5_KV575243v1_alt"
[653] "chr5_KZ208910v1_alt"     "chr5_MU273356v1_alt"
[655] "chr6_KQ090017v1_alt"     "chr6_MU273357v1_alt"
[657] "chr7_KZ208913v1_alt"     "chr7_KZ559106v1_alt"
[659] "chr7_MU273358v1_alt"     "chr8_KZ559107v1_alt"
[661] "chr9_KQ090018v1_alt"     "chr9_KQ090019v1_alt"
[663] "chr10_KQ090020v1_alt"    "chr11_KN538368v1_alt"
[665] "chr11_KZ559110v1_alt"    "chr11_KZ559111v1_alt"
[667] "chr11_MU273368v1_alt"    "chr12_KQ090023v1_alt"
[669] "chr12_KZ208918v1_alt"    "chr12_KZ559112v1_alt"
[671] "chr13_KQ090024v1_alt"    "chr13_KQ090025v1_alt"
[673] "chr14_KZ208919v1_alt"    "chr14_ML143368v1_alt"
[675] "chr15_KQ031389v1_alt"    "chr15_MU273375v1_alt"
[677] "chr16_KQ031390v1_alt"    "chr16_KQ090026v1_alt"
[679] "chr16_KQ090027v1_alt"    "chr16_KZ208921v1_alt"
[681] "chr17_KV766197v1_alt"    "chr17_KV766198v1_alt"
[683] "chr17_KZ559114v1_alt"    "chr17_MU273378v1_alt"
[685] "chr18_KQ458385v1_alt"    "chr18_KZ559116v1_alt"
[687] "chr19_KV575246v1_alt"    "chr19_KV575247v1_alt"
[689] "chr19_KV575248v1_alt"    "chr19_KV575249v1_alt"
[691] "chr19_KV575250v1_alt"    "chr19_KV575251v1_alt"
[693] "chr19_KV575252v1_alt"    "chr19_KV575253v1_alt"
[695] "chr19_KV575254v1_alt"    "chr19_KV575255v1_alt"
[697] "chr19_KV575256v1_alt"    "chr19_KV575257v1_alt"
[699] "chr19_KV575258v1_alt"    "chr19_KV575259v1_alt"
[701] "chr19_KV575260v1_alt"    "chr19_MU273387v1_alt"
[703] "chr22_KN196485v1_alt"    "chr22_KN196486v1_alt"
[705] "chr22_KQ458387v1_alt"    "chr22_KQ458388v1_alt"
[707] "chr22_KQ759761v1_alt"    "chrX_KV766199v1_alt"
[709] "chrX_MU273395v1_alt"     "chrX_MU273396v1_alt"
[711] "chrX_MU273397v1_alt"    </code></pre>
</div>
<p>Similarly, the function <code>seqinfo()</code> can be used to get the
full sequence information stored in the object.</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seqinfo</span><span class="op">(</span><span class="va">genome</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Seqinfo object with 711 sequences (1 circular) from hg38 genome:
  seqnames             seqlengths isCircular genome
  chr1                  248956422      FALSE   hg38
  chr2                  242193529      FALSE   hg38
  chr3                  198295559      FALSE   hg38
  chr4                  190214555      FALSE   hg38
  chr5                  181538259      FALSE   hg38
  ...                         ...        ...    ...
  chr22_KQ759761v1_alt     145162      FALSE   hg38
  chrX_KV766199v1_alt      188004      FALSE   hg38
  chrX_MU273395v1_alt      619716      FALSE   hg38
  chrX_MU273396v1_alt      294119      FALSE   hg38
  chrX_MU273397v1_alt      330493      FALSE   hg38</code></pre>
</div>
<p>Finally, the nature of BSgenome objects being akin to a list of
sequences, the operators <code>$</code> and <code>[[]]</code> can both
be used to extract individual sequences from the BSgenome object.</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genome</span><span class="op">$</span><span class="va">chr1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>248956422-letter MaskedDNAString object (# for masking)
seq: ####################################...####################################
masks:
  maskedwidth  maskedratio active names                               desc
1    18470101 7.419010e-02   TRUE AGAPS                      assembly gaps
2        5309 2.132502e-05   TRUE   AMB           intra-contig ambiguities
3   119060341 4.782377e-01  FALSE    RM                       RepeatMasker
4     1647959 6.619468e-03  FALSE   TRF Tandem Repeats Finder [period&lt;=12]
all masks together:
  maskedwidth maskedratio
    137685771   0.5530517
all active masks together:
  maskedwidth maskedratio
     18475410  0.07421142</code></pre>
</div>
<p>For instance, we can extract the sequence of the Y chromosome and
assign it to a new object <code>chrY</code>.</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrY</span> <span class="op">&lt;-</span> <span class="va">genome</span><span class="op">[[</span><span class="st">"chrY"</span><span class="op">]</span><span class="op">]</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="using-genome-sequences">Using genome sequences<a class="anchor" aria-label="anchor" href="#using-genome-sequences"></a>
</h3>
<p>From this point, genome sequences can be treated very much like
biological strings (e.g. <code>DNAString</code>) described earlier, in
the <em><a href="https://bioconductor.org/packages/3.22/Biostrings" class="external-link">Biostrings</a></em>
package.</p>
<p>For instance, the function <code>countPattern()</code> can be used to
count the number of occurences of a given pattern in a given genome
sequence.</p>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">countPattern</span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"CANNTG"</span>, subject <span class="op">=</span> <span class="va">chrY</span>, fixed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 141609</code></pre>
</div>
<div id="note" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="note" class="callout-inner">
<h3 class="callout-title">Note</h3>
<div class="callout-content">
<p>In the example above, the argument <code>fixed = FALSE</code> is used
to indicate that the pattern contain <a href="https://en.wikipedia.org/wiki/Nucleic_acid_notation#IUPAC_notation" class="external-link">IUPAC
ambiguity codes</a>.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>The <code>Biostrings</code> package defines classes to represent
sequences of nucleotides and amino acids.</li>
<li>The <code>Biostrings</code> package also defines methods to
efficiently process biological sequences.</li>
<li>The <code>BSgenome</code> package provides genome sequences for a
range of model organisms immediately available as Bioconductor
objects.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-04-annotations"><p>Content from <a href="04-annotations.html">Working with annotations</a></p>
<hr>
<p>Last updated on 2025-11-11 |

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/episodes/04-annotations.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<hr>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What Bioconductor packages provides methods to efficiently fetch and
use gene annotations?</li>
<li>How can I use gene annotation packages to convert between different
gene identifiers?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain how gene annotations are managed in the Bioconductor
project.</li>
<li>Identify Bioconductor packages and methods available to fetch and
use gene annotations.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="install-packages">Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a>
</h2>
<hr class="half-width">
<p>Before we can proceed into the following sections, we install some
Bioconductor packages that we will need. First, we check that the <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em>
package is installed before trying to use it; otherwise we install it.
Then we use the <code>BiocManager::install()</code> function to install
the necessary packages.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">requireNamespace</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"biomaRt"</span>, <span class="st">"org.Hs.eg.db"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<hr class="half-width">
<p>Packages dedicated to query gene annotations exist in the ‘Software’
and ‘Annotation’ categories of the Bioconductor <a href="https://www.bioconductor.org/packages/release/BiocViews.html" class="external-link">biocViews</a>,
according to their nature.</p>
<p>In the ‘Software’ section, we find packages that do not actually
contain gene annotations, but rather dynamically <em>query</em> them
from online resources (e.g.,<a href="https://www.ensembl.org/biomart/martview" class="external-link">Ensembl BioMart</a>).
One such Bioconductor package is <em><a href="https://bioconductor.org/packages/3.22/biomaRt" class="external-link">biomaRt</a></em>.</p>
<p>Instead, in the ‘Annotation’ section, we find packages that do
contain annotations. Examples include <em><a href="https://bioconductor.org/packages/3.22/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>,
<em><a href="https://bioconductor.org/packages/3.22/EnsDb.Hsapiens.v86" class="external-link">EnsDb.Hsapiens.v86</a></em>,
and <em><a href="https://bioconductor.org/packages/3.22/TxDb.Hsapiens.UCSC.hg38.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg38.knownGene</a></em>.</p>
<p>In this episode, we will demonstrate the two approaches:</p>
<ul>
<li>Querying annotations from the Ensembl Biomart API using the <em><a href="https://bioconductor.org/packages/3.22/biomaRt" class="external-link">biomaRt</a></em>
package.</li>
<li>Querying annotations from the <em><a href="https://bioconductor.org/packages/3.22/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>
annotation package.</li>
</ul></section><section><h2 class="section-heading" id="online-resources-or-bioconductor-annotation-packages">Online resources or Bioconductor annotation packages?<a class="anchor" aria-label="anchor" href="#online-resources-or-bioconductor-annotation-packages"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="accessing-the-latest-information">Accessing the latest information<a class="anchor" aria-label="anchor" href="#accessing-the-latest-information"></a>
</h3>
<p>Bioconductor’s 6-month release cycle implies that packages available
from the latest stable release branch will not be updated for six months
(only bugfixes are allowed, not functional updates). As a result,
annotation packages may contain information that is out-of-date by up to
six months.</p>
<p>Instead, independent online resources often have different policies
driving the release of updated information. Some databases are
frequently updated, while others may not have been updated in years.</p>
<p>Accessing up-to-date information must also be balanced with
reproducibility. Having downloaded the ‘latest’ information at one point
is time is no good if one hasn’t recorded at least <em>when</em> that
information was downloaded.</p>
</div>
<div class="section level3">
<h3 id="storage-requirements">Storage requirements<a class="anchor" aria-label="anchor" href="#storage-requirements"></a>
</h3>
<p>By nature, Bioconductor annotation packages are larger than software
packages. Just like any other R package, annotation packages must be
installed on the user’s computer before they can be used. This can
rapidly use add up to using an amount of disk space that is not
negligible.</p>
<p>Conversely, online resources are generally accessed programmatically,
and generally only require users to record code to replicate analyses
reproducibly.</p>
</div>
<div class="section level3">
<h3 id="internet-connectivity">Internet connectivity<a class="anchor" aria-label="anchor" href="#internet-connectivity"></a>
</h3>
<p>When using online resources, it is often a good idea to write
annotations ownloaded from online resources to a local file, and refer
to that local file during analyses.</p>
<p>If online resources were to become unavailable for any reason (e.g.,
downtime, loss of internet connection), analyses that use local files
can carry on while those that rely on those online resources cannot.</p>
<p>In contrast, Bioconductor annotation packages only require internet
connectivity at the time of installation. Once installed, they do not
require internet connectivity, as they rely on information stored
locally.</p>
</div>
<div class="section level3">
<h3 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h3>
<p>Bioconductor annotation packages are naturally versioned, meaning
that users can confidently report the version of the package used in
their analysis. Just like software packages, users control if and when
annotation packages should be updated on their computer.</p>
<p>Online resources have different policies to facilitate reproducible
analyses. Some online resources keep archived versions of their
annotations, allowing users to consistently access the same information
over time. When this is not the case, it may be necessary to download a
copy of the annotation at one point in time, and preciously keep that
copy throughout the lifetime of the project to ensure the use of a
consistent set of annotations.</p>
</div>
<div class="section level3">
<h3 id="consistency">Consistency<a class="anchor" aria-label="anchor" href="#consistency"></a>
</h3>
<p>As we will see in the practical examples of this episode,
Bioconductor annotation packages generally re-use a consistent set of
data structures. This allows users familiar with one annotation package
to rapidly get started with others.</p>
<p>Independent online resources often organise their data in different
ways, which requires users to write custom code to access, retrieve, and
process their respective data.</p>
</div>
</section><section><h2 class="section-heading" id="querying-annotations-from-ensembl-biomart">Querying annotations from Ensembl BioMart<a class="anchor" aria-label="anchor" href="#querying-annotations-from-ensembl-biomart"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="the-ensembl-biomart">The Ensembl BioMart<a class="anchor" aria-label="anchor" href="#the-ensembl-biomart"></a>
</h3>
<p><a href="https://www.ensembl.org/biomart/martview" class="external-link">Ensembl
BioMart</a> is a robust data mining tool designed to facilitate access
to the vast array of biological data available through the Ensembl
project.</p>
<p>The <a href="https://www.ensembl.org/biomart/martview" class="external-link">BioMart web
interface</a> enables researchers to efficiently query and retrieve data
on genes, proteins, and other genomic features across multiple species.
It allows users to filter, sort, and export data based on various
attributes such as gene IDs, chromosomal locations, and functional
annotations.</p>
</div>
<div class="section level3">
<h3 id="the-bioconductor-biomart-package">The Bioconductor <code>biomaRt</code> package<a class="anchor" aria-label="anchor" href="#the-bioconductor-biomart-package"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.22/biomaRt" class="external-link">biomaRt</a></em>
is a Bioconductor software package that enables retrieval of large
amounts of data from Ensembl BioMart tables directly from an R session
where those annotations can be used.</p>
<p>Let us first load the package:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">biomaRt</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="listing-available-marts">Listing available marts<a class="anchor" aria-label="anchor" href="#listing-available-marts"></a>
</h3>
<p>Ensembl BioMart organises its diverse biological information into
four databases also known as ‘marts’ or ‘biomarts’. Each mart focuses on
a different type of data.</p>
<p>Users must select the mart corresponds to the type of data they are
interested in before they can query any information from it.</p>
<p>The function <code>listMarts()</code> can be used to display the
names of those marts. This is convenient as users do not need to
memorise the name of the marts, and the function will also return an
updated list of names if any mart is renamed, added, or removed.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">listMarts</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>               biomart                version
1 ENSEMBL_MART_ENSEMBL      Ensembl Genes 115
2   ENSEMBL_MART_MOUSE      Mouse strains 115
3     ENSEMBL_MART_SNP  Ensembl Variation 115
4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 115</code></pre>
</div>
<p>In this demonstration, we will use the biomart called
<code>ENSEMBL_MART_ENSEMBL</code>, which contains the Ensembl gene
set.</p>
<p>Notably, the <code>version</code> columns also indicates the version
of the biomart. The Ensembl BioMart is updated regularly (multiple times
per year). By default, <em><a href="https://bioconductor.org/packages/3.22/biomaRt" class="external-link">biomaRt</a></em>
functions access the latest version of each biomart. This is not ideal
for reproducibility.</p>
<p>Thankfully, Ensembl BioMart archives past versions of its mars in a
way that is accessible both programmatically, and on its website.</p>
<p>The function <code>listEnsemblArchives()</code> can be used to
display all the versions of Ensembl Biomart accessible.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">listEnsemblArchives</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             name     date                                 url version
1  Ensembl GRCh37 Feb 2014          https://grch37.ensembl.org  GRCh37
2     Ensembl 115 Sep 2025 https://sep2025.archive.ensembl.org     115
3     Ensembl 114 May 2025 https://may2025.archive.ensembl.org     114
4     Ensembl 113 Oct 2024 https://oct2024.archive.ensembl.org     113
5     Ensembl 112 May 2024 https://may2024.archive.ensembl.org     112
6     Ensembl 111 Jan 2024 https://jan2024.archive.ensembl.org     111
7     Ensembl 110 Jul 2023 https://jul2023.archive.ensembl.org     110
8     Ensembl 109 Feb 2023 https://feb2023.archive.ensembl.org     109
9     Ensembl 108 Oct 2022 https://oct2022.archive.ensembl.org     108
10    Ensembl 107 Jul 2022 https://jul2022.archive.ensembl.org     107
11    Ensembl 106 Apr 2022 https://apr2022.archive.ensembl.org     106
12    Ensembl 105 Dec 2021 https://dec2021.archive.ensembl.org     105
13    Ensembl 104 May 2021 https://may2021.archive.ensembl.org     104
14    Ensembl 103 Feb 2021 https://feb2021.archive.ensembl.org     103
15    Ensembl 102 Nov 2020 https://nov2020.archive.ensembl.org     102
16    Ensembl 101 Aug 2020 https://aug2020.archive.ensembl.org     101
17    Ensembl 100 Apr 2020 https://apr2020.archive.ensembl.org     100
18     Ensembl 80 May 2015 https://may2015.archive.ensembl.org      80
19     Ensembl 77 Oct 2014 https://oct2014.archive.ensembl.org      77
20     Ensembl 75 Feb 2014 https://feb2014.archive.ensembl.org      75
21     Ensembl 54 May 2009 https://may2009.archive.ensembl.org      54
   current_release
1
2                *
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21                </code></pre>
</div>
<p>In the output above, the key piece of information is the
<code>url</code> column, which provides the URL that <em><a href="https://bioconductor.org/packages/3.22/biomaRt" class="external-link">biomaRt</a></em>
functions will need to access data from the corresponding snapshot of
the Ensembl BioMart.</p>
<p>At the time of writing, the current release is Ensembl 112, so let us
use the corresponding url
<code>https://may2024.archive.ensembl.org</code> to ensure reproducible
results no matter when this lesson is delivered.</p>
</div>
<div class="section level3">
<h3 id="connecting-to-a-biomart">Connecting to a biomart<a class="anchor" aria-label="anchor" href="#connecting-to-a-biomart"></a>
</h3>
<p>The two pieces of information collected above – the name of a biomart
and the URL of a snapshot – is all that is needed to connect to a
BioMart database reproducibly.</p>
<p>The function <code>useMart()</code> can then be used to create a
connection. The connection is traditionally stored in an object called
<code>mart</code>, to be reused in subsequent steps for querying
information from the online mart.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mart</span> <span class="op">&lt;-</span> <span class="fu">useMart</span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>, host <span class="op">=</span> <span class="st">"https://may2024.archive.ensembl.org"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="listing-available-data-sets">Listing available data sets<a class="anchor" aria-label="anchor" href="#listing-available-data-sets"></a>
</h3>
<p>Each biomart contains a number of data sets.</p>
<p>The function <code>listDatasets()</code> can be used to display
information about those data sets. This is convenient as users do not
need to memorise the name of the data sets, and the information returned
by the function includes a short description of each data set, as well
as its version.</p>
<p>In the example below, we restrict the output table to the first few
rows, as the full table comprises 214 rows.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">listDatasets</span><span class="op">(</span><span class="va">mart</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                       dataset                           description
1 abrachyrhynchus_gene_ensembl Pink-footed goose genes (ASM259213v1)
2     acalliptera_gene_ensembl      Eastern happy genes (fAstCal1.3)
3   acarolinensis_gene_ensembl       Green anole genes (AnoCar2.0v2)
4    acchrysaetos_gene_ensembl       Golden eagle genes (bAquChr1.2)
5    acitrinellus_gene_ensembl        Midas cichlid genes (Midas_v5)
6    amelanoleuca_gene_ensembl       Giant panda genes (ASM200744v2)
      version
1 ASM259213v1
2  fAstCal1.3
3 AnoCar2.0v2
4  bAquChr1.2
5    Midas_v5
6 ASM200744v2</code></pre>
</div>
<p>In the output above, the key piece of information is the
<code>dataset</code> column, which provides the identifier that <em><a href="https://bioconductor.org/packages/3.22/biomaRt" class="external-link">biomaRt</a></em>
functions will need to access data from the corresponding biomart
table.</p>
<p>In this demonstration, we will use the Ensembl gene set for Homo
sapiens, which is not visible in the output above.</p>
<p>Given the number of data sets available, let us programmatically
filter the table of information using pattern matching rather than
searching the table manually:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">subset</span><span class="op">(</span><span class="fu">listDatasets</span><span class="op">(</span><span class="va">mart</span><span class="op">)</span>, <span class="fu">grepl</span><span class="op">(</span><span class="st">"sapiens"</span>, <span class="va">dataset</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                 dataset              description    version
80 hsapiens_gene_ensembl Human genes (GRCh38.p14) GRCh38.p14</code></pre>
</div>
<p>From the output above, we identify the desired data set identifier as
<code>hsapiens_gene_ensembl</code>.</p>
</div>
<div class="section level3">
<h3 id="connecting-to-a-data-set">Connecting to a data set<a class="anchor" aria-label="anchor" href="#connecting-to-a-data-set"></a>
</h3>
<p>Having chosen the data set that we want to use, we need to call the
function <code>useMart()</code> again, this time specifying the selected
data set.</p>
<p>Typically, one would copy paste the previous call to
<code>useMart()</code> and edit as needed. It is also common practice to
replace the <code>mart</code> object with the new connection.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mart</span> <span class="op">&lt;-</span> <span class="fu">useMart</span><span class="op">(</span></span>
<span>  biomart <span class="op">=</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>,</span>
<span>  dataset <span class="op">=</span> <span class="st">"hsapiens_gene_ensembl"</span>,</span>
<span>  host <span class="op">=</span> <span class="st">"https://may2024.archive.ensembl.org"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="listing-information-available-in-a-data-set">Listing information available in a data set<a class="anchor" aria-label="anchor" href="#listing-information-available-in-a-data-set"></a>
</h3>
<p>BioMart tables contain many pieces of information also known as
‘attributes’. So many, in fact, that they have been grouped into
categories also known as ‘pages’.</p>
<p>The function <code>listAttributes()</code> can be used to display
information about those attributes. This is convenient as users do not
need to memorise the name of the attributes, and the information
returned by the function includes a short description of each attribute,
as well as its page categorisation.</p>
<p>In the example below, we restrict the output table to the first few
rows, as the full table comprises 3157 rows.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">listAttributes</span><span class="op">(</span><span class="va">mart</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                           name                  description         page
1               ensembl_gene_id               Gene stable ID feature_page
2       ensembl_gene_id_version       Gene stable ID version feature_page
3         ensembl_transcript_id         Transcript stable ID feature_page
4 ensembl_transcript_id_version Transcript stable ID version feature_page
5            ensembl_peptide_id            Protein stable ID feature_page
6    ensembl_peptide_id_version    Protein stable ID version feature_page</code></pre>
</div>
<p>In the output above, the key piece of information is the
<code>name</code> column, which provides the identifier that <em><a href="https://bioconductor.org/packages/3.22/biomaRt" class="external-link">biomaRt</a></em>
functions will need to query that information from the corresponding
biomart data set.</p>
<p>The choice of attributes to query now depends on what it is we wish
to achieve.</p>
<p>For instance, let us imagine that we have a set of gene identifiers,
for which we wish to query:</p>
<ul>
<li>The gene symbol</li>
<li>The name of the chromosome where the gene is located</li>
<li>The start and end position of the gene on that chromosome</li>
<li>The strand on which the gene is encoded</li>
</ul>
<p>Users would often manually explore the full table of attributes to
identify the ones they wish to include in their query. It is also
possible to programmatically filter the table of attribute, based on
experience and intuition, to narrow down the search:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">subset</span><span class="op">(</span><span class="fu">listAttributes</span><span class="op">(</span><span class="va">mart</span><span class="op">)</span>, <span class="fu">grepl</span><span class="op">(</span><span class="st">"position"</span>, <span class="va">name</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu">grepl</span><span class="op">(</span><span class="st">"feature"</span>, <span class="va">page</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             name     description         page
10 start_position Gene start (bp) feature_page
11   end_position   Gene end (bp) feature_page</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="querying-information-from-a-biomart-table">Querying information from a BioMart table<a class="anchor" aria-label="anchor" href="#querying-information-from-a-biomart-table"></a>
</h3>
<p>We have now all the information that we need to perform the actual
query:</p>
<ul>
<li>A connection to a BioMart data set</li>
<li>The list of attributes available in that data set</li>
</ul>
<p>The function <code>getBM()</code> is the main <em><a href="https://bioconductor.org/packages/3.22/biomaRt" class="external-link">biomaRt</a></em>
query function. Given a set of filters and corresponding values, it
retrieves the attributes requested by the user from the BioMart data set
it is connected to.</p>
<p>In the example below, we manually create a vector of arbitrary gene
identifiers for our query. In practice, the query will often originate
from an earlier analysis (e.g., differential gene expression).</p>
<p>The example below also queries attributes that we have not introduced
yet. In the previous section, we described how one may search the table
of attributes returned by <code>listAttributes()</code> to identify
attributes to include in their query.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_gene_ids</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span></span>
<span>  <span class="st">"ENSG00000133101"</span>,</span>
<span>  <span class="st">"ENSG00000145386"</span>,</span>
<span>  <span class="st">"ENSG00000134057"</span>,</span>
<span>  <span class="st">"ENSG00000157456"</span>,</span>
<span>  <span class="st">"ENSG00000147082"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">getBM</span><span class="op">(</span></span>
<span>  attributes <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"ensembl_gene_id"</span>,</span>
<span>    <span class="st">"hgnc_symbol"</span>,</span>
<span>    <span class="st">"chromosome_name"</span>,</span>
<span>    <span class="st">"start_position"</span>,</span>
<span>    <span class="st">"end_position"</span>,</span>
<span>    <span class="st">"strand"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  filters <span class="op">=</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span>  values <span class="op">=</span> <span class="va">query_gene_ids</span>,</span>
<span>  mart <span class="op">=</span> <span class="va">mart</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  ensembl_gene_id hgnc_symbol chromosome_name start_position end_position
1 ENSG00000133101       CCNA1              13       36431520     36442870
2 ENSG00000134057       CCNB1               5       69167135     69178245
3 ENSG00000145386       CCNA2               4      121816444    121823883
4 ENSG00000147082       CCNB3               X       50202713     50351914
5 ENSG00000157456       CCNB2              15       59105126     59125045
  strand
1      1
2      1
3     -1
4      1
5      1</code></pre>
</div>
<p>Note that we also included the filtering attribute
<code>ensembl_gene_id</code> to the attributes retrieved from the data
set. This is key to reliably match the newly retrieved attributes to
those used in the query.</p>
</div>
</section><section><h2 class="section-heading" id="querying-annotations-from-annotation-packages">Querying annotations from annotation packages<a class="anchor" aria-label="anchor" href="#querying-annotations-from-annotation-packages"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="families-of-annotation-packages">Families of annotation packages<a class="anchor" aria-label="anchor" href="#families-of-annotation-packages"></a>
</h3>
<p>To balance the need for comprehensive information while maintaining
reasonable package sizes, Bioconductor annotation packages are organised
by release, data type, and species.</p>
<p>The major families of Bioconductor annotation packages are:</p>
<ul>
<li>
<code>OrgDb</code> packages provide mapping between various types of
gene identifiers and pathway information.</li>
<li>
<code>EnsDb</code> packages provide individual releases of Ensembl
annotations.</li>
<li>
<code>TxDb</code> packages provide individual releases of UCSC
annotations.</li>
</ul>
<p>All those families of annotations derive from the
<code>AnnotationDb</code> base class defined in the <em><a href="https://bioconductor.org/packages/3.22/AnnotationDbi" class="external-link">AnnotationDbi</a></em>
package. As a result, any of those annotation packages can be accessed
using the same set of R functions, as demonstrated in the following
sections.</p>
</div>
<div class="section level3">
<h3 id="using-an-orgdb-package">Using an OrgDb package<a class="anchor" aria-label="anchor" href="#using-an-orgdb-package"></a>
</h3>
<p>In this example, we will use the <em><a href="https://bioconductor.org/packages/3.22/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>
package to demonstrate the use of gene annotations for the human
species.</p>
<p>Let us first load the package:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<p>Each <code>OrgDb</code> package contains an object named identically
to the package itself. That object contains the annotations that the
package is meant to disseminate.</p>
<p>Aside from querying information, the whole object can be called to
print information about the annotations it contains, including the date
at which the snapshots of annotations that it contains were made.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">org.Hs.eg.db</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>OrgDb object:
| DBSCHEMAVERSION: 2.1
| Db type: OrgDb
| Supporting package: AnnotationDbi
| DBSCHEMA: HUMAN_DB
| ORGANISM: Homo sapiens
| SPECIES: Human
| EGSOURCEDATE: 2025-Sep24
| EGSOURCENAME: Entrez Gene
| EGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA
| CENTRALID: EG
| TAXID: 9606
| GOSOURCENAME: Gene Ontology
| GOSOURCEURL: http://current.geneontology.org/ontology/go-basic.obo
| GOSOURCEDATE: 2025-07-22
| GOEGSOURCEDATE: 2025-Sep24
| GOEGSOURCENAME: Entrez Gene
| GOEGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA
| KEGGSOURCENAME: KEGG GENOME
| KEGGSOURCEURL: ftp://ftp.genome.jp/pub/kegg/genomes
| KEGGSOURCEDATE: 2011-Mar15
| GPSOURCENAME: UCSC Genome Bioinformatics (Homo sapiens)
| GPSOURCEURL: ftp://hgdownload.cse.ucsc.edu/goldenPath/hg38/database
| GPSOURCEDATE: UTC-Sep26
| ENSOURCEDATE: 2025-Sep03
| ENSOURCENAME: Ensembl
| ENSOURCEURL: ftp://ftp.ensembl.org/pub/current_fasta
| UPSOURCENAME: Uniprot
| UPSOURCEURL: http://www.UniProt.org/
| UPSOURCEDATE: Sat Oct  4 09:48:11 2025</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Please see: help('select') for usage information</code></pre>
</div>
<p>That same object is the one that needs to be supplied to <em><a href="https://bioconductor.org/packages/3.22/AnnotationDbi" class="external-link">AnnotationDbi</a></em>
functions for running queries and retrieving annotations.</p>
</div>
<div class="section level3">
<h3 id="listing-information-available-in-an-annotation-package">Listing information available in an annotation package<a class="anchor" aria-label="anchor" href="#listing-information-available-in-an-annotation-package"></a>
</h3>
<p>The function <code>columns()</code> can be used to display the
annotations available in the object.</p>
<p>Here, the word ‘column’ refers to columns of tables used to store
information in database, the very same concept as ‘attributes’ in
BioMart. In other words, columns represent all the types of annotations
that may be retrieved from the object.</p>
<p>This is convenient as users do not need to memorise the names of the
columns of annotations available in the package.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">columns</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
 [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"
[11] "GENETYPE"     "GO"           "GOALL"        "IPI"          "MAP"
[16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"
[21] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"
[26] "UNIPROT"     </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="listing-keys-and-key-types">Listing keys and key types<a class="anchor" aria-label="anchor" href="#listing-keys-and-key-types"></a>
</h3>
<p>In database terminology, <em>keys</em> are the values by which
information may be queried from a database table.</p>
<p>Information being organised in columns, <em>key types</em> are the
names of the columns in which the key values are stored.</p>
<p>Given the variable number of columns in database tables, some tables
may allow information to be queried by more than one key. As a result,
it is crucial to specify both the keys and the type of key as part of
the query.</p>
<p>The function <code>keytypes()</code> can be used to display the names
of the columns that may be used to query information from the
object.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">keytypes</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
 [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"
[11] "GENETYPE"     "GO"           "GOALL"        "IPI"          "MAP"
[16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"
[21] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"
[26] "UNIPROT"     </code></pre>
</div>
<p>The function <code>keys()</code> can be used to display all the
possible values for a given key type.</p>
<p>It is generally better practice to specify the type of key being
queried (to avoid ambiguity), although database tables typically have a
‘primary key’ used if users do not specify a type themselves.</p>
<p>In the example below, we restrict the list of gene symbol keys to the
first few values, as the full set comprises 193512 values.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">keys</span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "A1BG"     "A2M"      "NAT1"     "NAT2"     "NATP"     "SERPINA3"</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="querying-information-from-an-annotation-package">Querying information from an annotation package<a class="anchor" aria-label="anchor" href="#querying-information-from-an-annotation-package"></a>
</h3>
<p>The function <code>select()</code> is the main <em><a href="https://bioconductor.org/packages/3.22/AnnotationDbi" class="external-link">AnnotationDbi</a></em>
query function. Given an <code>AnnotationDb</code> object, key values,
and columns (and optionally the type of key supplied if not the primary
key), it retrieves the columns requested by the user from the annotation
object.</p>
<p>In the example below, we re-use the vector of arbitrary gene
identifiers used in the BioMart example a few sections above.</p>
<p>As you can see from the output of the <code>columns()</code>
function, the annotation object does not contain some of the attributes
that we queried in the Biomart example. In this case, let us query:</p>
<ul>
<li>the gene symbol</li>
<li>the gene name</li>
<li>the gene type</li>
</ul>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>  keys <span class="op">=</span> <span class="va">query_gene_ids</span>,</span>
<span>  columns <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"SYMBOL"</span>,</span>
<span>    <span class="st">"GENENAME"</span>,</span>
<span>    <span class="st">"GENETYPE"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>          ENSEMBL SYMBOL  GENENAME       GENETYPE
1 ENSG00000133101  CCNA1 cyclin A1 protein-coding
2 ENSG00000145386  CCNA2 cyclin A2 protein-coding
3 ENSG00000134057  CCNB1 cyclin B1 protein-coding
4 ENSG00000157456  CCNB2 cyclin B2 protein-coding
5 ENSG00000147082  CCNB3 cyclin B3 protein-coding</code></pre>
</div>
<p>One small but notable difference with <em><a href="https://bioconductor.org/packages/3.22/biomaRt" class="external-link">biomaRt</a></em>
is that the output of <code>select()</code> automatically contains the
column that correspond to the key type used in the query. In other
words, there is no need to specify the key type(s) again in the
column(s) to retrieve.</p>
</div>
<div class="section level3">
<h3 id="vectorized-11-mapping">Vectorized 1:1 mapping<a class="anchor" aria-label="anchor" href="#vectorized-11-mapping"></a>
</h3>
<p>It is sometimes possible for annotations to display 1-to-many
relationships. For instance, individual genes typically have a unique
Ensembl gene identifier, while they may be known under multiple gene
name aliases.</p>
<p>The <code>select()</code> function demonstrated in the previous
section automatically returns <em>all</em> values in the columns
requested, for the key specified. This is possible thanks to the tabular
format in which annotations are returned; rows are added, repeating
values as necessary to display them on the same row as every other
values they are associated with.</p>
<p>In some cases, that behaviour is not desirable. Instead, users may
wish to retrieve a single value for each key that they input. One common
scenario arises during differential gene expression (DGE), where gene
identifiers are used to uniquely identify genes throughout the analysis,
while gene symbols are added to the final table of DGE statistics, to
provide more readable human-friendly gene identifiers. However, it is
not desirable to duplicate rows of DGE statistics, and thus only a
single gene symbol is required to annotate each gene.</p>
<p>The function <code>mapIds()</code> can be used for this purpose. A
major difference between the functions <code>mapIds()</code> and
<code>select()</code> are their arguments <code>column</code> (singular)
and <code>columns</code> (plural), respectively. The function
<code>mapIds()</code> accepts a single column name and returns a named
character vector where names are the input query values, and values are
the corresponding values in the requested column.</p>
<p>To deal with 1-to-many relationships, the function
<code>mapIds()</code> has an argument <code>multiVals</code> which can
be used to specify how the function should handle multiple values. The
default is to take the first value and ignore any other value.</p>
<p>In the example below, we query the gene symbol for a set of Ensembl
gene identifiers.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mapIds</span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>  keys <span class="op">=</span> <span class="va">query_gene_ids</span>,</span>
<span>  column <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>  keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>ENSG00000133101 ENSG00000145386 ENSG00000134057 ENSG00000157456 ENSG00000147082
        "CCNA1"         "CCNA2"         "CCNB1"         "CCNB2"         "CCNB3" </code></pre>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Load the packages <em><a href="https://bioconductor.org/packages/3.22/EnsDb.Hsapiens.v86" class="external-link">EnsDb.Hsapiens.v86</a></em>
and <em><a href="https://bioconductor.org/packages/3.22/TxDb.Hsapiens.UCSC.hg38.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg38.knownGene</a></em>.
Then, display the columns of annotations available in those
packages.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="fu">columns</span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "ENTREZID"            "EXONID"              "EXONIDX"
 [4] "EXONSEQEND"          "EXONSEQSTART"        "GENEBIOTYPE"
 [7] "GENEID"              "GENENAME"            "GENESEQEND"
[10] "GENESEQSTART"        "INTERPROACCESSION"   "ISCIRCULAR"
[13] "PROTDOMEND"          "PROTDOMSTART"        "PROTEINDOMAINID"
[16] "PROTEINDOMAINSOURCE" "PROTEINID"           "PROTEINSEQUENCE"
[19] "SEQCOORDSYSTEM"      "SEQLENGTH"           "SEQNAME"
[22] "SEQSTRAND"           "SYMBOL"              "TXBIOTYPE"
[25] "TXCDSSEQEND"         "TXCDSSEQSTART"       "TXID"
[28] "TXNAME"              "TXSEQEND"            "TXSEQSTART"
[31] "UNIPROTDB"           "UNIPROTID"           "UNIPROTMAPPINGTYPE" </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span></span>
<span><span class="fu">columns</span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "CDSCHROM"   "CDSEND"     "CDSID"      "CDSNAME"    "CDSPHASE"
 [6] "CDSSTART"   "CDSSTRAND"  "EXONCHROM"  "EXONEND"    "EXONID"
[11] "EXONNAME"   "EXONRANK"   "EXONSTART"  "EXONSTRAND" "GENEID"
[16] "TXCHROM"    "TXEND"      "TXID"       "TXNAME"     "TXSTART"
[21] "TXSTRAND"   "TXTYPE"    </code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Bioconductor provides a wide range annotation packages.</li>
<li>Some Bioconductor software packages can be used to programmatically
access online resources.</li>
<li>Users should carefully choose their source of annotations based on
their needs and expectations.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-05-s4"><p>Content from <a href="05-s4.html">The S4 class</a></p>
<hr>
<p>Last updated on 2025-11-11 |

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/episodes/05-s4.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>A merged episodes from <a href="https://carpentries-incubator.github.io/bioc-project/05-s4.html" class="external-link">S4
classes in Bioconductor</a> and <a href="https://carpentries-incubator.github.io/bioc-project/09-summarizedexperiment.html" class="external-link">The
SummarizedExperiment class</a> from the carpentries incibator lesson <a href="https://carpentries-incubator.github.io/bioc-project/" class="external-link">Bioconductor
project</a>.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is the S4 class system?</li>
<li>How does Bioconductor use S4 classes?</li>
<li>How is the Bioconductor <code>DataFrame</code> different from the
base <code>data.frame</code>?</li>
<li>How is information organized in SummarizedExperiment objects?</li>
<li>How can that information be added, edited, and accessed?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain what S4 classes, generics, and methods are.</li>
<li>Identify S4 classes at the core of the Bioconductor package
infrastructure.</li>
<li>Create various S4 objects and apply relevant S4 methods.</li>
<li>Describe how both experimental data and metadata can be stored in a
single object.</li>
<li>Explain why this is crucial to keep data and metadata synchronised
throughout analyses.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="install-packages">Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a>
</h2>
<hr class="half-width">
<p>Before we can proceed into the following sections, we install some
Bioconductor packages that we will need. First, we check that the <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em>
package is installed before trying to use it; otherwise we install it.
Then we use the <code>BiocManager::install()</code> function to install
the necessary packages.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">requireNamespace</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"S4Vectors"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"SummarizedExperiment"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="s4-classes-and-methods">S4 classes and methods<a class="anchor" aria-label="anchor" href="#s4-classes-and-methods"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="the-methods-package">The methods package<a class="anchor" aria-label="anchor" href="#the-methods-package"></a>
</h3>
<p>The S4 class system is implemented in the base package <a href="https://stat.ethz.ch/R-manual/R-devel/library/methods/html/00Index.html" class="external-link">methods</a>.
As such, the concept is not specific to the Bioconductor project and can
be found in various independent packages as well. The subject is
thoroughly documented in the online book <a href="https://adv-r.had.co.nz/S4.html" class="external-link">Advanced R</a>, by Hadley
Wickham. Most Bioconductor users will never need to get overly familiar
with the intricacies of the S4 class system. Rather, the key to an
efficient use of packages in the Bioconductor project relies on a
sufficient understanding of key motivations for using the S4 class
system, as well as best practices for user-facing functionality,
including classes, generics, and methods. In the following sections of
this episode, we focus on the essential functionality and user
experience of S4 classes and methods in the context of the Bioconductor
project.</p>
<p>On one side, <a href="reference.html#s4-class">S4 classes</a> provide
data structures capable of storing arbitrarily complex information in
computational objects that can be assigned to variable names in an R
session. On the other side, <a href="reference.html#s4-method">S4
generics and methods</a> define functions that may be applied to process
those objects.</p>
<p>Over the years, the Bioconductor project has used the S4 class system
to develop a number of classes and methods capable of storing and
processing data for most biological assays, including raw and processed
assay data, experimental metadata for individual features and samples,
as well as other assay-specific information as relevant. Gaining
familiarity with the standard S4 classes commonly used throughout
Bioconductor packages is a key step in building up confidence in users
wishing to follow best practices while developing analytical
workflows.</p>
<figure><img src="fig/bioc-s4.svg" alt="S4 classes, generics, and methods." class="figure mx-auto d-block"></figure><p><strong>S4 classes, generics, and methods.</strong> On the left, two
example classes named <code>S4Class1</code> and <code>S4Class2</code>
demonstrate the concept of inheritance. The class <code>S4Class1</code>
contains two slots named <code>SlotName1</code> and
<code>SlotName2</code> for storing data. Those two slots are restricted
to store objects of type <code>SlotType1</code> and
<code>SlotType2</code>, respectively. The class also defines validity
rules that check the integrity of data each time an object is updated.
The class <code>S4Class2</code> inherits all the slots and validity
rules from the class <code>S4Class1</code>, in addition to defining a
new slot named <code>SlotName3</code> and new validity rules. Example
code illustrates how objects of each class are typically created using
constructor functions named identically to the corresponding class. On
the right, one generic function and two methods demonstrate the concept
of polymorphism and the process of S4 method dispatch. The generic
function <code>S4Generic1()</code> defines the name of the function, as
well as its arguments. However, it does not provide any implementation
of that function. Instead, two methods are defined, each providing a
distinct implementation of the generic function for a particular class
of input. Namely, the first method defines an implementation of
<code>S4Generic1()</code> if an object of class <code>S4Class1</code> is
given as argument <code>x</code>, while the second method method
provides a different implementation of <code>S4Generic1()</code> if an
object of class <code>S4Class2</code> is given as argument
<code>x</code>. When the generic function <code>S4Generic1()</code> is
called, a process called <a href="reference.html#s4-method-dispatch">method dispatch</a> takes
place, whereby the appropriate implementation of the
<code>S4Generic1()</code> method is called according to the class of the
object passed to the argument <code>x</code>.</p>
</div>
<div class="section level3">
<h3 id="slots-and-validity">Slots and validity<a class="anchor" aria-label="anchor" href="#slots-and-validity"></a>
</h3>
<p>In contrast to the S3 class system available directly in base R (not
described in this lesson), the S4 class system provides a much stricter
definition of classes and methods for object-oriented programming (OOP)
in R. Like many programming languages that implement the OOP model, S4
classes are used to represent real-world entities as computational
objects that store information inside one or more internal components
called <em>slots</em>. The class definition declares the type of data
that may be stored in each slot; an error will be thrown if one would
attempt to store unsuitable data. Moreover, the class definition can
also include code that checks the validity of data stored in an object,
beyond their type. For instance, while a slot of type
<code>numeric</code> could be used to store a person’s age, but a
validity method could check that the value stored is, in fact,
positive.</p>
</div>
<div class="section level3">
<h3 id="inheritance">Inheritance<a class="anchor" aria-label="anchor" href="#inheritance"></a>
</h3>
<p>One of the core pillars of the OOP model is the possibility to
develop new classes that inherit and extend the functionality of
existing classes. The S4 class system implements this paradigm.</p>
<p>The definition of a new S4 classes can declare the name of other
classes to inherit from. The new classes will contain all the slots of
the parent class, in addition to any new slot added in the definition of
the new class itself.</p>
<p>The new class definition can also define new validity checks, which
are added to any validity check implement in each of the parent
classes.</p>
</div>
<div class="section level3">
<h3 id="generics-and-methods">Generics and methods<a class="anchor" aria-label="anchor" href="#generics-and-methods"></a>
</h3>
<p>While classes define the data structures that store information,
generics and methods define the functions that can be applied to objects
instantiated from those classes.</p>
<p>S4 generic functions are used to declare the name of functions that
are expected to behave differently depending on the class of objects
that are given as some of their essential arguments. Instead, S4 methods
are used define the distinct implementations of a generic function for
each particular combination of inputs.</p>
<p>When a generic function is called and given an S4 object, a process
called <a href="reference.html#s4-method-dispatch">method dispatch</a>
takes place, whereby the class of the object is used to determine the
appropriate method to execute.</p>
</div>
</section><section><h2 class="section-heading" id="the-s4vectors-package">The S4Vectors package<a class="anchor" aria-label="anchor" href="#the-s4vectors-package"></a>
</h2>
<hr class="half-width">
<p>The <em><a href="https://bioconductor.org/packages/3.22/S4Vectors" class="external-link">S4Vectors</a></em>
package defines the <code>Vector</code> and <code>List</code> virtual
classes and a set of generic functions that extend the semantic of
ordinary vectors and lists in R. Using the S4 class system, package
developers can easily implement vector-like or list-like objects as
concrete subclasses of <code>Vector</code> or <code>List</code>.</p>
<p>Virtual classes – such as <code>Vector</code> and <code>List</code> –
cannot be instantiated into objects themselves. Instead, those virtual
classes provide core functionality inherited by all the concrete classes
that are derived from them.</p>
<p>Instead, a few low-level concrete subclasses of general interest
(e.g. <code>DataFrame</code>, <code>Rle</code>, and <code>Hits</code>)
are implemented in the <em><a href="https://bioconductor.org/packages/3.22/S4Vectors" class="external-link">S4Vectors</a></em>
package itself, and many more are implemented in other packages
throughout the Bioconductor project (e.g., <em><a href="https://bioconductor.org/packages/3.22/IRanges" class="external-link">IRanges</a></em>).</p>
<p>Attach the package to the current R session as follows.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">S4Vectors</span><span class="op">)</span></span></code></pre>
</div>
<div id="note" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="note" class="callout-inner">
<h3 class="callout-title">Note</h3>
<div class="callout-content">
<p>The package startup messages printed in the console are worth noting
that the <em><a href="https://bioconductor.org/packages/3.22/S4Vectors" class="external-link">S4Vectors</a></em>
package masks a number of functions from the <a href="https://stat.ethz.ch/R-manual/R-devel/library/base/html/00Index.html" class="external-link">base</a>
package when the package is attached to the session. This means that the
<em><a href="https://bioconductor.org/packages/3.22/S4Vectors" class="external-link">S4Vectors</a></em>
package includes an implementation of those functions, and that – being
the latest package attached to the R session – its own implementation of
those functions will be found first on the R search path and used
instead of their original implementation in the <a href="https://stat.ethz.ch/R-manual/R-devel/library/base/html/00Index.html" class="external-link">base</a>
package.</p>
<p>In many cases, masked functions can be used as before without any
issue. Occasionally, it may be necessary to disambiguate calls to masked
function using the package name as well as the function name,
e.g. <code>base::anyDuplicated()</code>.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="the-dataframe-class">The DataFrame class<a class="anchor" aria-label="anchor" href="#the-dataframe-class"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="an-extension-to-the-concept-of-rectangular-data">An extension to the concept of rectangular data<a class="anchor" aria-label="anchor" href="#an-extension-to-the-concept-of-rectangular-data"></a>
</h3>
<p>The <code>DataFrame</code> class implemented in the <em><a href="https://bioconductor.org/packages/3.22/S4Vectors" class="external-link">S4Vectors</a></em>
package extends the concept of rectangular data familiar to users of the
<code>data.frame</code> class in base R, or <code>tibble</code> in the
tidyverse. Specifically, the <code>DataFrame</code> supports the storage
of any type of object (with <code>length</code> and <code>[</code>
methods) as columns.</p>
<p>On the whole, the <code>DataFrame</code> class provides a formal
definition of an S4 class that behaves very similarly to
<code>data.frame</code>, in terms of construction, subsetting,
splitting, combining, etc.</p>
<p>The <code>DataFrame()</code> constructor function should be used to
create new objects, comparably to the <code>data.frame()</code>
equivalent in base R. The help page for the function, accessible as
<code>?DataFrame</code>, can be consulted for more information.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF1</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>    Integers <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1L</span>, <span class="fl">2L</span>, <span class="fl">3L</span><span class="op">)</span>,</span>
<span>    Letters <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span>,</span>
<span>    Floats <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1.2</span>, <span class="fl">2.3</span>, <span class="fl">3.4</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">DF1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3 rows and 3 columns
   Integers     Letters    Floats
  &lt;integer&gt; &lt;character&gt; &lt;numeric&gt;
1         1           A       1.2
2         2           B       2.3
3         3           C       3.4</code></pre>
</div>
<p>In fact, <code>DataFrame</code> objects can be easily converted to
equivalent <code>data.frame</code> objects.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df1</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">DF1</span><span class="op">)</span></span>
<span><span class="va">df1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  Integers Letters Floats
1        1       A    1.2
2        2       B    2.3
3        3       C    3.4</code></pre>
</div>
<p>Vice versa, we can also convert <code>data.frame</code> objects to
<code>DataFrame</code> using the <code>as()</code> function.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">as</span><span class="op">(</span><span class="va">df1</span>, <span class="st">"DataFrame"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3 rows and 3 columns
   Integers     Letters    Floats
  &lt;integer&gt; &lt;character&gt; &lt;numeric&gt;
1         1           A       1.2
2         2           B       2.3
3         3           C       3.4</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="differences-with-the-base-data-frame">Differences with the base data.frame<a class="anchor" aria-label="anchor" href="#differences-with-the-base-data-frame"></a>
</h3>
<p>The most notable exceptions have to do with handling of row names.
First, row names are optional. This means calling
<code>rownames(x)</code> will return <code>NULL</code> if there are no
row names.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rownames</span><span class="op">(</span><span class="va">DF1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="cn">NULL</span></span></code></pre>
</div>
<p>This is different from <code>data.frame</code>, where
<code>rownames(x)</code> returns the equivalent of
<code>as.character(seq_len(nrow(x)))</code>.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rownames</span><span class="op">(</span><span class="va">df1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "1" "2" "3"</code></pre>
</div>
<p>However, returning <code>NULL</code> informs, for example,
combination functions that no row names are desired (they are often a
luxury when dealing with large data).</p>
<p>Furthermore, row names of <code>DataFrame</code> objects are not
required to be unique, in contrast to the <code>data.frame</code> in
base R. Row names are a frequent source of controversy in R, as they can
be used to uniquely identify and index observations in rectangular
format, without storing that information explicitly in a dedicated
column. When set, row names can be used to subset rectangular data using
the <code>[</code> operator. Meanwhile, non-unique row names defeat that
purpose and can lead to unexpected results, as only the first occurrence
of each selected row name is extracted. Instead, the tidyverse
<code>tibble</code> removed the ability to set row names altogether,
forcing users to store every bit of information explicitly in dedicated
columns, while providing functions to dedicated to efficiently filtering
rows in rectangular data, without the need for the <code>[</code>
operator.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF2</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>    Integers <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1L</span>, <span class="fl">2L</span>, <span class="fl">3L</span><span class="op">)</span>,</span>
<span>    Letters <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span>,</span>
<span>    Floats <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1.2</span>, <span class="fl">2.3</span>, <span class="fl">3.4</span><span class="op">)</span>,</span>
<span>    row.names <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"name1"</span>, <span class="st">"name1"</span>, <span class="st">"name2"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">DF2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3 rows and 3 columns
       Integers     Letters    Floats
      &lt;integer&gt; &lt;character&gt; &lt;numeric&gt;
name1         1           A       1.2
name1         2           B       2.3
name2         3           C       3.4</code></pre>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Using the example above, what does <code>DF2["name1", ]</code>
return? Why?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<pre><code>&gt; DF2["name1", ]
DataFrame with 1 row and 3 columns
       Integers     Letters    Floats
      &lt;integer&gt; &lt;character&gt; &lt;numeric&gt;
name1         1           A       1.2</code></pre>
<p>Only the first occurrence of a row matching the row name
<code>name1</code> is returned.</p>
<p>In this case, row names do not have a particular meaning, making it
difficult to justify the need for them. Instead, users could extract all
the rows that matching the row name <code>name1</code> more explicitly
as follows: <code>DF2[rownames(DF2) == "name1", ]</code>.</p>
<p>Users should be mindful of the motivation for using row names in any
given situation; what they represent, and how they should be used during
the analysis.</p>
</div>
</div>
</div>
</div>
<p>Finally, row names in <code>DataFrame</code> do not support partial
matching during subsetting, in contrast to <code>data.frame</code>. The
stricter behaviour of <code>DataFrame</code> prevents often unexpected
results faced by unsuspecting users.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF3</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>    Integers <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1L</span>, <span class="fl">2L</span>, <span class="fl">3L</span><span class="op">)</span>,</span>
<span>    Letters <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span>,</span>
<span>    Floats <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1.2</span>, <span class="fl">2.3</span>, <span class="fl">3.4</span><span class="op">)</span>,</span>
<span>    row.names <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"alpha"</span>, <span class="st">"beta"</span>, <span class="st">"gamma"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">df3</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">DF3</span><span class="op">)</span></span></code></pre>
</div>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Using the examples above, what are the outputs of
<code>DF3["a", ]</code> and <code>df3["a", ]</code>? Why are they
different?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<pre><code>&gt; DF3["a", ]
DataFrame with 1 row and 3 columns
      Integers     Letters    Floats
     &lt;integer&gt; &lt;character&gt; &lt;numeric&gt;
&lt;NA&gt;        NA          NA        NA
&gt; df3["a", ]
      Integers Letters Floats
alpha        1       A    1.2</code></pre>
<p>The <code>DataFrame</code> object did not perform partial row name
matching, and thus did not match any row and return a
<code>DataFrame</code> full of <code>NA</code> values. Instead, the
<code>data.frame</code> object performed partial row name matching,
matched the requested <code>"a"</code> to the <code>"alpha"</code> row
name, and returned the corresponding row as a new
<code>data.frame</code> object.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="indexing">Indexing<a class="anchor" aria-label="anchor" href="#indexing"></a>
</h3>
<p>Just like a regular <code>data.frame</code>, columns can be accessed
using <code>$</code>, <code>[</code>, and <code>[[</code>. Each operator
has a different purpose, and the most appropriate one will often depend
on what you are trying to achieve.</p>
<p>For example, the dollar operator <code>$</code> can be used to
extract a single column by name. That will often be a vector, but it may
depend on the nature of the data in that column. This operator can be
quite convenient in an interactive R session, as it will offer
autocompletion among available column names.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF3</span><span class="op">$</span><span class="va">Integers</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 2 3</code></pre>
</div>
<p>Similarly, the double bracket operator <code>[[</code> can also be
used to extract a single column. It is more flexible than <code>$</code>
as it can handle both character names and integer indices.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF3</span><span class="op">[[</span><span class="st">"Letters"</span><span class="op">]</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "A" "B" "C"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF3</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "A" "B" "C"</code></pre>
</div>
<p>The operator <code>[</code> is most convenient when it comes to
selecting simultaneously on rows and columns, or controlling whether a
single-column selection should be returned as a <code>DataFrame</code>
or a <code>vector</code>.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF3</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span>, <span class="st">"Letters"</span>, drop<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 2 rows and 1 column
          Letters
      &lt;character&gt;
beta            B
gamma           C</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="metadata-columns">Metadata columns<a class="anchor" aria-label="anchor" href="#metadata-columns"></a>
</h3>
<p>One of most notable novel functionality in <code>DataFrame</code>
relative to the base <code>data.frame</code> is the capacity to hold
metadata on the columns in another <code>DataFrame</code>.</p>
<figure><img src="fig/dataframe-mcols.svg" alt="Metadata columns." class="figure mx-auto d-block"></figure><p><strong>Metadata columns.</strong> Metadata columns are illustrated
in the context of a <code>DataFrame</code> object. On the left, a
<code>DataFrame</code> object called <code>DF</code> is created with
columns named <code>A</code> and <code>B</code>. On the right, the
metadata columns for <code>DF</code> are accessed using
<code>mcols(DF)</code>. In this example, two metadata columns are
created with names <code>meta1</code> and <code>meta2</code>. Metadata
columns are stored as a <code>DataFrame</code> that contains one row for
each column in the parent <code>DataFrame</code>.</p>
<p>The metadata columns are accessed using the function
<code>mcols()</code>, If no metadata column is defined,
<code>mcols()</code> simply returns <code>NULL.</code></p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF4</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>    Integers <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1L</span>, <span class="fl">2L</span>, <span class="fl">3L</span><span class="op">)</span>,</span>
<span>    Letters <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span>,</span>
<span>    Floats <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1.2</span>, <span class="fl">2.3</span>, <span class="fl">3.4</span><span class="op">)</span>,</span>
<span>    row.names <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"alpha"</span>, <span class="st">"beta"</span>, <span class="st">"gamma"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">DF4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="cn">NULL</span></span></code></pre>
</div>
<p>The function <code>mcols()</code> can also be used to add, edit, or
remove metadata columns. For instance, we can initialise metadata
columns as a <code>DataFrame</code> of two columns:</p>
<ul>
<li>one column indicating the type of value stored in the corresponding
column</li>
<li>one column indicating the number of distinct values observed in the
corresponding column</li>
</ul>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">DF4</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>    Type <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">DF4</span>, <span class="va">typeof</span><span class="op">)</span>,</span>
<span>    Distinct <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">DF4</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span> <span class="fu">length</span><span class="op">(</span><span class="fu">unique</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">}</span> <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">DF4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3 rows and 2 columns
                Type  Distinct
         &lt;character&gt; &lt;integer&gt;
Integers     integer         3
Letters    character         3
Floats        double         3</code></pre>
</div>
<div id="note-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="note-1" class="callout-inner">
<h3 class="callout-title">Note</h3>
<div class="callout-content">
<p>The row names of the metadata columns are automatically set to match
the column names of the parent <code>DataFrame</code>, clearly
indicating the pairing between columns and metadata.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="run-length-encoding-rle">Run-length encoding (RLE)<a class="anchor" aria-label="anchor" href="#run-length-encoding-rle"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="an-extension-to-the-concept-of-vector">An extension to the concept of vector<a class="anchor" aria-label="anchor" href="#an-extension-to-the-concept-of-vector"></a>
</h3>
<p>Similarly to the <code>DataFrame</code> class implemented in the
<em><a href="https://bioconductor.org/packages/3.22/S4Vectors" class="external-link">S4Vectors</a></em>,
the <code>Rle</code> class provides an S4 extension to the
<code>rle()</code> function from the base package. Specifically, the
<code>Rle</code> class supports the storage of atomic vectors in a
run-length encoding format.</p>
<figure><img src="fig/rle.svg" alt="Run-length encoding." class="figure mx-auto d-block"></figure><p><strong>Run-length encoding.</strong> The concept of run-length
encoding is demonstrated here using the example of a sequence of nucleic
acids. Before encoding, each nucleotide at each position in the sequence
is explicitly stored in memory. During the encoding, consecutive runs of
identical nucleotides are collapsed into two bits of information: the
identity of the nucleotide and the length of the run.</p>
<p>Run-length encoding can dramatically reduce the memory footprint of
vectors that contain frequent runs of identical information. For
instance, a compelling application of run-length encoding is the
representation of genomic coverage in sequencing experiments, where
large genomic regions devoid of any mapped reads result in long runs of
<code>0</code> values. Storing each individual value would be highly
inefficient from the standpoint of memory usage. Instead, the run-length
encoding process collapses such runs of redundant information from
arbitrarily long runs of identical information to two values: the
repeated value itself, and the number of times that it is repeated.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v1</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">rle1</span> <span class="op">&lt;-</span> <span class="fu">Rle</span><span class="op">(</span><span class="va">v1</span><span class="op">)</span></span>
<span><span class="va">rle1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>numeric-Rle of length 17 with 7 runs
  Lengths: 7 1 1 1 1 1 5
  Values : 0 1 2 3 2 1 0</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="indexing-1">Indexing<a class="anchor" aria-label="anchor" href="#indexing-1"></a>
</h3>
<p>Just like a regular <code>vector</code>, <code>Rle</code> objects can
be indexed using <code>[</code>.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rle1</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>numeric-Rle of length 3 with 1 run
  Lengths: 3
  Values : 0</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<p>As vector-like objects, <code>Rle</code> objects can also be stored
as columns of <code>DataFrame</code> objects, alongside other
vector-like objects.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v2</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="fl">2</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rle2</span> <span class="op">&lt;-</span> <span class="fu">Rle</span><span class="op">(</span><span class="va">v2</span><span class="op">)</span></span>
<span><span class="va">DF5</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>    vector <span class="op">=</span> <span class="va">v2</span>,</span>
<span>    rle <span class="op">=</span> <span class="va">rle2</span>,</span>
<span>    equal <span class="op">=</span> <span class="va">v2</span> <span class="op">==</span> <span class="va">rle2</span></span>
<span><span class="op">)</span></span>
<span><span class="va">DF5</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 10 rows and 3 columns
      vector   rle equal
   &lt;numeric&gt; &lt;Rle&gt; &lt;Rle&gt;
1          1     1  TRUE
2          1     1  TRUE
3          1     1  TRUE
4          1     1  TRUE
5          1     1  TRUE
6          2     2  TRUE
7          2     2  TRUE
8          2     2  TRUE
9          2     2  TRUE
10         2     2  TRUE</code></pre>
</div>
<hr>
</div>
</section><section><h2 class="section-heading" id="summarizedexperiment-class">SummarizedExperiment class<a class="anchor" aria-label="anchor" href="#summarizedexperiment-class"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="motivation">Motivation<a class="anchor" aria-label="anchor" href="#motivation"></a>
</h3>
<p>Experiments are multifaceted data sets typically composed of at least
two key pieces of information necessary for any analysis:</p>
<ul>
<li>Assay data, typically a matrix representing measurements of a set of
features in a set of samples (e.g., RNA-sequencing).</li>
<li>Sample metadata, typically a <code>data.frame</code> of metadata
representing information about samples (e.g., treatment group).</li>
</ul>
<p>All those pieces of information must be kept synchronised – same
samples, same order – for downstream analyses to accurately process the
information and produce reliable results.</p>
<p>It is also very common for analytical workflows to analyse subsets of
samples or identify outliers that need to be removed to allow for more
accurate downstream analyses. In such cases, all aspects of the
experiments must be subsetted to the same set of samples – in the same
order – to preserve consistency in the data set and correct results.</p>
<p>The <code>SummarizedExperiment</code> – implemented in the <em><a href="https://bioconductor.org/packages/3.22/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
package – provides a container that accommodates those essential aspects
of individual experiments into a single object coordinates data and
metadata during subsetting and reordering operations. Its flexibility
accommodating many biological data types and comprehensive set of
features make it a popular data structure re-used throughout the
Bioconductor and a key part of the Bioconductor ecosystem. For instance,
familiarity with the <code>SummarizedExperiment</code> is a prerequisite
for working with the <em><a href="https://bioconductor.org/packages/3.22/DESeq2" class="external-link">DESeq2</a></em>
package for differential expression analysis, and the <em><a href="https://bioconductor.org/packages/3.22/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
extension class for single-cell analyses.</p>
</div>
<div class="section level3">
<h3 id="class-structure">Class structure<a class="anchor" aria-label="anchor" href="#class-structure"></a>
</h3>
<p><code>SummarizedExperiment</code> is a matrix-like container where
rows represent features of interest (e.g. genes, transcripts, exons,
etc.) and columns represent samples.</p>
<p>The objects can contain one or more assays, each represented by a
matrix-like object, as long as they be of the same dimensions.</p>
<p>Information about the features is stored in a <code>DataFrame</code>
object, nested within the <code>SummarizedExperiment</code> object, and
accessible using the function <code>rowData()</code>. Each row of the
<code>DataFrame</code> provides information on the feature in the
corresponding row of the SummarizedExperiment object. That information
may include annotations independent of the experiment (e.g., gene
identifier) as well as quality control metrics computed from assay data
during workflows.</p>
<p>Similarly, information about the samples is stored in another
<code>DataFrame</code> object, also nested within the
<code>SummarizedExperiment</code> object, and accessible using the
function <code>colData()</code>.</p>
<p>The following graphic displays the class geometry and highlights the
vertical (column) and horizontal (row) relationships. It was obtained
from the vignette of the <em><a href="https://bioconductor.org/packages/3.22/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
package.</p>
<figure><img src="fig/summarizedexperiment.svg" alt="Schematic representation of the SummarizedExperiment class." class="figure mx-auto d-block"></figure>
</div>
<div class="section level3">
<h3 id="creating-a-summarizedexperiment-object">Creating a SummarizedExperiment object<a class="anchor" aria-label="anchor" href="#creating-a-summarizedexperiment-object"></a>
</h3>
<p>Let us first load the package:</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span></code></pre>
</div>
<p>Then, let us import assay data from a file that we downloaded during
the lesson setup.</p>
<p>The file is a simple text file in which the first column contains
made-up feature identifiers and all other columns contain simulated data
for made-up samples. As such, we can use the base R function
<code>read.csv</code> to parse the file into a <code>data.frame</code>
object.</p>
<p>In the example below, we indicate that the row names can be found in
the first column, so that the function immediately sets the row names
accordingly in the output object. Hadn’t we specified it, the function
would have parsed it as a regular column and left the row names to the
default integer indexing.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_data</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/counts.csv"</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">count_data</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        sample_1 sample_2 sample_3 sample_4
gene_1       109       84       91      105
gene_2       111       97       98      108
gene_3        89      121      105       99
gene_4       105      109      122      101
gene_5        82       97      112       83
gene_6        89       96       90      116
gene_7       121       95       88      106
gene_8       101      101       86      103
gene_9        91      119       89       87
gene_10       81      111       81      118
gene_11       93      118       93       99
gene_12      103      111      116      103
gene_13       89      126      103      100
gene_14      101      107      111       79
gene_15       96       91      103      108
gene_16      110      102      128      103
gene_17       95      106      118      100
gene_18       99      115      114      102
gene_19      114      105       94      118
gene_20      110       88       99      102
gene_21      116       95       94      105
gene_22      114       96      107       91
gene_23       97      120       93       90
gene_24       91       84      118       97
gene_25       99      106       97      110</code></pre>
</div>
<p>One assay data matrix is enough to create a
<code>SummarizedExperiment</code> object, although without sample
metadata, only unsupervised analyses – that do not require information
about the samples – are possible.</p>
<p>In the example below, we create a <code>SummarizedExperiment</code>
object in which we store the matrix of count data under the name
‘counts’. Note that the argument ‘assays=’ (plural) can accept more than
one assay – as discussed above – which is why we encapsulate our only
assay matrix in a named <code>list</code> that also gives us the
opportunity to assign a name to the assay. Naming assays becomes crucial
during workflows that contain multiple assays, in order to identify and
retrieve individual assays unambiguously.</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span></span>
<span>  assays <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count_data</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SummarizedExperiment
dim: 25 4
metadata(0):
assays(1): counts
rownames(25): gene_1 gene_2 ... gene_24 gene_25
rowData names(0):
colnames(4): sample_1 sample_2 sample_3 sample_4
colData names(0):</code></pre>
</div>
<p>In the output above, the summary view of the object reminds us that
the assay – and thus the overall <code>SummarizedExperiment</code>
object – contains information for 25 features in 4 samples, it contains
a single assay named ‘counts’, the features seem to be named from
‘gene_1’ to ‘gene_25’ (only the first and last ones are shown), and the
samples are named from <code>sample_1</code> to <code>sample_4</code>.
The object does not contain any row metadata nor column metadata.</p>
<p>To create a more comprehensive <code>SummarizedExperiment</code>
object, let us import gene metadata and sample metadata for another two
files that we downloaded during the lesson setup.</p>
<p>The files are formatted similarly to the count data, so we use again
the base R function <code>read.csv()</code> to parse them into
<code>data.frame</code> objects.</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_metadata</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/sample_metadata.csv"</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">sample_metadata</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>         condition batch
sample_1         A     1
sample_2         A     2
sample_3         B     1
sample_4         B     2</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_metadata</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/gene_metadata.csv"</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">gene_metadata</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        chromosome
gene_1           4
gene_2           4
gene_3           5
gene_4           4
gene_5           5
gene_6           1
gene_7           2
gene_8           1
gene_9           3
gene_10          1
gene_11          1
gene_12          5
gene_13          5
gene_14          1
gene_15          3
gene_16          4
gene_17          2
gene_18          5
gene_19          1
gene_20          3
gene_21          5
gene_22          5
gene_23          1
gene_24          4
gene_25          5</code></pre>
</div>
<p>We can re-create the <code>SummarizedExperiment</code> object, this
time including the gene and sample metadata:</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span></span>
<span>  assays <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count_data</span><span class="op">)</span>,</span>
<span>  colData <span class="op">=</span> <span class="va">sample_metadata</span>,</span>
<span>  rowData <span class="op">=</span> <span class="va">gene_metadata</span></span>
<span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SummarizedExperiment
dim: 25 4
metadata(0):
assays(1): counts
rownames(25): gene_1 gene_2 ... gene_24 gene_25
rowData names(1): chromosome
colnames(4): sample_1 sample_2 sample_3 sample_4
colData names(2): condition batch</code></pre>
</div>
<p>Comparing the output above with the previous ‘assay-only’ version of
the <code>SummarizedExperiment</code> object, we can see that the
<code>rowData</code> and <code>colData</code> components now contain 1
and 4 metadata, respectively.</p>
</div>
<div class="section level3">
<h3 id="accessing-information">Accessing information<a class="anchor" aria-label="anchor" href="#accessing-information"></a>
</h3>
<p>A number of functions give access to the various components of
<code>SummarizedExperiment</code> objects.</p>
<p>The <code>assays()</code> function returns the list of assays stored
in the object. The output is always a <code>List</code>, event if the
object contains a single assay.</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assays</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>List of length 1
names(1): counts</code></pre>
</div>
<p>The <code>assayNames()</code> function returns a character vector of
the assay names. This is most useful when the object contains larger
numbers of assays, as the <code>assays()</code> function (see above) may
not display all of them. Knowing the names of the various assays is key
to accessing any individual assay.</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assayNames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "counts"</code></pre>
</div>
<p>The <code>assay()</code> function can be used to retrieve a single
assay from the object. For this, the function should be given the name
or the integer position of the desired assay. If unspecified, the
function automatically returns the first assay in the object.</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       sample_1 sample_2 sample_3 sample_4
gene_1      109       84       91      105
gene_2      111       97       98      108
gene_3       89      121      105       99
gene_4      105      109      122      101
gene_5       82       97      112       83
gene_6       89       96       90      116</code></pre>
</div>
<p>The <code>colData()</code> and <code>rowData()</code> functions can
be used to retrieve sample metadata and row metadata, respectively.</p>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 4 rows and 2 columns
           condition     batch
         &lt;character&gt; &lt;integer&gt;
sample_1           A         1
sample_2           A         2
sample_3           B         1
sample_4           B         2</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 25 rows and 1 column
        chromosome
         &lt;integer&gt;
gene_1           4
gene_2           4
gene_3           5
gene_4           4
gene_5           5
...            ...
gene_21          5
gene_22          5
gene_23          1
gene_24          4
gene_25          5</code></pre>
</div>
<p>Separately, the <code>$</code> operator can be used to access a
single column of sample metadata. A useful feature of this operator is
the autocompletion that is triggered automatically in RStudio or using
the tabulation key in terminal applications.</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span><span class="op">$</span><span class="va">batch</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 2 1 2</code></pre>
</div>
<p>Notably, there is no operator for accessing a single column of
feature metadata. For this, users need to first access the full
<code>DataFrame</code> returned by <code>rowData()</code> before
accessing a column using the standard <code>$</code> or <code>[[</code>
operators, e.g.</p>
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[[</span><span class="st">"chromosome"</span><span class="op">]</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 4 4 5 4 5 1 2 1 3 1 1 5 5 1 3 4 2 5 1 3 5 5 1 4 5</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="adding-and-editing-information">Adding and editing information<a class="anchor" aria-label="anchor" href="#adding-and-editing-information"></a>
</h3>
<p>Information can be added to <code>SummarizedExperiment</code> after
their creation. In fact, this is the basis for workflows that compute
normalised assay values – adding those to the list of assays –, and
quality control metrics for either features or samples – adding those to
the <code>rowData</code> and <code>colData</code> components, as
appropriate – progressively growing the amount of information stored
within the overall object.</p>
<p>Most of the functions for accessing information, described in the
previous section, have a counterpart function for adding new values or
editing existing ones. Note that editing is merely the result of adding
values under a name already in use, which has the effect of replacing
existing values.</p>
<p>In the example below, we add an assay named ‘logcounts’ which is the
result of applying a log-transformation to the ‘counts’ assay after
adding a pseucocount of one:</p>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"logcounts"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">log1p</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SummarizedExperiment
dim: 25 4
metadata(0):
assays(2): counts logcounts
rownames(25): gene_1 gene_2 ... gene_24 gene_25
rowData names(1): chromosome
colnames(4): sample_1 sample_2 sample_3 sample_4
colData names(2): condition batch</code></pre>
</div>
<p>In the output above, we see that the object now contains two assays:
the ‘counts’ assay included in the object when it was first created, and
the ‘logcounts’ assay added just now.</p>
<p>Similarly, the <code>colData()</code> and <code>rowData()</code>
functions – as well as the <code>$</code> operator – can be used to add
and edit values in the corresponding components.</p>
<p>In the example below, we compute the sum of counts for each sample,
and store the result in the sample metadata table under the new name
‘sum_counts’.</p>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[[</span><span class="st">"sum_counts"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">colSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 4 rows and 3 columns
           condition     batch sum_counts
         &lt;character&gt; &lt;integer&gt;  &lt;numeric&gt;
sample_1           A         1       2506
sample_2           A         2       2600
sample_3           B         1       2550
sample_4           B         2       2533</code></pre>
</div>
<p>In this next example, we compute the average count for each feature,
and store the result in the feature metadata table under the new name
‘mean_counts’.</p>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[[</span><span class="st">"mean_counts"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">rowSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 25 rows and 2 columns
        chromosome mean_counts
         &lt;integer&gt;   &lt;numeric&gt;
gene_1           4         389
gene_2           4         414
gene_3           5         414
gene_4           4         437
gene_5           5         374
...            ...         ...
gene_21          5         410
gene_22          5         408
gene_23          1         400
gene_24          4         390
gene_25          5         412</code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>S4 classes store information in slots, and check the validity of the
information every an object is updated.</li>
<li>To ensure the continued integrity of S4 objects, users should not
access slots directly, but using dedicated functions.</li>
<li>S4 generics invoke different implementations of the method depending
on the class of the object that they are given.</li>
<li>The S4 class <code>DataFrame</code> extends the functionality of
base <code>data.frame</code>, for instance with the capacity to hold
information about each column in metadata columns.</li>
<li>The S4 class <code>Rle</code> extends the functionality of the base
<code>vector</code>, for instance with the capacity to encode repetitive
vectors in a memory-efficient format.</li>
<li>The <code>SummarizedExperiment</code> class provides a single
container for storing both assay data and metadata.</li>
<li>Assay data and metadata are kept synchronised through subsetting and
reordering operations.</li>
<li>A comprehensive set of functions are available to access, add, and
edit information stored in the various components of the
<code>SummarizedExperiment</code> objects.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-06-application-microbiome"><p>Content from <a href="06-application-microbiome.html">Microbiome application</a></p>
<hr>
<p>Last updated on 2025-11-11 |

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/episodes/06-application-microbiome.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>In this example we will explore some bioconductor packages for the
analysis of microbiome data. We will focus on managing and storing data
in a structures and reproducible way and then fine tune some
visualizations. We will follow on with a similar set on analysis in the
next session for <a href="">communicating &amp; reproducing
analysis</a></p>
<p>During this we will be making use of two main packages from the
bioconductor library:</p>
<ul>
<li><a href="https://www.bioconductor.org/packages/release/bioc/html/microbiome.html" class="external-link"><em>Microbiome</em></a></li>
<li><a href="https://www.bioconductor.org/packages/release/bioc/html/phyloseq.html" class="external-link"><em>Phyloseq</em></a></li>
</ul>
<p>We will also introduce a new data class - <code>phyloseq-class</code>
from the package <a href="https://joey711.github.io/phyloseq/" class="external-link"><em>Phyloseq</em></a> that
uses the S4 class system.</p>
<blockquote>
<p>The <a href="https://www.bioconductor.org/packages/release/bioc/html/phyloseq.html" class="external-link"><em>Phyloseq</em></a>
package is designed to store and manage phylogenetic sequencing data,
including abundance data, phylogenetic trees, taxonomic assignments, and
sample metadata, within a single experiment-level S4 object. This S4
object-oriented approach provides a structured and organized way to
handle complex microbiome data, facilitating reproducible analyses and
data sharing.</p>
</blockquote>
<p>At its core is the <code>phyloseq-class</code> object, which
integrates multiple data types: OTU/ASV tables, taxonomic annotations,
sample metadata, and phylogenetic trees. While originally developed for
16S rRNA microbiome studies, phyloseq-class objects are also widely used
in:</p>
<ul>
<li>eDNA metabarcoding studies (e.g., biodiversity monitoring in aquatic
or terrestrial environments)</li>
<li>Fungal ITS sequencing projects</li>
<li>Metagenomic and metatranscriptomic datasets with annotated taxonomic
profiles</li>
<li>Environmental genomics and ecological community profiling</li>
</ul>
<section><h2 class="section-heading" id="install-packages">Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a>
</h2>
<hr class="half-width">
<p>Install our <a href="https://www.bioconductor.org/packages/release/bioc/html/microbiome.html" class="external-link"><em>Microbiome</em></a>
and <a href="https://www.bioconductor.org/packages/release/bioc/html/phyloseq.html" class="external-link"><em>Phyloseq</em></a>
packages using <code>BiocManager::install()</code></p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw">require</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The following initializes usage of Bioc devel</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span>version<span class="op">=</span><span class="st">'devel'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"phyloseq"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"microbiome"</span><span class="op">)</span></span></code></pre>
</div>
<p>Now load in the packages we need for this analysis</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">microbiome</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">phyloseq</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">knitr</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="example-datasets">Example datasets<a class="anchor" aria-label="anchor" href="#example-datasets"></a>
</h2>
<hr class="half-width">
<p>We will explore some example datasets that come with the packages <a href="https://microbiome.github.io/tutorials/Data.html" class="external-link">Microbiome*</a>
and <a href="https://joey711.github.io/phyloseq/Example-Data" class="external-link"><em>Phyloseq</em></a>.
We will take some time to explore what information they contain, how we
can view the datasets and how they differ.</p>
<p><strong>Explore dataset 1 - <code>dietswap</code></strong></p>
<p>The first dataset we will load is from <a href="https://dx.doi.org/10.1038/ncomms7342" class="external-link">O’Keefe S et al. Nature
Communications, 2015</a> made available by <a href="https://microbiome.github.io/tutorials/Data.html" class="external-link"><em>Microbiome</em></a>
package. This study contains microbiome data from a study with African
and African American groups undergoing a two-week diet swap. This data
set is based on the Human Intestinal Tract (HIT)Chip phylogenetic 16S
microarray (<a href="https://doi.org/10.1111/j.1462-2920.2009.01900.x" class="external-link">Rajilic-Stojanovic
et al. 2009</a>). This profiling technology differs from the more widely
used 16S rRNA amplicon sequencing. Column metadata includes the subject
identifier, sex, nationality, group information, sample identifier, time
point information, time point information within group and BMI group.
Row metadata contains taxonomic information on the Phylum, Family and
Genus level.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">data</span><span class="op">(</span><span class="va">dietswap</span><span class="op">)</span></span></code></pre>
</div>
<p><strong>Explore dataset 2 - <code>peerj32</code></strong></p>
<p>The second dataset we will load is the <code>peerj32</code> data from
<a href="https://doi.org/10.7717/peerj.32" class="external-link">Lahti et al. PeerJ 1:e32,
2013</a> made available by <a href="https://microbiome.github.io/tutorials/Data.html" class="external-link"><em>Microbiome</em></a>
package. This characterizes associations between human intestinal
microbiota and blood serum lipids. Note that this data set contains an
additional data matrix of lipid species.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">data</span><span class="op">(</span><span class="va">peerj32</span><span class="op">)</span></span></code></pre>
</div>
<p><strong>Explore dataset 3 - <code>GlobalPatterns</code></strong></p>
<p>The third dataset we will load is the <code>GlobalPatterns</code>
data from <a href="https://doi.org/10.1073/pnas.1000080107" class="external-link">Caporaso et
al. PNAS, 108, 4516-4522, 2011</a> made available by <a href="https://joey711.github.io/phyloseq/Example-Data" class="external-link"><em>Phyloseq</em></a>.
This work compared the microbial communities from 25 environmental
samples and three known “mock communities” – a total of 9 sample types –
at a depth averaging 3.1 million reads per sample. Authors were able to
reproduce diversity patterns seen in many other published studies, while
also investigating technical issues/bias by applying the same techniques
to simulated microbial communities of known composition.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">data</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span></span></code></pre>
</div>
<div id="exploring-data" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exploring-data" class="callout-inner">
<h3 class="callout-title">Exploring data</h3>
<div class="callout-content">
<p>Explore the datasets <code>dietswap</code>, <code>peerj32</code> and
<code>GlobalPatterns</code> that you just loaded. You can do this by
navigating to your “Environment” pane, or in the console try the
following <code>dietswap@</code> and see what options are available to
you. Can you view the metadata for this data set? Why does
<code>peerj32@</code> not work in the same way?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>To see the full data within the <code>peerj32</code> object try
typing <code>peerj32$</code> into the console.</p>
</div>
</div>
</div>
</div>
<div id="missing-information" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="missing-information" class="callout-inner">
<h3 class="callout-title">Missing information</h3>
<div class="callout-content">
<p>Can you identify what information is missing in each of the three
<code>phyloseq-class</code> objects for the datasets above?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>Check the contents of each of the three datasets separately and you
can see that both <code>dietswap</code> and
<code>peerj32$phyloseq</code> objects only contain the minimum data
required - <code>otu_table()</code>, <code>sample_data()</code> and
<code>tax_table()</code>. The <code>GlobalPatterns</code> data also
contained additional taxonomic ranks and <code>phy_tree()</code> which
is a phylogenetic tree that we can use when performing analysis.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="sc">&gt;</span> dietswap</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>phyloseq<span class="sc">-</span>class experiment<span class="sc">-</span>level object</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">otu_table</span>()   OTU Table<span class="sc">:</span>         [ <span class="dv">130</span> taxa and <span class="dv">222</span> samples ]</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="fu">sample_data</span>() Sample Data<span class="sc">:</span>       [ <span class="dv">222</span> samples by <span class="dv">8</span> sample variables ]</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="fu">tax_table</span>()   Taxonomy Table<span class="sc">:</span>    [ <span class="dv">130</span> taxa by <span class="dv">3</span> taxonomic ranks ]</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="sc">&gt;</span> peerj32<span class="sc">$</span>phyloseq</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>phyloseq<span class="sc">-</span>class experiment<span class="sc">-</span>level object</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="fu">otu_table</span>()   OTU Table<span class="sc">:</span>         [ <span class="dv">130</span> taxa and <span class="dv">44</span> samples ]</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="fu">sample_data</span>() Sample Data<span class="sc">:</span>       [ <span class="dv">44</span> samples by <span class="dv">5</span> sample variables ]</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="fu">tax_table</span>()   Taxonomy Table<span class="sc">:</span>    [ <span class="dv">130</span> taxa by <span class="dv">3</span> taxonomic ranks ]</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="sc">&gt;</span> GlobalPatterns</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>phyloseq<span class="sc">-</span>class experiment<span class="sc">-</span>level object</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="fu">otu_table</span>()   OTU Table<span class="sc">:</span>         [ <span class="dv">19216</span> taxa and <span class="dv">26</span> samples ]</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="fu">sample_data</span>() Sample Data<span class="sc">:</span>       [ <span class="dv">26</span> samples by <span class="dv">7</span> sample variables ]</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="fu">tax_table</span>()   Taxonomy Table<span class="sc">:</span>    [ <span class="dv">19216</span> taxa by <span class="dv">7</span> taxonomic ranks ]</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="fu">phy_tree</span>()    Phylogenetic Tree<span class="sc">:</span> [ <span class="dv">19216</span> tips and <span class="dv">19215</span> internal nodes ]</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="visualising-data">Visualising data<a class="anchor" aria-label="anchor" href="#visualising-data"></a>
</h2>
<hr class="half-width">
<p>Now we will have a go at working through some data visualizations and
improving asesthetics ready for publications or sharing with
colleagues.</p>
<p>For this exercise we will work through using the
<code>GlobalPatterns</code> data.</p>
<blockquote>
<p>The <code>GlobalPatterns</code> data is from <a href="https://doi.org/10.1073/pnas.1000080107" class="external-link">Caporaso et al. PNAS,
108, 4516-4522, 2011</a> made available by <a href="https://joey711.github.io/phyloseq/Example-Data" class="external-link"><em>Phyloseq</em></a>.
This work compared the microbial communities from 25 environmental
samples and three known “mock communities”</p>
</blockquote>
<p>First a re-fresh of what the sample data contains for this
dataset</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GlobalPatterns</span><span class="op">@</span><span class="va">sam_data</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Sample Data:        [26 samples by 7 sample variables]:
         X.SampleID  Primer Final_Barcode Barcode_truncated_plus_T
CL3             CL3 ILBC_01        AACGCA                   TGCGTT
CC1             CC1 ILBC_02        AACTCG                   CGAGTT
SV1             SV1 ILBC_03        AACTGT                   ACAGTT
M31Fcsw     M31Fcsw ILBC_04        AAGAGA                   TCTCTT
M11Fcsw     M11Fcsw ILBC_05        AAGCTG                   CAGCTT
M31Plmr     M31Plmr ILBC_07        AATCGT                   ACGATT
M11Plmr     M11Plmr ILBC_08        ACACAC                   GTGTGT
F21Plmr     F21Plmr ILBC_09        ACACAT                   ATGTGT
M31Tong     M31Tong ILBC_10        ACACGA                   TCGTGT
M11Tong     M11Tong ILBC_11        ACACGG                   CCGTGT
LMEpi24M   LMEpi24M ILBC_13        ACACTG                   CAGTGT
SLEpi20M   SLEpi20M ILBC_15        ACAGAG                   CTCTGT
AQC1cm       AQC1cm ILBC_16        ACAGCA                   TGCTGT
AQC4cm       AQC4cm ILBC_17        ACAGCT                   AGCTGT
AQC7cm       AQC7cm ILBC_18        ACAGTG                   CACTGT
NP2             NP2 ILBC_19        ACAGTT                   AACTGT
NP3             NP3 ILBC_20        ACATCA                   TGATGT
NP5             NP5 ILBC_21        ACATGA                   TCATGT
TRRsed1     TRRsed1 ILBC_22        ACATGT                   ACATGT
TRRsed2     TRRsed2 ILBC_23        ACATTC                   GAATGT
TRRsed3     TRRsed3 ILBC_24        ACCACA                   TGTGGT
TS28           TS28 ILBC_25        ACCAGA                   TCTGGT
TS29           TS29 ILBC_26        ACCAGC                   GCTGGT
Even1         Even1 ILBC_27        ACCGCA                   TGCGGT
Even2         Even2 ILBC_28        ACCTCG                   CGAGGT
Even3         Even3 ILBC_29        ACCTGT                   ACAGGT
         Barcode_full_length         SampleType
CL3              CTAGCGTGCGT               Soil
CC1              CATCGACGAGT               Soil
SV1              GTACGCACAGT               Soil
M31Fcsw          TCGACATCTCT              Feces
M11Fcsw          CGACTGCAGCT              Feces
M31Plmr          CGAGTCACGAT               Skin
M11Plmr          GCCATAGTGTG               Skin
F21Plmr          GTAGACATGTG               Skin
M31Tong          TGTGGCTCGTG             Tongue
M11Tong          TAGACACCGTG             Tongue
LMEpi24M         CATGAACAGTG         Freshwater
SLEpi20M         AGCCGACTCTG         Freshwater
AQC1cm           GACCACTGCTG Freshwater (creek)
AQC4cm           CAAGCTAGCTG Freshwater (creek)
AQC7cm           ATGAAGCACTG Freshwater (creek)
NP2              TCGCGCAACTG              Ocean
NP3              GCTAAGTGATG              Ocean
NP5              GAACGATCATG              Ocean
TRRsed1          CACGTGACATG Sediment (estuary)
TRRsed2          TGCGCTGAATG Sediment (estuary)
TRRsed3          GATGTATGTGG Sediment (estuary)
TS28             GCATCGTCTGG              Feces
TS29             CTAGTCGCTGG              Feces
Even1            TGACTCTGCGG               Mock
Even2            TCTGATCGAGG               Mock
Even3            AGAGAGACAGG               Mock
                                          Description
CL3          Calhoun South Carolina Pine soil, pH 4.9
CC1          Cedar Creek Minnesota, grassland, pH 6.1
SV1        Sevilleta new Mexico, desert scrub, pH 8.3
M31Fcsw       M3, Day 1, fecal swab, whole body study
M11Fcsw      M1, Day 1, fecal swab, whole body study
M31Plmr       M3, Day 1, right palm, whole body study
M11Plmr      M1, Day 1, right palm, whole body study
F21Plmr     F1, Day 1,  right palm, whole body study
M31Tong          M3, Day 1, tongue, whole body study
M11Tong          M1, Day 1, tongue, whole body study
LMEpi24M Lake Mendota Minnesota, 24 meter epilimnion
SLEpi20M Sparkling Lake Wisconsin, 20 meter eplimnion
AQC1cm                   Allequash Creek, 0-1cm depth
AQC4cm                  Allequash Creek, 3-4 cm depth
AQC7cm                  Allequash Creek, 6-7 cm depth
NP2            Newport Pier, CA surface water, Time 1
NP3            Newport Pier, CA surface water, Time 2
NP5            Newport Pier, CA surface water, Time 3
TRRsed1                Tijuana River Reserve, depth 1
TRRsed2                Tijuana River Reserve, depth 2
TRRsed3                Tijuana River Reserve, depth 2
TS28                                          Twin #1
TS29                                          Twin #2
Even1                                           Even1
Even2                                           Even2
Even3                                           Even3</code></pre>
</div>
<p>Next we often need to do some data pruning or subsetting. In this
case we are making sure we remove taxa that doesn’t have any sequences
present.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GP</span> <span class="op">&lt;-</span> <span class="fu">prune_species</span><span class="op">(</span><span class="fu">speciesSums</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">GlobalPatterns</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in prune_species(speciesSums(GlobalPatterns) &gt; 0, GlobalPatterns): 'prune_species' is deprecated.
Use 'prune_taxa' instead.
See help("Deprecated") and help("phyloseq-deprecated").</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in speciesSums(GlobalPatterns): 'speciesSums' is deprecated.
Use 'taxa_sums' instead.
See help("Deprecated") and help("phyloseq-deprecated").</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compare the two datasets</span></span>
<span><span class="va">GlobalPatterns</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GP</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 18988 taxa and 26 samples ]
sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
tax_table()   Taxonomy Table:    [ 18988 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 18988 tips and 18987 internal nodes ]</code></pre>
</div>
<div class="section level3">
<h3 id="alpha-diversity-plots">Alpha diversity plots<a class="anchor" aria-label="anchor" href="#alpha-diversity-plots"></a>
</h3>
<p>For microbiome (or community analysis) a common strategy is to
visualize the alpha diversity of samples. Essentially this is how many
unique taxa were identified in each sample.</p>
<p>At the simplest level we can plot all the alpha diversity measures
from the function <code>plot_richness()</code> from the <a href="https://www.bioconductor.org/packages/release/bioc/html/phyloseq.html" class="external-link"><em>Phyloseq</em></a>
package.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_richness</span><span class="op">(</span><span class="va">GP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: `aes_string()` was deprecated in ggplot2 3.0.0.
ℹ Please use tidy evaluation idioms with `aes()`.
ℹ See also `vignette("ggplot2-in-packages")` for more information.
ℹ The deprecated feature was likely used in the phyloseq package.
  Please report the issue at &lt;https://github.com/joey711/phyloseq/issues&gt;.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
generated.</code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-8-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Note that in this case, the Fisher calculation results in a warning
(but still plots).</p>
<p>There is a lot going on here so let’s try just look at two different
alpha diversity measures.</p>
<p>We can avoid this by specifying a measures argument to plot_richness,
which will include just the alpha-diversity measures that we want.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_richness</span><span class="op">(</span><span class="va">GP</span>, measures<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"Chao1"</span>, <span class="st">"Shannon"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-9-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="getting-help-on-functions" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="getting-help-on-functions" class="callout-inner">
<h3 class="callout-title">Getting help on functions</h3>
<div class="callout-content">
<p>You can find more information on the funciton by typing
<code>?plot_richness</code> into the console. Have a look and see what
measures are available using this function.</p>
</div>
</div>
</div>
<p>We can specify a sample variable on which to group/organize samples
along the horizontal (<code>x</code>) axis. An experimentally meaningful
categorical variable is usually a good choice – in this case, the
“<code>SampleType</code>” variable works much better than attempting to
interpret the sample names directly (as in the previous plot):</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_richness</span><span class="op">(</span><span class="va">GP</span>, x<span class="op">=</span><span class="st">"SampleType"</span>, measures<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"Chao1"</span>, <span class="st">"Shannon"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-10-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Now suppose we wanted to use an external variable in the plot that
isn’t in the <code>GP</code> dataset already – for example, a logical
that indicated whether or not the samples are human-associated. First,
define this new variable, <code>human</code>, as a factor (other vectors
could also work; or other data you might have describing the
samples).</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sampleData</span><span class="op">(</span><span class="va">GP</span><span class="op">)</span><span class="op">$</span><span class="va">human</span> <span class="op">&lt;-</span> <span class="fu">getVariable</span><span class="op">(</span><span class="va">GP</span>, <span class="st">"SampleType"</span><span class="op">)</span> <span class="op">%in%</span> </span>
<span>  <span class="fu">c</span><span class="op">(</span><span class="st">"Feces"</span>, <span class="st">"Mock"</span>, <span class="st">"Skin"</span>, <span class="st">"Tongue"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in getVariable(GP, "SampleType"): 'getVariable' is deprecated.
Use 'get_variable' instead.
See help("Deprecated") and help("phyloseq-deprecated").</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in sampleData(`*tmp*`): 'sampleData' is deprecated.
Use 'sample_data' instead.
See help("Deprecated") and help("phyloseq-deprecated").</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in `sampleData&lt;-`(`*tmp*`, value = new("sample_data", .Data = list(: 'sampleData&lt;-' is deprecated.
Use 'sample_data&lt;-' instead.
See help("Deprecated") and help("phyloseq-deprecated").</code></pre>
</div>
<p>Now tell <code>plot_richness</code> to map the new human variable on
the horizontal axis, and shade the points in different color groups,
according to which “<code>SampleType</code>” they belong.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_richness</span><span class="op">(</span></span>
<span>  <span class="va">GP</span>,</span>
<span>  x <span class="op">=</span> <span class="st">"human"</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"SampleType"</span>,</span>
<span>  measures <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Chao1"</span>, <span class="st">"Shannon"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-12-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can merge samples that are from the environment
(<code>SampleType</code>), and make the points bigger with a ggplot2
layer. First, merge the samples.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GPst</span> <span class="op">&lt;-</span> <span class="fu">merge_samples</span><span class="op">(</span><span class="va">GP</span>, <span class="st">"SampleType"</span><span class="op">)</span></span>
<span><span class="co"># repair variables that were damaged during merge (coerced to numeric)</span></span>
<span><span class="fu">sample_data</span><span class="op">(</span><span class="va">GPst</span><span class="op">)</span><span class="op">$</span><span class="va">SampleType</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="fu">sample_names</span><span class="op">(</span><span class="va">GPst</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">sample_data</span><span class="op">(</span><span class="va">GPst</span><span class="op">)</span><span class="op">$</span><span class="va">human</span> <span class="op">&lt;-</span> <span class="fu">as.logical</span><span class="op">(</span><span class="fu">sample_data</span><span class="op">(</span><span class="va">GPst</span><span class="op">)</span><span class="op">$</span><span class="va">human</span><span class="op">)</span></span></code></pre>
</div>
<p>Now we can plot this environment-merged version of the data. First
store the default ggplot graphic as <code>p</code>, then add an
additional <code>geom_point</code> layer with a large size and slight
transparency.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">plot_richness</span><span class="op">(</span></span>
<span>  <span class="va">GPst</span>,</span>
<span>  x <span class="op">=</span> <span class="st">"human"</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"SampleType"</span>,</span>
<span>  measures <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Chao1"</span>, <span class="st">"Shannon"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-14-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p><strong>More details about ggplot2</strong></p>
<p>For those interested in why this works so concisely
<code>(p + geom_point(size=4, alpha=0.7))</code>, it is because the rest
of the aesthetic mapping and data are contained in the ggplot object,
<code>p</code>, and so is inherited in the call to the ggplot2 geometric
object layer function, <code>geom_point</code>, by default since we
didn’t specify alternative <code>aes</code> or <code>data</code>
arguments. Although we could have if we wanted to. This perhaps sounds
more confusing than it is, and I find it easier to understand by
inspecting the examples I’ve shown here.</p>
<p>You’ll also notice that the original smaller points are still on the
plot. This is because they were the first layer, and our larger points
are semi-transparent. I find this kind of distracting, and doesn’t add
any information or clarity. The good news is that layers can be removed
from a ggplot object with standard list notation (using the dollar sign
<code>$</code>).</p>
<p>First, check which lists are present in <code>p</code>.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">layers</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$geom_point
geom_point: na.rm = TRUE
stat_identity: na.rm = TRUE
position_identity

$geom_errorbar
mapping: ymax = ~value + se, ymin = ~value - se
geom_errorbar: na.rm = FALSE, orientation = NA, lineend = butt, width = 0.1
stat_identity: na.rm = FALSE
position_identity </code></pre>
</div>
<p>We can see that the first layer is the one specifying the original
points, which are small. We can use negative indexing to “pop” it out,
then add a new <code>geom_point</code> layer with larger point size (the
following two lines).</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">layers</span> <span class="op">&lt;-</span> <span class="va">p</span><span class="op">$</span><span class="va">layers</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span>size<span class="op">=</span><span class="fl">5</span>, alpha<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-application-microbiome-rendered-unnamed-chunk-16-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
<div class="section level3">
<h3 id="phylogenetic-trees">Phylogenetic trees<a class="anchor" aria-label="anchor" href="#phylogenetic-trees"></a>
</h3>
<p><a href="https://joey711.github.io/phyloseq/plot_tree-examples.html" class="external-link"></a></p>
</div>
<div class="section level3">
<h3 id="ordination-plots">Ordination plots<a class="anchor" aria-label="anchor" href="#ordination-plots"></a>
</h3>
<p><a href="https://joey711.github.io/phyloseq/plot_ordination-examples.html" class="external-link"></a></p>
<hr>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Bespoke often domain specific data classes are often built to serve
a specific set of data, however by utilising fundamental R principals
such as S4 class structures transferring formats between packages for
analysis is much easier.</li>
<li>Look for widely accepted data classes in your field before
formatting your data for a specific package early on.</li>
<li>Packages that generate <a href="https://ggplot2.tidyverse.org/" class="external-link"><em>ggplot2</em></a> compatible
figures allow for <em>easier</em> manipulations and customise using
familar syntax</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-07-reproducible-analysis"><p>Content from <a href="07-reproducible-analysis.html">Communicating &amp; Reproducing Analysis</a></p>
<hr>
<p>Last updated on 2025-11-11 |

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/episodes/07-reproducible-analysis.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>In this exercise we will cover some aspects relevant to communicating
and most importantly <strong>re</strong>producing analysis.</p>
<p>We will use the libraries and data we covered in the previous episode
<a href="https://siobhonlegan.com/2025-11-13-COMBINE-WA/10-application-microbiome.html" class="external-link">Application
for microbiome data</a>.</p>
<section><h2 class="section-heading" id="why-the-emphasis-on-reproducible-analysis">Why the emphasis on reproducible analysis<a class="anchor" aria-label="anchor" href="#why-the-emphasis-on-reproducible-analysis"></a>
</h2>
<hr class="half-width">
<p><a href="https://r4ds.had.co.nz/introduction.html" class="external-link"><img src="images/clipboard-464455724.png" alt="Data lifecycle" class="figure"></a></p>
<p>Research is rarely a straight line. Imagine this common scenario:</p>
<ul>
<li>A research student runs a preliminary analysis and shares results
with their supervisor for feedback.</li>
<li>The supervisor suggests changes—perhaps a different filtering
threshold, an alternative normalization method, or adding new
samples.</li>
<li>The student now needs to reanalyse the data, often weeks later.
Without a clear record of the original steps, this can mean starting
from scratch, risking inconsistencies and wasted time.</li>
</ul>
<p>The same challenge applies to research staff working in teams. When
multiple people collaborate on a project, unclear workflows can lead
to:</p>
<ul>
<li>Confusion about which version of the analysis is current.</li>
<li>Difficulty reproducing figures or tables for publications.</li>
<li>Problems when revisiting the project months or years later.</li>
</ul>
<p><strong>Reproducibility = Transparency + Efficiency</strong></p>
<p>Reproducible analysis ensures that:</p>
<ul>
<li>Every step of your workflow is documented.</li>
<li>Analyses can be rerun easily after updates or corrections.</li>
<li>Collaborators (and future you!) can understand and verify your
work.</li>
</ul>
<p><strong>Our Approach in This Workshop</strong></p>
<p>For this exercise, we will use R Markdown and Quarto
notebooks—powerful tools for reproducible research. These allow you
to:</p>
<ul>
<li>Combine code, results, and narrative in one document.</li>
<li>Automatically regenerate outputs when data or code changes.</li>
<li>Share complete, executable workflows with colleagues or
supervisors.</li>
</ul></section><section><h2 class="section-heading" id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h2>
<hr class="half-width">
<p>Now load in the required packages for this analysis</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">microbiome</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">knitr</span><span class="op">)</span></span>
<span><span class="co"># load the peerj32 dataset</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="va">peerj32</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="reproducible-data">Reproducible data<a class="anchor" aria-label="anchor" href="#reproducible-data"></a>
</h2>
<hr class="half-width">
<p>Let’s say you have spent a long time getting your data formatted and
read into R.</p>
<p>Now you will want to do some visualizations on this data and it is
likely this will need to be repeated (many) times in the future.</p>
<p>You can save your data in and <code>.rda</code> or <code>.rds</code>
format so that it will allow you to quickly read it back into the work
space and pick up on more figures. It will also made it easy to read in
and generate reports in a reproducible way (which we will do
shortly).</p>
<p>Have a go at saving the <code>peerj32</code> object that is in your
environment to your local directory. If you have followed the previous
episodes in this lesson you should already have a <strong>data/</strong>
directory.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">save</span><span class="op">(</span><span class="va">peerj32</span>, file <span class="op">=</span> <span class="st">"data/peerj32.rda"</span><span class="op">)</span></span></code></pre>
</div>
<p>You can quickly load the <code>.rda</code> file back in</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">load</span><span class="op">(</span><span class="st">"data/peerj32.rda"</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="making-reproducible-and-easy-to-share-reports">Making reproducible and easy to share reports<a class="anchor" aria-label="anchor" href="#making-reproducible-and-easy-to-share-reports"></a>
</h2>
<hr class="half-width">
<p>Next you will download a have prepared a <code>.qmd</code> file and
then work through the code chunks and “knit” the document to a
reproducible a <code>.html</code> file for communication.</p>
<div id="the-situation---sharing-results" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="the-situation---sharing-results" class="callout-inner">
<h3 class="callout-title">The Situation - Sharing Results</h3>
<div class="callout-content">
<p>Imagine you need to give your supervisors a summary of your findings
early on, instead of spending weeks neatly formatting and exporting
plots one by one in R you may like to share this html file early in your
analysis to get some general feedback and direction before find tuning
your analysis.</p>
</div>
</div>
</div>
<p>It is best practice to not re-install the packages each time when
making these reports, so ensure you have the required packages installed
prior to running.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install your CRAN packages for the qmd</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"gtsummary"</span><span class="op">)</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"reshape2"</span><span class="op">)</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"knitr"</span><span class="op">)</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"DT"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## install your Bioconductor packages</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw">require</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"microbiome"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"phyloseq"</span><span class="op">)</span></span></code></pre>
</div>
<p>Download the <code>.qmd</code> file</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span></span>
<span>  url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/siobhon-egan/2025-11-13-COMBINE-WA/refs/heads/main/episodes/files/microbiome-data-report.qmd"</span>,</span>
<span>  destfile <span class="op">=</span> <span class="st">"data/microbiome-report.qmd"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>Have a go at rendering the <code>.qmd</code> file.</p>
<div id="challenge" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Practical opening and “knitting” the <code>.qmd</code> file to html
output.</p>
</div>
</div>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: `includeHTML()` was provided a `path` that appears to be a complete HTML document.
✖ Path: files/microbiome-data-report.html
ℹ Use `tags$iframe()` to include an HTML document. You can either ensure `path` is accessible in your app or document (see e.g. `shiny::addResourcePath()`) and pass the relative path to the `src` argument. Or you can read the contents of `path` and pass the contents to `srcdoc`.</code></pre>
</div>
<!--html_preserve-->
&lt;!DOCTYPE html&gt;


<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.23">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Siobhon Egan">
<meta name="dcterms.date" content="2025-11-13">
<title>
Microbiome data report for Supervisor meeting
</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>
<script src="microbiome-data-report_files/libs/clipboard/clipboard.min.js"></script><script src="microbiome-data-report_files/libs/quarto-html/quarto.js" type="module"></script><script src="microbiome-data-report_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script><script src="microbiome-data-report_files/libs/quarto-html/popper.min.js"></script><script src="microbiome-data-report_files/libs/quarto-html/tippy.umd.min.js"></script><script src="microbiome-data-report_files/libs/quarto-html/anchor.min.js"></script><link href="microbiome-data-report_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="microbiome-data-report_files/libs/quarto-html/quarto-syntax-highlighting-226bd0f977fa82dfae4534cac220d79a.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="microbiome-data-report_files/libs/bootstrap/bootstrap.min.js"></script><p><link href="microbiome-data-report_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet"><link href="microbiome-data-report_files/libs/bootstrap/bootstrap-be9206a58a27969b3ea3001da8f26fef.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light"></p>


<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
Microbiome data report for Supervisor meeting
</h1>
</div>
<div class="quarto-title-meta">
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">SH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode sh" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="op">&lt;</span>div<span class="op">&gt;</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="op">&lt;</span>div <span class="va">class</span><span class="op">=</span><span class="st">"quarto-title-meta-heading"</span><span class="op">&gt;</span>Author<span class="op">&lt;</span>/div<span class="op">&gt;</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="op">&lt;</span>div <span class="va">class</span><span class="op">=</span><span class="st">"quarto-title-meta-contents"</span><span class="op">&gt;</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>         <span class="op">&lt;</span>p<span class="op">&gt;</span>Siobhon <span class="ex">Egan</span> <span class="op">&lt;</span>/p<span class="op">&gt;</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>      <span class="op">&lt;</span>/div<span class="op">&gt;</span></span></code></pre>
</div>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">SH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode sh" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="op">&lt;</span>div<span class="op">&gt;</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="op">&lt;</span>div <span class="va">class</span><span class="op">=</span><span class="st">"quarto-title-meta-heading"</span><span class="op">&gt;</span>Published<span class="op">&lt;</span>/div<span class="op">&gt;</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="op">&lt;</span>div <span class="va">class</span><span class="op">=</span><span class="st">"quarto-title-meta-contents"</span><span class="op">&gt;</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>  <span class="op">&lt;</span>p <span class="va">class</span><span class="op">=</span><span class="st">"date"</span><span class="op">&gt;</span>November <span class="ex">13,</span> <span class="dv">2025</span><span class="op">&lt;</span>/p<span class="op">&gt;</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="op">&lt;</span>/div<span class="op">&gt;</span></span></code></pre>
</div>
</header></main>
</div>
</section><p>
The following datasets and analysis have been used from tutorials
available in the <a href="">microbiome package</a>, for
<a href="https://microbiome.github.io/tutorials/Heatmap.html" class="external-link">heatmaps</a>
</p>
<section class="level2"><h2 class="anchored" data-anchor-id="information" id="information">
Information
<a class="anchor" aria-label="anchor" href="#information"></a>
</h2>
<p>
Prepare a summary table of the participants in your dataset
</p>
<div class="cell">
<details class="code-fold"><summary>
Show the code
</summary><div id="cb1" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span><span class="op">@</span><span class="va">sam_data</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">group</span>, <span class="va">sex</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 × 4
# Groups:   group, sex [4]
  group   sex     time     n
  &lt;fct&gt;   &lt;fct&gt;  &lt;int&gt; &lt;int&gt;
1 LGG     female     1     5
2 LGG     female     2     5
3 LGG     male       1     3
4 LGG     male       2     3
5 Placebo female     1    10
6 Placebo female     2    10
7 Placebo male       1     4
8 Placebo male       2     4</code></pre>
</div>
<details class="code-fold"><summary>
Show the code
</summary><div id="cb3" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span><span class="op">@</span><span class="va">sam_data</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">group</span>, <span class="va">sex</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 3
# Groups:   group [2]
  group   sex        n
  &lt;fct&gt;   &lt;fct&gt;  &lt;int&gt;
1 LGG     female    10
2 LGG     male       6
3 Placebo female    20
4 Placebo male       8</code></pre>
</div>
</div>
</section><section class="level2"><h2 class="anchored" data-anchor-id="alpha-diversity" id="alpha-diversity">
Alpha diversity
<a class="anchor" aria-label="anchor" href="#alpha-diversity"></a>
</h2>
</section><section class="level2"><h2 class="anchored" data-anchor-id="beta-diversity" id="beta-diversity">
Beta diversity
<a class="anchor" aria-label="anchor" href="#beta-diversity"></a>
</h2>
</section><section class="level2"><h2 class="anchored" data-anchor-id="heatmap-of-abundance" id="heatmap-of-abundance">
Heatmap of abundance
<a class="anchor" aria-label="anchor" href="#heatmap-of-abundance"></a>
</h2>
<p>
Abundance of microbial taxa (using the Z transformed data)
</p>
<div class="cell">
<details class="code-fold"><summary>
Show the code
</summary><div id="cb5" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the abundances heatmap, round values to 2 decimals</span></span>
<span><span class="va">dfm</span> <span class="op">&lt;-</span> <span class="fu">melt</span><span class="op">(</span><span class="fu">round</span><span class="op">(</span><span class="fu">abundances</span><span class="op">(</span><span class="va">pseqz</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">dfm</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Taxa"</span>, <span class="st">"Sample"</span>, <span class="st">"value"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">heat</span><span class="op">(</span><span class="va">dfm</span>, <span class="st">"Taxa"</span>, <span class="st">"Sample"</span>, <span class="st">"value"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>text<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>size<span class="op">=</span><span class="fl">10</span><span class="op">)</span>, </span>
<span>        axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>        legend.key.size <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fl">1.2</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output-display">
<div>
<figure class="figure"><p>
<img src="microbiome-data-report_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img figure" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h2 class="anchored" data-anchor-id="correlation-between-microbes-and-lipids" id="correlation-between-microbes-and-lipids">
Correlation between microbes and lipids
<a class="anchor" aria-label="anchor" href="#correlation-between-microbes-and-lipids"></a>
</h2>
<p>
Cross-correlate columns of two data sets from related to microbiome and
blood serum lipids associations
(<a href="https://peerj.com/articles/32/" class="external-link">PeerJ 1:e32</a>).
</p>
<p>
The function returns correlations, raw p-values, and fdr estimates (not
strictly proper as the comparisons are not independent). Keep only those
elements that have at least only one significant correlation (n.signif):
</p>
<div class="cell">
<details class="code-fold"><summary>
Show the code
</summary><div id="cb6" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load example data </span></span>
<span><span class="va">otu</span> <span class="op">&lt;-</span> <span class="va">peerj32</span><span class="op">$</span><span class="va">microbes</span> </span>
<span><span class="va">lipids</span> <span class="op">&lt;-</span> <span class="va">peerj32</span><span class="op">$</span><span class="va">lipids</span> </span>
<span></span>
<span><span class="co"># Define data sets to cross-correlate</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">log10</span><span class="op">(</span><span class="va">otu</span><span class="op">)</span> <span class="co"># OTU Log10 (44 samples x 130 genera)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">lipids</span><span class="op">)</span> <span class="co"># Lipids (44 samples x 389 lipids)</span></span>
<span></span>
<span><span class="co"># Cross correlate data sets</span></span>
<span><span class="va">correlations</span> <span class="op">&lt;-</span> <span class="fu">associate</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, method <span class="op">=</span> <span class="st">"spearman"</span>, mode <span class="op">=</span> <span class="st">"matrix"</span>, p.adj.threshold <span class="op">=</span> <span class="fl">0.05</span>, n.signif <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Or, alternatively, the same output is also available in a handy table format</span></span>
<span><span class="va">correlation.table</span> <span class="op">&lt;-</span> <span class="fu">associate</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, method <span class="op">=</span> <span class="st">"spearman"</span>, mode <span class="op">=</span> <span class="st">"table"</span>, p.adj.threshold <span class="op">=</span> <span class="fl">0.05</span>, n.signif <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="fu">head</span><span class="op">(</span><span class="va">correlation.table</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output-display">
<table class="table caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 7%">
<col style="width: 45%">
<col style="width: 15%">
<col style="width: 17%">
<col style="width: 14%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">
</th>
<th style="text-align: left;">
X1
</th>
<th style="text-align: left;">
X2
</th>
<th style="text-align: right;">
Correlation
</th>
<th style="text-align: right;">
p.adj
</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">
1648
</td>
<td style="text-align: left;">
Ruminococcus gnavus et rel.
</td>
<td style="text-align: left;">
TG(54:5).2
</td>
<td style="text-align: right;">
0.7164958
</td>
<td style="text-align: right;">
0.0022842
</td>
</tr>
<tr class="even">
<td style="text-align: left;">
384
</td>
<td style="text-align: left;">
Moraxellaceae
</td>
<td style="text-align: left;">
PC(40:3e)
</td>
<td style="text-align: right;">
-0.6941863
</td>
<td style="text-align: right;">
0.0029225
</td>
</tr>
<tr class="odd">
<td style="text-align: left;">
1829
</td>
<td style="text-align: left;">
Uncultured Bacteroidetes
</td>
<td style="text-align: left;">
TG(56:2).1
</td>
<td style="text-align: right;">
-0.6987375
</td>
<td style="text-align: right;">
0.0029225
</td>
</tr>
<tr class="even">
<td style="text-align: left;">
349
</td>
<td style="text-align: left;">
Lactobacillus plantarum et rel.
</td>
<td style="text-align: left;">
PC(40:3)
</td>
<td style="text-align: right;">
-0.6877976
</td>
<td style="text-align: right;">
0.0031520
</td>
</tr>
<tr class="odd">
<td style="text-align: left;">
1198
</td>
<td style="text-align: left;">
Ruminococcus gnavus et rel.
</td>
<td style="text-align: left;">
TG(52:5)
</td>
<td style="text-align: right;">
0.6806216
</td>
<td style="text-align: right;">
0.0037518
</td>
</tr>
<tr class="even">
<td style="text-align: left;">
264
</td>
<td style="text-align: left;">
Moraxellaceae
</td>
<td style="text-align: left;">
PC(38:4).1
</td>
<td style="text-align: right;">
-0.6700504
</td>
<td style="text-align: right;">
0.0038414
</td>
</tr>
</tbody>
</table>
</div>
</div>
<section class="level3"><h3 class="anchored" data-anchor-id="association-heatmaps" id="association-heatmaps">
Association heatmaps
<a class="anchor" aria-label="anchor" href="#association-heatmaps"></a>
</h3>
<p>
Rearrange the data and plot the heatmap and mark significant
correlations with stars to reproduce microbiota-lipidome heatmap from
<a href="https://peerj.com/articles/32/" class="external-link">Lahti et al. PeerJ (2013)</a>
(the ordering of rows and columns may be different):
</p>
<div class="cell">
<details class="code-fold"><summary>
Show the code
</summary><div id="cb7" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">heat</span><span class="op">(</span><span class="va">correlation.table</span>, <span class="st">"X1"</span>, <span class="st">"X2"</span>, </span>
<span>          fill <span class="op">=</span> <span class="st">"Correlation"</span>, </span>
<span>          star <span class="op">=</span> <span class="st">"p.adj"</span>, </span>
<span>          p.adj.threshold <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> </span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>text<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>size<span class="op">=</span><span class="fl">10</span><span class="op">)</span>, </span>
<span>          axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, </span>
<span>          legend.key.size <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fl">1.3</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output-display">
<div>
<figure class="figure"><p>
<img src="microbiome-data-report_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img figure" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level3"><h3 class="anchored" data-anchor-id="heatmap-of-significant-correlations" id="heatmap-of-significant-correlations">
Heatmap of significant correlations
<a class="anchor" aria-label="anchor" href="#heatmap-of-significant-correlations"></a>
</h3>
<p>
This time we are modifying the plot to only include significant
correlations with a threshold of P &lt; 0.05. The cells marked with a
white cross indicate most significant correlations with a significance
of P &lt; 0.02.
</p>
<div class="cell">
<details class="code-fold"><summary>
Show the code
</summary><div id="cb8" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Order the rows and columns with levels argument if needed:</span></span>
<span><span class="va">correlation.table</span><span class="op">$</span><span class="va">X1</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">correlation.table</span><span class="op">$</span><span class="va">X1</span>, levels <span class="op">=</span> <span class="fu">unique</span><span class="op">(</span><span class="fu">as.character</span><span class="op">(</span><span class="va">correlation.table</span><span class="op">$</span><span class="va">X1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">correlation.table</span><span class="op">$</span><span class="va">X2</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">correlation.table</span><span class="op">$</span><span class="va">X2</span>, levels <span class="op">=</span> <span class="fu">unique</span><span class="op">(</span><span class="fu">as.character</span><span class="op">(</span><span class="va">correlation.table</span><span class="op">$</span><span class="va">X2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set black-and-white theme</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="fu">theme_set</span><span class="op">(</span><span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Pick only the correlations with p.adj&lt;0.05</span></span>
<span><span class="co"># Note: this will leave other cells empty</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="va">subtable</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">correlation.table</span>, <span class="va">p.adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Arrange the figure</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">subtable</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">X1</span>, y <span class="op">=</span> <span class="va">X2</span>, fill <span class="op">=</span> <span class="va">Correlation</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">geom_tile</span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">scale_fill_gradientn</span><span class="op">(</span><span class="st">"Correlation"</span>, </span>
<span>                       breaks <span class="op">=</span> <span class="fu">seq</span><span class="op">(</span>from <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, to <span class="op">=</span> <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>, </span>
<span>                   colours <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"darkblue"</span>, <span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>, <span class="st">"darkred"</span><span class="op">)</span>, </span>
<span>                   limits <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Polish texts</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>axis.text.x<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust<span class="op">=</span><span class="fl">1</span>, face <span class="op">=</span> <span class="st">"italic"</span><span class="op">)</span>,</span>
<span>               axis.text.y<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">xlab</span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">""</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Mark the most significant cells with stars using threshold of P &lt; 0.02</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">geom_text</span><span class="op">(</span>data <span class="op">=</span> <span class="fu">subset</span><span class="op">(</span><span class="va">correlation.table</span>, <span class="va">p.adj</span> <span class="op">&lt;</span> <span class="fl">0.02</span><span class="op">)</span>, </span>
<span>               <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">X1</span>, y <span class="op">=</span> <span class="va">X2</span>, label <span class="op">=</span> <span class="st">"+"</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"white"</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output-display">
<div>
<figure class="figure"><p>
<img src="microbiome-data-report_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img figure" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold"><summary>
Show the code
</summary><div id="cb9" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.1 (2025-06-13)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Perth
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] gtsummary_2.4.0   knitr_1.50        reshape2_1.4.4    dplyr_1.1.4
[5] microbiome_1.30.0 ggplot2_4.0.0     phyloseq_1.52.0

loaded via a namespace (and not attached):
 [1] gtable_0.3.6            xfun_0.53               htmlwidgets_1.6.4
 [4] rhdf5_2.52.1            Biobase_2.68.0          lattice_0.22-7
 [7] rhdf5filters_1.20.0     vctrs_0.6.5             tools_4.5.1
[10] generics_0.1.4          biomformat_1.36.0       stats4_4.5.1
[13] parallel_4.5.1          tibble_3.3.0            cluster_2.1.8.1
[16] pkgconfig_2.0.3         Matrix_1.7-3            data.table_1.17.8
[19] RColorBrewer_1.1-3      S7_0.2.0                S4Vectors_0.46.0
[22] lifecycle_1.0.4         GenomeInfoDbData_1.2.14 compiler_4.5.1
[25] farver_2.1.2            stringr_1.5.2           Biostrings_2.76.0
[28] codetools_0.2-20        permute_0.9-8           GenomeInfoDb_1.44.3
[31] htmltools_0.5.8.1       yaml_2.3.10             tidyr_1.3.1
[34] pillar_1.11.1           crayon_1.5.3            MASS_7.3-65
[37] vegan_2.7-2             iterators_1.0.14        foreach_1.5.2
[40] nlme_3.1-168            tidyselect_1.2.1        digest_0.6.37
[43] Rtsne_0.17              stringi_1.8.7           purrr_1.1.0
[46] splines_4.5.1           ade4_1.7-23             fastmap_1.2.0
[49] grid_4.5.1              cli_3.6.5               magrittr_2.0.4
[52] utf8_1.2.6              survival_3.8-3          ape_5.8-1
[55] withr_3.0.2             UCSC.utils_1.4.0        scales_1.4.0
[58] rmarkdown_2.30          XVector_0.48.0          httr_1.4.7
[61] igraph_2.2.0            multtest_2.64.0         evaluate_1.0.5
[64] IRanges_2.42.0          mgcv_1.9-3              rlang_1.1.6
[67] Rcpp_1.1.0              glue_1.8.0              BiocGenerics_0.54.1
[70] rstudioapi_0.17.1       jsonlite_2.0.0          R6_2.6.1
[73] Rhdf5lib_1.30.0         plyr_1.8.9             </code></pre>
</div>
</div>
</section></section><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button,
          { trigger: "manual",
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config);
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined;
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              }
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            }
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><section><h2 class="section-heading" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<hr class="half-width">
<p>When sharing R Markdown (.Rmd) or Quarto (.qmd) notebooks, it’s good
practice to include information about your R session at the end of the
document. This helps others (and future you!) understand:</p>
<ul>
<li>Which version of R was used</li>
<li>What packages were loaded</li>
<li>What platform the code was run on</li>
</ul>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.5.2 (2025-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] knitr_1.50        microbiome_1.32.0 ggplot2_4.0.0     phyloseq_1.54.0

loaded via a namespace (and not attached):
 [1] gtable_0.3.6        xfun_0.54           rhdf5_2.54.0
 [4] Biobase_2.70.0      lattice_0.22-5      rhdf5filters_1.22.0
 [7] vctrs_0.6.5         tools_4.5.2         generics_0.1.4
[10] biomformat_1.38.0   stats4_4.5.2        parallel_4.5.2
[13] tibble_3.3.0        cluster_2.1.8.1     pkgconfig_2.0.3
[16] Matrix_1.7-4        data.table_1.17.8   RColorBrewer_1.1-3
[19] S7_0.2.0            S4Vectors_0.48.0    lifecycle_1.0.4
[22] compiler_4.5.2      farver_2.1.2        stringr_1.6.0
[25] Biostrings_2.78.0   Seqinfo_1.0.0       codetools_0.2-19
[28] permute_0.9-8       htmltools_0.5.8.1   yaml_2.3.10
[31] pillar_1.11.1       crayon_1.5.3        tidyr_1.3.1
[34] MASS_7.3-65         vegan_2.7-2         iterators_1.0.14
[37] foreach_1.5.2       nlme_3.1-168        tidyselect_1.2.1
[40] digest_0.6.37       Rtsne_0.17          stringi_1.8.7
[43] purrr_1.2.0         dplyr_1.1.4         reshape2_1.4.4
[46] splines_4.5.2       fastmap_1.2.0       ade4_1.7-23
[49] grid_4.5.2          cli_3.6.5           magrittr_2.0.4
[52] survival_3.8-3      ape_5.8-1           withr_3.0.2
[55] scales_1.4.0        XVector_0.50.0      igraph_2.2.1
[58] multtest_2.66.0     evaluate_1.0.5      IRanges_2.44.0
[61] mgcv_1.9-1          rlang_1.1.6         Rcpp_1.1.0
[64] glue_1.8.0          BiocManager_1.30.26 renv_1.1.5
[67] BiocGenerics_0.56.0 jsonlite_2.0.0      R6_2.6.1
[70] Rhdf5lib_1.32.0     plyr_1.8.9         </code></pre>
</div>
<p>Some other similar handy functions include:</p>
<ul>
<li>
<code>devtools::session_info()</code> - more detailed, especially
for package dependencies, and identifies where packages were installed
from.</li>
<li>
<code>renv::snapshot()</code> – for managing project-specific
environments.</li>
</ul></section><section><h2 class="section-heading" id="recommendations">Recommendations<a class="anchor" aria-label="anchor" href="#recommendations"></a>
</h2>
<hr class="half-width">
<p>Some recommendations for rendering your <code>.qmd</code>:</p>
<ul>
<li>Hide warnings and messages from your R chunks</li>
<li>Keeping your code in the output can be good to see what data was
used for analysis and if any filtering or transformations were used,
however it is not always neccessary and can make it difficult for other
“non-coders” to read. In the “YAML” at the top you will see the option
<code>code-fold: true</code> is used to allow us to toggle the code on
and off.</li>
<li>Include the <code>sessionInfo()</code> at the bottom of your file to
keep track of package versions used!</li>
</ul>
<hr>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Saving your data in <code>.rda</code> format can make it easier to
load and repeat data analysis.</li>
<li>A notebook in <code>.rmd</code> or <code>.qmd</code> format and
knitted to html can be an efficient way to store and share data
workflows and analysis</li>
<li>Use <code>sessionInfo()</code> when using R to generate reports as a
handy way to track package versions used to generate analysis and
figures.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/" class="external-link">Source</a></p>
				<p><a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:siobhon.egan@murdoch.edu.au">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://siobhon-egan.github.io/2025-11-13-COMBINE-WA/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://siobhon-egan.github.io/2025-11-13-COMBINE-WA/aio.html",
  "identifier": "https://siobhon-egan.github.io/2025-11-13-COMBINE-WA/aio.html",
  "dateCreated": "2025-11-13",
  "dateModified": "2025-11-11",
  "datePublished": "2025-11-11"
}

  </script><script>
		feather.replace();
	</script>
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

