<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Intermediate R - biological data: Bioconductor Introduction &amp; Install</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png"><link rel="manifest" href="../favicons/incubator/site.webmanifest"><link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../02-bioconductor.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Intermediate R - biological data
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Intermediate R - biological data
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Intermediate R - biological data
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 10%" class="percentage">
    10%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 10%" aria-valuenow="10" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../02-bioconductor.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-introduction.html">1. Introduction</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        2. Bioconductor Introduction &amp; Install
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#what-is-bioconductor">What is Bioconductor?</a></li>
<li><a href="#biocmanager">BiocManager</a></li>
<li><a href="#bioconductor-releases-and-current-version">Bioconductor releases and current version</a></li>
<li><a href="#check-for-updates">Check for updates</a></li>
<li><a href="#exploring-the-package-repository">Exploring the package repository</a></li>
<li><a href="#installing-packages">Installing packages</a></li>
<li><a href="#uninstalling-packages">Uninstalling packages</a></li>
<li><a href="#getting-help">Getting help</a></li>
<li><a href="#references">References</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-biological-sequences.html">3. Biological sequences</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-annotations.html">4. Working with annotations</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-s4.html">5. The S4 class</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-application-microbiome.html">6. Microbiome application</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-reproducible-analysis.html">7. Communicating &amp; Reproducing Analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/01-introduction.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/03-biological-sequences.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/01-introduction.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction
        </a>
        <a class="chapter-link float-end" href="../instructor/03-biological-sequences.html" rel="next">
          Next: Biological sequences...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Bioconductor Introduction &amp; Install</h1>
        <p>Last updated on 2025-11-11 |

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/episodes/02-bioconductor.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 20 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What does the Bioconductor project comprise?</li>
<li>How does the Bioconductor project relate to the CRAN
repository?</li>
<li>How can I learn to use Bioconductor packages effectively?</li>
<li>How do I join and communicate with the Bioconductor community?</li>
<li>How do I install Bioconductor packages?</li>
<li>How do I check if newer versions of my installed packages are
available?</li>
<li>How do I update Bioconductor packages?</li>
<li>How do I find out the name of packages available from the
Bioconductor repositories?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Describe the Bioconductor project globally.</li>
<li>Gain a global view of the Bioconductor project in the R
ecosystem.</li>
<li>Identify sources of information to watch for future updates about
the Bioconductor project.</li>
<li>Install BiocManager.</li>
<li>Install Bioconductor packages.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="what-is-bioconductor">What is Bioconductor?<a class="anchor" aria-label="anchor" href="#what-is-bioconductor"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="a-brief-history-of-bioconductor">A brief history of Bioconductor<a class="anchor" aria-label="anchor" href="#a-brief-history-of-bioconductor"></a></h3>
<p>The Bioconductor project was started in the Fall of 2001, as an
initiative for the collaborative creation of extensible software for
computational biology and bioinformatics (Gentleman, Carey, Bates,
Bolstad, Dettling, Dudoit, Ellis, Gautier, Ge, Gentry, Hornik, Hothorn,
Huber, Iacus, Irizarry, Leisch, Li, Maechler, Rossini, Sawitzki, Smith,
Smyth, Tierney, Yang, and Zhang, 2004). From the very start, the stated
mission of the project was to develop tools for the statistical analysis
and comprehension of large datasets and technological artifacts in
rigorously and robustly designed experiments. Beyond statistical
analyses, the interpretation of statistical results is supported by
packages providing biological context, visualization, and
reproducibility.</p>
<p>Over the years, software packages contributed to the Bioconductor
project have reflected the evolution and emergence of several
high-throughput technologies, from microarrays to single-cell genomics,
through many variations of sequencing experiments (e.g., RNA-seq,
ChIP-seq, DNA-seq), analyses (e.g., sequence variation, copy number
variation, single nucleotide polymorphisms), and data modalities (e.g.,
flow cytometry, proteomics, microscopy and image analysis).</p>
<p>Crucially, the project has not only released software packages
implementing novel statistical tests and methodologies, but also
produced a diverse range of packages types granting access to databases
of molecular annotations and experimental datasets.</p>
<p>The Bioconductor project culminates at an annual conference in North
America in the summer, while regional conferences offer great
opportunities for networking in Europe, Asia, and North America. The
project is committed to promote a diverse and inclusive community,
including a <a href="https://www.bioconductor.org/about/code-of-conduct/" class="external-link">Code of
Conduct</a> enforced by a Code of Conduct committee.</p>
<figure><img src="../fig/bioc-timeline.svg" alt="Timeline of major Bioconductor milestones alongside technological advancements." class="figure mx-auto d-block"></figure><p><strong>Timeline of major Bioconductor milestones alongside
technological advancements.</strong> Above the timeline, the figure
marks the first occurence of major events. Within the timeline, the name
of packages providing core infrastructure indicate the release date.
Below the timeline, major technological advancements contextualise the
evolution of the Bioconductor project over time.</p>
</div>
<div class="section level3">
<h3 id="a-scientific-project">A scientific project<a class="anchor" aria-label="anchor" href="#a-scientific-project"></a></h3>
<p>The original publication describes the aims and methods of the
project at its inception Gentleman, Carey, Bates et al. (2004).</p>
<p>Huber, Carey, Gentleman, Anders, Carlson, Carvalho, Bravo, Davis,
Gatto, Girke, Gottardo, Hahne, Hansen, Irizarry, Lawrence, Love,
MacDonald, Obenchain, Oles, Pages, Reyes, Shannon, Smyth, Tenenbaum,
Waldron, and Morgan (2015) illustrates the progression of the project,
including descriptions of core infrastructure and case studies, from the
perspective of both users and developers.</p>
<p>Amezquita, Lun, Becht, Carey, Carpp, Geistlinger, Marini,
Rue-Albrecht, Risso, Soneson, Waldron, Pages, Smith, Huber, Morgan,
Gottardo, and Hicks (2020) reviewed further developments of the project
in the wake of single-cell genomics technologies.</p>
<p>Many more publications and book chapters cite the Bioconductor
project, with recent example listed on the <a href="https://www.bioconductor.org/help/publications/" class="external-link">Bioconductor
website</a>.</p>
</div>
<div class="section level3">
<h3 id="a-package-repository">A package repository<a class="anchor" aria-label="anchor" href="#a-package-repository"></a></h3>
<div class="section level4">
<h4 id="overview-and-relationship-to-cran">Overview and relationship to CRAN<a class="anchor" aria-label="anchor" href="#overview-and-relationship-to-cran"></a></h4>
<p>Undoubtedly, software packages are the best-known aspect of the
Bioconductor project. Since its inception in 2001, the repository has
grown over time to host thousands of packages.</p>
<p>The Bioconductor project has extended the preexisting CRAN repository
– much larger and general-purpose in scope – to comprise R packages
primarily catering for bioinformatics and computational biology
analyses.</p>
<div id="going-further" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="going-further" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>The <a href="discuss.html#bioconductor-and-CRAN">Discussion</a>
article of this lesson includes a section discussing the relationship of
Bioconductor and CRAN in further details.</p>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="the-bioconductor-release-cycle">The Bioconductor release cycle<a class="anchor" aria-label="anchor" href="#the-bioconductor-release-cycle"></a></h4>
<p>The Bioconductor project also extended the packaging infrastructure
of the CRAN repository to better support the deployment and management
of packages at the user level (Gentleman, Carey, Bates et al., 2004). In
particular, the Bioconductor projects features a 6-month release cycle
(typically around April and October), which sees a snapshot of the
current version of all packages in the Bioconductor repository earmarked
for a specific version of R. R itself is released on an annual basis
(typically around April), meaning that for each release of R, two
compatible releases of Bioconductor packages are available.</p>
<p>As such, Bioconductor package developers are required to always use
the version of R that will be associated with the next release of the
Bioconductor project. This means using the development version of R
between October and April, and the release version of R between April
and October.</p>
<p>Crucially, the strict Bioconductor release cycle prevents users from
installing temporally distant versions of packages that were very likely
never tested together. This practice reflects the development cycle of
packages of both CRAN and Bioconductor, where contemporaneous packages
are regularly tested by automated systems to ensure that the latest
software updates in package dependencies do not break downstream
packages, or prompts those package maintainers to update their own
software as a consequence.</p>
<p>Prior to each Bioconductor release, packages that do not pass the
requires suites of automated tests are deprecated and subsequently
removed from the repository. This ensures that each Bioconductor release
provides a suite of packages that are mutually compatible, traceable,
and guaranteed to function for the associated version of R.</p>
<figure><img src="../fig/bioc-release-cycle.svg" alt="Timeline of release dates for selected Bioconductor and R versions." class="figure mx-auto d-block"></figure><p><strong>Timeline of release dates for selected Bioconductor and R
versions.</strong> The upper section of the timeline indicates versions
and approximate release dates for the R project. The lower section of
the timeline indicates versions and release dates for the Bioconductor
project. Source: <a href="https://bioconductor.org/about/release-announcements/" class="external-link">Bioconductor</a>.</p>
</div>
<div class="section level4">
<h4 id="package-types">Package types<a class="anchor" aria-label="anchor" href="#package-types"></a></h4>
<p>Packages are broadly divided in four major categories:</p>
<ul><li>software</li>
<li>annotation data</li>
<li>experiment data</li>
<li>workflows</li>
</ul><p><a href="reference.html#software-package">Software packages</a>
themselves can be subdivided into packages that provide infrastructure
(i.e., classes) to store and access data, and packages that provide
methodological tools to process data stored in those data structures.
This separation of structure and analysis is at the core of the
Bioconductor project, encouraging developers of new methodological
software packages to thoughtfully re-use existing data containers where
possible, and reducing the cognitive burden imposed on users who can
more easily experiment with alternative workflows without the need to
learn and convert between different data structures.</p>
<p><a href="reference.html#annotationdata-package">Annotation data
packages</a> provide self-contained databases of diverse genomic
annotations (e.g., gene identifiers, biological pathways). Different
collections of annotation packages can be found in the Bioconductor
project. They are identifiable by their respective naming pattern, and
the information that they contain. For instance, the so-called
<code>OrgDb</code> packages (e.g., the <em><a href="https://bioconductor.org/packages/3.22/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>
package) provide information mapping different types of gene identifiers
and pathway databases; the so-called <code>EnsDb</code> (e.g., <em><a href="https://bioconductor.org/packages/3.22/EnsDb.Hsapiens.v86" class="external-link">EnsDb.Hsapiens.v86</a></em>)
packages encapsulate individual versions of the Ensembl annotations in
Bioconductor packages; and the so-called <code>TxDb</code> packages
(e.g., <em><a href="https://bioconductor.org/packages/3.22/TxDb.Hsapiens.UCSC.hg38.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg38.knownGene</a></em>)
encapsulate individual versions UCSC gene annotation tables.</p>
<p><a href="reference.html#experimentdata-package">Experiment data
packages</a> provide self-contained datasets that are often used by
software package developers to demonstrate the use of their package on
well-known standard datasets in their <a href="reference.html#vignette">package vignettes</a>.</p>
<p>Finally, <a href="reference.html#workflow-package">workflow
packages</a> exclusively provide collections of vignettes that
demonstrate the combined usage of several other packages as a coherent
workflow, but do not provide any new source code or functionality
themselves.</p>
<div id="challenge-the-bioconductor-website" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-the-bioconductor-website" class="callout-inner">
<h3 class="callout-title">Challenge: The Bioconductor website</h3>
<div class="callout-content">
<p>The Bioconductor website is accessible at <a href="https://bioconductor.org/" class="external-link">https://bioconductor.org/</a>.</p>
<p>Browse the website to find information answering the following
questions:</p>
<ol style="list-style-type: decimal"><li>How many packages does the current release of the Bioconductor
project include?</li>
<li>How many packages of each type does this number include?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>The following solution includes numbers that were valid at the time
of writing (Bioconductor release 3.13); numbers will inevitably be
different for future releases of the Bioconductor project.</p>
<ol style="list-style-type: decimal"><li>On the page <a href="https://bioconductor.org/" class="external-link">https://bioconductor.org/</a>, in the
section “Install”, we can read:</li>
</ol><blockquote>
<p>Discover 2042 software packages available in Bioconductor release
3.13.</p>
</blockquote>
<ol start="2" style="list-style-type: decimal"><li>On the page <a href="https://bioconductor.org/" class="external-link">https://bioconductor.org/</a>, in the
section “News”, click on the link that reads “Bioconductor Bioc
<code>X.Y</code> Released” (<code>X.Y</code> being the version of the
current Bioconductor release when you go through this exercise
yourself). On the linked page, we can read:</li>
</ol><blockquote>
<p>We are pleased to announce Bioconductor 3.13, consisting of 2042
software packages, 406 experiment data packages, 965 annotation
packages, and 29 workflows.</p>
<p>There are 133 new software packages, 22 new data experiment packages,
7 new annotation packages, 1 new workflow, no new books, and many
updates and improvements to existing packages; Bioconductor 3.13 is
compatible with R 4.1.0, and is supported on Linux, 32- and 64-bit
Windows, and macOS 10.14.6 Mojave or higher. This release will include
an updated Bioconductor Docker containers.</p>
</blockquote>
</div>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="package-classification-using-biocviews">Package classification using biocViews<a class="anchor" aria-label="anchor" href="#package-classification-using-biocviews"></a></h4>
<p>The Bioconductor project uses <a href="reference.html#biocviews">biocViews</a>, a set of terms from a
controlled vocabulary, to classify Bioconductor packages and facilitate
their discovery by thematic search on the <a href="https://www.bioconductor.org/packages/release/BiocViews.html" class="external-link">Bioconductor
website</a>.</p>
<p>Each Bioconductor package is tagged with a small set of terms chosen
from the available controlled vocabulary, to describe the type and
functionality of the package. Terms are initially selected by the
package authors, and subsequently refined during package review or
updates to the controlled vocabulary.</p>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Visit the listing of all packages on the Bioconductor <a href="https://www.bioconductor.org/packages/release/BiocViews.html" class="external-link">biocViews</a>
web page. Use the “Autocomplete biocViews search” box in the upper left
to filter packages by category and explore the graph of software
packages by expanding and contracting individual terms.</p>
<ol style="list-style-type: decimal"><li>What biocView terms can be used to identify packages that have been
tagged for RNA sequencing analysis? ChIP-seq? Epigenetics? Variant
annotation? Proteomics? Single-cell genomics?</li>
<li>In the <code>RNASeq</code> category, two very popular packages are
<em><a href="https://bioconductor.org/packages/3.22/DESeq2" class="external-link">DESeq2</a></em> and
<em><a href="https://bioconductor.org/packages/3.22/edgeR" class="external-link">edgeR</a></em>.
Which one is more popular in terms of download statistics (i.e., lower
rank)?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<ol style="list-style-type: decimal"><li>
<code>RNAseq</code>, <code>ChIPSeq</code>, <code>Epigenetics</code>,
<code>VariantAnnotation</code>, <code>Proteomics</code>,
<code>SingleCell</code>.</li>
<li>For Bioconductor release <code>3.14</code>, <em><a href="https://bioconductor.org/packages/3.22/DESeq2" class="external-link">DESeq2</a></em> and
<em><a href="https://bioconductor.org/packages/3.22/edgeR" class="external-link">edgeR</a></em> are
listed at ranks 23 and 28 respectively. In other words, the two packages
are among the most frequently downloaded packages in the Bioconductor
project, in this instance with a small advantage in favour of <em><a href="https://bioconductor.org/packages/3.22/edgeR" class="external-link">edgeR</a></em>.</li>
</ol></div>
</div>
</div>
</div>
<div id="going-further-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="going-further-1" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>The Bioconductor package <em><a href="https://bioconductor.org/packages/3.22/biocViews" class="external-link">biocViews</a></em>
is used to support and manage the infrastructure of the controlled
vocabulary. It can also be used to programmatically inspect and subset
the list of terms available using their relationship as a graph.</p>
<p>Furthermore, the <em><a href="https://bioconductor.org/packages/3.22/BiocPkgTools" class="external-link">BiocPkgTools</a></em>
package can be used to browse packages under different <a href="reference.html#biocviews">biocViews</a> (Su, Carey, Shepherd,
Ritchie, Morgan, and Davis, 2019).</p>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="packages-interoperability">Packages interoperability<a class="anchor" aria-label="anchor" href="#packages-interoperability"></a></h4>
<p>At the core of the Bioconductor philosophy is the notion of
interoperability. Namely, the capacity of packages to operate on the
same data structures. Importantly, interoperability benefits both users
and developers.</p>
<p>Users can more easily write arbitrarily complex workflows that
combine multiple packages. With packages operating on the same data
structure, users can maximize their attention to the practical steps of
their workflow, and minimize time spent in often complex and error-prone
conversions between different data structures specific to each package.
Comparative benchmarks are also easier to implement and can be evaluated
more fairly when competing software packages offering similar
functionality operate on input and outputs stored in the same data
structures.</p>
<p>Similarly, developers of new packages can focus on the implementation
of novel functionality, borrowing existing data structures that offer
robust and trusted infrastructure for storage, verification, and
indexing of information.</p>
<p>Ultimately, the figure below illustrates how many different
Bioconductor packages - as well as base R packages - can be combined to
carry out a diverse range of analyses, from importing sequencing data
into an R session, to the annotation, integration and visualization of
data and results.</p>
<figure><img src="../fig/bioc-sequencing-ecosystem.svg" alt="Sequencing Ecosystem." class="figure mx-auto d-block"></figure><p><strong>Sequencing Ecosystem</strong> Major data processing steps
(blue) and relevant software packages (pink) are listed in the context
of archetypal workflows for various types of genomics analyses. The
sequential relation of workflow steps and software package illustrates
the importance of interoperability between software package in order to
assemble complete end-to-end workflows.</p>
</div>
</div>
<div class="section level3">
<h3 id="conferences-courses-and-workshops">Conferences, courses and workshops<a class="anchor" aria-label="anchor" href="#conferences-courses-and-workshops"></a></h3>
<p>The Bioconductor community regularly organizes a number of events
throughout the year and across the world. For example:</p>
<ul><li>The annual BioC summer conference in North America</li>
<li>Regional conference in winter (e.g. BioC Europe, BioC Asia)</li>
<li>Summer schools (e.g., CSAMA)</li>
<li>Online meetings open to all community members (e.g., Bioconductor
Developers Forum)</li>
</ul><p>Course materials are regularly uploaded on the <a href="https://bioconductor.org/help/course-materials/" class="external-link">Bioconductor
website</a> following each of those events. In particular, <a href="https://www.bioconductor.org/books/release/" class="external-link">online books</a> are
being developed and maintained by community members.</p>
<p>The Bioconductor <a href="https://www.youtube.com/user/bioconductor" class="external-link">YouTube channel</a> is
used to publish video recordings of conference presentations including
talks and workshops, as well as editions of the regular Bioconductor
developers forum (link needed).</p>
</div>
<div class="section level3">
<h3 id="online-communication-channels">Online communication channels<a class="anchor" aria-label="anchor" href="#online-communication-channels"></a></h3>
<div class="section level4">
<h4 id="support-site">Support site<a class="anchor" aria-label="anchor" href="#support-site"></a></h4>
<p>The Bioconductor <a href="https://support.bioconductor.org/" class="external-link">support
site</a> provides a platform where users and developers can communicate
freely (following the Bioconductor <a href="https://www.bioconductor.org/about/code-of-conduct/" class="external-link">Code of
Conduct</a>) to discuss issues on a range of subjects, ranging from
packages to conceptual questions about best practices.</p>
</div>
<div class="section level4">
<h4 id="slack-workspace">Slack workspace<a class="anchor" aria-label="anchor" href="#slack-workspace"></a></h4>
<p>The Bioconductor <a href="https://bioc-community.herokuapp.com/" class="external-link">Slack workspace</a> is an
open space that all community members are welcome to join (for free) and
use for rapid interactions. Currently, the “Pro” pricing plan kindly
supported by core funding provides:</p>
<ul><li>Unlimited message archive</li>
<li>Unlimited apps</li>
<li>Group video calls with screen sharing</li>
<li>Work securely with other organizations using Slack Connect</li>
</ul><p>A wide range of channels have been created to discuss a range of
subjects, and community members can freely join the discussion on those
channels of create new ones to discuss new subjects.</p>
<p>Important announcements are posted on the <code>#general</code>
channel.</p>
<div id="note" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="note" class="callout-inner">
<h3 class="callout-title">Note</h3>
<div class="callout-content">
<p>Users are encouraged to use the Bioconductor <a href="https://support.bioconductor.org/" class="external-link">support site</a> to raise
issues that are relevant to the wider community. The Slack workspace is
often most useful for live discussions, and widely subscribed channels
(e.g. <code>#general</code>) should be used with moderation.</p>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="developer-mailing-list">Developer Mailing List<a class="anchor" aria-label="anchor" href="#developer-mailing-list"></a></h4>
<p>The <a href="mailto:bioc-devel@r-project.org">bioc-devel@r-project.org</a>
mailing list is used for communication between package developers, and
announcements from the Biocondutor core team.</p>
</div>
</div>
<div class="section level3">
<h3 id="a-scientific-and-technical-community">A scientific and technical community<a class="anchor" aria-label="anchor" href="#a-scientific-and-technical-community"></a></h3>
<ul><li>
<a href="https://bioconductor.org/about/scientific-advisory-board/" class="external-link">Scientific
Advisory Board (SAB)</a> Meet Annually, External and Internal leader in
the field who act as project advisors. No Term limits.</li>
<li>
<a href="https://bioconductor.org/about/technical-advisory-board/" class="external-link">Technical
Advisory Board (TAB)</a>. Meet monthly to consider technical aspects of
core infastructure and scientific direction of the project. 15 members,
3 year term. Annual open-to-all elections to rotate members. Current
officers are Vince Carey (chair), Levi Waldron (vice Chair) Charlotte
Soneson (Secretary).</li>
<li>
<a href="https://bioconductor.org/about/community-advisory-board/" class="external-link">Community
Advisory Board (CAB)</a> Meet monthly to consider community outreach,
events, education and training. 15 members, 3 year term. Annual
open-to-all elections to rotate members. Current officers are Aedin
Culhane (chair), Matt Ritchie (co Chair), Lori Kern (Secretary).</li>
<li><a href="https://www.bioconductor.org/about/code-of-conduct/" class="external-link">Code
of Conduct committee</a></li>
</ul><div id="note-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="note-1" class="callout-inner">
<h3 class="callout-title">Note</h3>
<div class="callout-content">
<p>At least 1 member of TAB/CAB sits on both to act at the liason to
ensure communication of the board.</p>
</div>
</div>
</div>
<hr></div>
</section><section><h2 class="section-heading" id="biocmanager">BiocManager<a class="anchor" aria-label="anchor" href="#biocmanager"></a></h2>
<hr class="half-width"><p>The <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em>
package is the entry point into the Bioconductor package repository.
Technically, this is the only Bioconductor package distributed on the
CRAN repository.</p>
<p>It provides functions to safely install Bioconductor packages and
check for available updates.</p>
<p>Once the package is installed, the function
<code>BiocManager::install()</code> can be used to install packages from
the Bioconductor repository. The function is also capable of installing
packages from other repositories (e.g., CRAN), if those packages are not
found in the Bioconductor repository first.</p>
<figure><img src="../fig/bioc-install.svg" alt="The package BiocManager is available from the CRAN repository and used to install packages from the Bioconductor repository." class="figure mx-auto d-block"></figure><p><strong>The package BiocManager is available from the CRAN repository
and used to install packages from the Bioconductor repository.</strong>
The function <code>install.packages()</code> from the base R package
<code>utils</code> can be used to install the <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em>
package distributed on the CRAN repository. In turn, the function
<code>BiocManager::install()</code> can be used to install packages
available on the Bioconductor repository. Notably, the
<code>BiocManager::install()</code> function will fall back on the CRAN
repository if a package cannot be found in the Bioconductor
repository.</p>
<p>Install the package using the code below.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre>
</div>
<div id="going-further-2" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="going-further-2" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>A number of packages that are not part of the base R installation
also provide functions to install packages from various repositories.
For instance:</p>
<ul><li><code>devtools::install()</code></li>
<li><code>remotes::install_bioc()</code></li>
<li><code>remotes::install_bitbucket()</code></li>
<li><code>remotes::install_cran()</code></li>
<li><code>remotes::install_dev()</code></li>
<li><code>remotes::install_github()</code></li>
<li><code>remotes::install_gitlab()</code></li>
<li><code>remotes::install_git()</code></li>
<li><code>remotes::install_local()</code></li>
<li><code>remotes::install_svn()</code></li>
<li><code>remotes::install_url()</code></li>
<li><code>renv::install()</code></li>
</ul><p>Those functions are beyond the scope of this lesson, and should be
used with caution and adequate knowledge of their specific behaviors.
The general recommendation is to use <code>BiocManager::install()</code>
over any other installation mechanism because it ensures proper
versioning of Bioconductor packages.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="bioconductor-releases-and-current-version">Bioconductor releases and current version<a class="anchor" aria-label="anchor" href="#bioconductor-releases-and-current-version"></a></h2>
<hr class="half-width"><p>Once the <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em>
package is installed, the <code>BiocManager::version()</code> function
displays the version (i.e., release) of the Bioconductor project that is
currently active in the R session.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">version</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] '3.22'</code></pre>
</div>
<p>Using the correct version of R and Bioconductor packages is a key
aspect of reproducibility. The <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em>
packages uses the version of R running in the current session to
determine the version of Biocondutor packages that can be installed in
the current R library.</p>
<p>The Bioconductor project produces two releases each year, one around
April and another one around October. The April release of Bioconductor
coincides with the annual release of R. The October release of
Bioconductor continues to use the same version of R for that annual
cycle (i.e., until the next release, in April).</p>
<figure><img src="../fig/bioc-release-cycle.svg" alt="Timeline of release dates for selected Bioconductor and R versions." class="figure mx-auto d-block"></figure><p><strong>Timeline of release dates for selected Bioconductor and R
versions.</strong> The upper section of the timeline indicates versions
and approximate release dates for the R project. The lower section of
the timeline indicates versions and release dates for the Bioconductor
project. Source: <a href="https://bioconductor.org/about/release-announcements/" class="external-link">Bioconductor</a>.</p>
<p>During each 6-month cycle of package development, Bioconductor tests
packages for compatibility with the version of R that will be available
for the next release cycle. Then, each time a new Bioconductor release
is produced, the version of every package in the Bioconductor repository
is incremented, including the package <em><a href="https://bioconductor.org/packages/3.22/BiocVersion" class="external-link">BiocVersion</a></em>
which determines the version of the Bioconductor project.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">packageVersion</span><span class="op">(</span><span class="st">"BiocVersion"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] '3.22.0'</code></pre>
</div>
<p>This is the case for every package, even those which have not been
updated at all since the previous release. That new version of each
package is earmarked for the corresponding version of R; in other words,
that version of the package can only be installed and accessed in an R
session that uses the correct version of R. This version increment is
essential to associate a each version of a Bioconductor package with a
unique release of the Bioconductor project.</p>
<p>Following the April release, this means that users must install the
new version of R to access the newly released versions of Bioconductor
packages.</p>
<p>Instead, in October, users can continue to use the same version of R
to access the newly released version of Bioconductor packages. However,
to update an R library from the April release to the October release of
Bioconductor, users need to call the function
<code>BiocManager::install()</code> specifying the correct version of
Bioconductor as the <code>version</code> option, for instance:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.14"</span><span class="op">)</span></span></code></pre>
</div>
<p>This needs to be done only once, as the <em><a href="https://bioconductor.org/packages/3.22/BiocVersion" class="external-link">BiocVersion</a></em>
package will be updated to the corresponding version, indicating the
version of Bioconductor in use in this R library.</p>
<div id="going-further-3" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="going-further-3" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>The <a href="discuss.html#the-bioconductor-release-cycle">Discussion</a>
article of this lesson includes a section discussing the release cycle
of the Bioconductor project.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="check-for-updates">Check for updates<a class="anchor" aria-label="anchor" href="#check-for-updates"></a></h2>
<hr class="half-width"><p>The <code>BiocManager::valid()</code> function inspects the version
of packages currently installed in the user library, and checks whether
a new version is available for any of them on the Bioconductor
repository.</p>
<p>If everything is up-to-date, the function will simply print
<code>TRUE</code>.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">valid</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: 1 packages out-of-date; 0 packages too new</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
* sessionInfo()

R version 4.5.2 (2025-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] bibtex_0.5.1     RefManageR_1.4.0 BiocStyle_2.38.0

loaded via a namespace (and not attached):
 [1] httr_1.4.7          cli_3.6.5           knitr_1.50
 [4] rlang_1.1.6         xfun_0.54           stringi_1.8.7
 [7] renv_1.1.5          generics_0.1.4      jsonlite_2.0.0
[10] glue_1.8.0          backports_1.5.0     plyr_1.8.9
[13] htmltools_0.5.8.1   rmarkdown_2.29      evaluate_1.0.5
[16] fastmap_1.2.0       yaml_2.3.10         lifecycle_1.0.4
[19] BiocManager_1.30.26 stringr_1.6.0       compiler_4.5.2
[22] Rcpp_1.1.0          timechange_0.3.0    digest_0.6.37
[25] R6_2.6.1            magrittr_2.0.4      tools_4.5.2
[28] lubridate_1.9.4     xml2_1.4.1

Bioconductor version '3.22'

  * 1 packages out-of-date
  * 0 packages too new

create a valid installation with

  BiocManager::install("rmarkdown", update = TRUE, ask = FALSE, force = TRUE)

more details: BiocManager::valid()$too_new, BiocManager::valid()$out_of_date</code></pre>
</div>
<p>Conveniently, if any package can be updated, the function generates
and displays the command needed to update those packages. Users simply
need to copy-paste and run that command in their R console.</p>
<div id="example-of-out-of-date-package-library" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="example-of-out-of-date-package-library" class="callout-inner">
<h3 class="callout-title">Example of out-of-date package library</h3>
<div class="callout-content">
<p>In the example below, the <code>BiocManager::valid()</code> function
did not return <code>TRUE</code>. Instead, it includes information about
the active user session, and displays the exact call to
<code>BiocManager::install()</code> that the user should run to replace
all the outdated packages detected in the user library with the latest
version available in CRAN or Bioconductor.</p>
<pre><code>&gt; BiocManager::valid()

* sessionInfo()

R version 4.1.0 (2021-05-18)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 11.6

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base

loaded via a namespace (and not attached):
[1] BiocManager_1.30.16 compiler_4.1.0      tools_4.1.0         renv_0.14.0

Bioconductor version '3.13'

  * 18 packages out-of-date
  * 0 packages too new

create a valid installation with

  BiocManager::install(c(
    "cpp11", "data.table", "digest", "hms", "knitr", "lifecycle", "matrixStats", "mime", "pillar", "RCurl",
    "readr", "remotes", "S4Vectors", "shiny", "shinyWidgets", "tidyr", "tinytex", "XML"
  ), update = TRUE, ask = FALSE)

more details: BiocManager::valid()$too_new, BiocManager::valid()$out_of_date

Warning message:
18 packages out-of-date; 0 packages too new </code></pre>
<p>Specifically, in this example, the message tells the user to run the
following command to bring their installation up to date:</p>
<pre><code><span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"cpp11"</span>, <span class="st">"data.table"</span>, <span class="st">"digest"</span>, <span class="st">"hms"</span>, <span class="st">"knitr"</span>, <span class="st">"lifecycle"</span>, <span class="st">"matrixStats"</span>, <span class="st">"mime"</span>, <span class="st">"pillar"</span>, <span class="st">"RCurl"</span>,</span>
<span>    <span class="st">"readr"</span>, <span class="st">"remotes"</span>, <span class="st">"S4Vectors"</span>, <span class="st">"shiny"</span>, <span class="st">"shinyWidgets"</span>, <span class="st">"tidyr"</span>, <span class="st">"tinytex"</span>, <span class="st">"XML"</span></span>
<span>  <span class="op">)</span>, update <span class="op">=</span> <span class="cn">TRUE</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exploring-the-package-repository">Exploring the package repository<a class="anchor" aria-label="anchor" href="#exploring-the-package-repository"></a></h2>
<hr class="half-width"><p>The Bioconductor <a href="reference.html#biocviews">biocViews</a>,
demonstrated in the earlier episode <a href="https://carpentries-incubator.github.io/bioc-project/02-introduction-to-bioconductor/index.html#package-classification-using-biocviews" class="external-link">Introduction
to Bioconductor</a>, are a great way to discover new packages by
thematically browsing the hierarchical classification of Bioconductor
packages.</p>
<p>In addition, the <code>BiocManager::available()</code> function
returns the complete list of package names that are can be installed
from the Bioconductor and CRAN repositories. For instance the total
number of numbers that could be installed using <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em></p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">available</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 26633</code></pre>
</div>
<p>Specifically, the union of current Bioconductor repositories and
other repositories on the search path can be displayed as follows.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">repositories</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span>                                                <span class="va">BioCsoft</span></span>
<span>           <span class="st">"https://bioconductor.org/packages/3.22/bioc"</span></span>
<span>                                                 <span class="va">BioCann</span></span>
<span><span class="st">"https://bioconductor.org/packages/3.22/data/annotation"</span></span>
<span>                                                 <span class="va">BioCexp</span></span>
<span><span class="st">"https://bioconductor.org/packages/3.22/data/experiment"</span></span>
<span>                                           <span class="va">BioCworkflows</span></span>
<span>      <span class="st">"https://bioconductor.org/packages/3.22/workflows"</span></span>
<span>                                               <span class="va">BioCbooks</span></span>
<span>          <span class="st">"https://bioconductor.org/packages/3.22/books"</span></span>
<span>                                             <span class="va">carpentries</span></span>
<span>                    <span class="st">"https://carpentries.r-universe.dev"</span></span>
<span>                                     <span class="va">carpentries_archive</span></span>
<span>                    <span class="st">"https://carpentries.github.io/drat"</span></span>
<span>                                                    <span class="va">CRAN</span></span>
<span>                              <span class="st">"https://cran.rstudio.com"</span> </span></code></pre>
</div>
<p>Each repository URL can be accessed in a web browser, displaying the
full list of packages available from that repository. For instance,
navigate to <a href="https://bioconductor.org/packages/3.14/bioc" class="external-link">https://bioconductor.org/packages/3.14/bioc</a>.</p>
<div id="going-further-4" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="going-further-4" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>The function <code>BiocManager::repositories()</code> can be combined
with the base function <code>available.packages()</code> to query
packages available specifically from any package repository, e.g. the
Bioconductor <a href="reference.html#software-package">software
package</a> repository.</p>
<pre><code>&gt; db = available.packages(repos = BiocManager::repositories()["BioCsoft"])
&gt; dim(db)
[1] 1948   17
&gt; head(rownames(db))
[1] "a4"          "a4Base"      "a4Classif"   "a4Core"      "a4Preproc"
[6] "a4Reporting"</code></pre>
</div>
</div>
</div>
<p>Conveniently, <code>BiocManager::available()</code> includes a
<code>pattern=</code> argument, particularly useful to navigate
annotation resources (the original use case motivating it). For
instance, a range of <a href="reference.html#annotationdata-package">Annotation data
packages</a> available for the mouse model organism can be listed as
follows.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">available</span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"*Mmusculus"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "BSgenome.Mmusculus.UCSC.mm10"        "BSgenome.Mmusculus.UCSC.mm10.masked"
 [3] "BSgenome.Mmusculus.UCSC.mm39"        "BSgenome.Mmusculus.UCSC.mm8"
 [5] "BSgenome.Mmusculus.UCSC.mm8.masked"  "BSgenome.Mmusculus.UCSC.mm9"
 [7] "BSgenome.Mmusculus.UCSC.mm9.masked"  "EnsDb.Mmusculus.v75"
 [9] "EnsDb.Mmusculus.v79"                 "PWMEnrich.Mmusculus.background"
[11] "TxDb.Mmusculus.UCSC.mm10.ensGene"    "TxDb.Mmusculus.UCSC.mm10.knownGene"
[13] "TxDb.Mmusculus.UCSC.mm39.knownGene"  "TxDb.Mmusculus.UCSC.mm39.refGene"
[15] "TxDb.Mmusculus.UCSC.mm9.knownGene"  </code></pre>
</div>
</section><section><h2 class="section-heading" id="installing-packages">Installing packages<a class="anchor" aria-label="anchor" href="#installing-packages"></a></h2>
<hr class="half-width"><p>The <code>BiocManager::install()</code> function is used to install
or update packages.</p>
<p>The function takes a character vector of package names, and attempts
to install them from the Bioconductor repository.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"S4Vectors"</span>, <span class="st">"BiocGenerics"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>However, if any package cannot be found in the Bioconductor
repository, the function also searches for those packages in
repositories listed in the global option <code>repos</code>.</p>
<div id="contribute" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="contribute" class="callout-inner">
<h3 class="callout-title">Contribute !</h3>
<div class="callout-content">
<p>Add an example of non-Bioconductor package that can be installed
using BioManager. Preferably, a package that will be used later in this
lesson.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="uninstalling-packages">Uninstalling packages<a class="anchor" aria-label="anchor" href="#uninstalling-packages"></a></h2>
<hr class="half-width"><p>Bioconductor packages can be removed from the R library like any
other R package, using the base R function
<code>remove.packages()</code>. In essence, this function simply removes
installed packages and updates index information as necessary. As a
result, it will not be possible to attach the package to a session or
browse the documentation of that package anymore.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remove.packages</span><span class="op">(</span><span class="st">"S4Vectors"</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="getting-help">Getting help<a class="anchor" aria-label="anchor" href="#getting-help"></a></h2>
<hr class="half-width"><p>For time we will not go over the full episode on <a href="https://carpentries-incubator.github.io/bioc-project/04-getting-help.html" class="external-link">Getting
Help</a>, instead below is a some summary of key points and sources:</p>
<p><strong>The Bioconductor website</strong></p>
<p>The main [Bioconductor website][bioc-website] provides a host of
resources, all freely available without even the need to install R or
any Bioconductor package.</p>
<p>In particular, the <a href="https://www.bioconductor.org/packages/release/BiocViews.html" class="external-link">biocViews</a>
page is a great way to thematically explore the collection of packages
and identify Bioconductor packages providing a certain
functionality.</p>
<p>Furthermore, the website also collects materials from [courses and
conferences][bioc-courses-conferences-materials], including
presentations, video recordings, and teaching materials.</p>
<p><strong>Package landing pages</strong></p>
<p>Each package accepted in the Bioconductor project is granted a
landing page on the main Bioconductor website, e.g. <em><a href="https://bioconductor.org/packages/3.22/S4Vectors" class="external-link">S4Vectors</a></em>.</p>
<p>Package landing pages contains useful information that can be
consulted without the need to install the package itself. This is
particularly useful while browsing the Bioconductor repository is search
of packages suitable for a specific task.</p>
<p><strong>Package vignettes</strong></p>
<p>When Bioconductor packages are installed in the user’s R library, the
vignettes associated with that particular version of the package are
also installed on the user’s computer. Those locally installed vignettes
are the gold standard reference for the version of the package that is
currently installed in the R library and used in the R session. They can
be accessed using the function <code>browseVignettes()</code>, for
instance:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">browseVignettes</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre>
</div>
<p><strong>Package help pages</strong></p>
<p>Bioconductor requires every user-facing package function to be
documented in one of the package help pages (often referred to as “man
pages”, after the name of the package sub-directory where they are
stored).</p>
<p>Help pages can be accessed using the <code>help()</code> function or
the question mark symbol <code>?</code>.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">help</span><span class="op">(</span>topic <span class="op">=</span> <span class="st">"install"</span>, package <span class="op">=</span> <span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">?</span><span class="fu">BiocManager</span><span class="fu">::</span><span class="va">install</span></span></code></pre>
</div>
<p><strong>The Bioconductor support site</strong></p>
<p>The <a href="https://support.bioconductor.org/" class="external-link">Bioconductor support
site</a> provides a platform for the community of users and developers
to ask questions, and help each other through the doubts and challenges
related to Bioconductor packages and analytical workflows.</p>
<p><strong>Workflow packages</strong></p>
<p>Bioconductor workflow packages are special in the way that they are
only expected to contain vignettes, without any additional code or
functionality of their own. Instead, the vignettes of workflow packages
exclusively import functionality from other packages, and demonstrate
how to combine functions from those packages into an integrated workflow
that users are likely to face in their day-to-day work.</p>
<p>Workflow packages can be browsed in a dedicated section of the
[biocViews page][biocviews-workflow].</p>
<p><strong>Slack workspace</strong></p>
<p>The <a href="https://bioc-community.herokuapp.com/" class="external-link">Bioconductor
Slack workspace</a> was created in 2016.</p>
<p>The workspace can be freely joined using this [Heroku
app][bioc-slack-heroku] to generate invitations for individual email
addresses.</p>
<div class="section level3">
<h3 id="how-to-efficiently-ask-for-help">How to efficiently ask for help<a class="anchor" aria-label="anchor" href="#how-to-efficiently-ask-for-help"></a></h3>
<p>Most community members provide help voluntarily, on their own spare
time and without any form of compensation. As such, it is important to
ask questions as clearly as possible, providing relevant information, to
help those volunteers identify the source of the issue as rapidly as
possible, thus saving time for both the helper and the original
poster.</p>
<p>Depending on the question, some key information includes:</p>
<ul><li>Operating system</li>
<li>Version of R</li>
<li>Version of Bioconductor</li>
<li>Version of individual packages installed in the R library</li>
<li>Version of individual packages attached to the current session</li>
<li>Third-party software, libraries, and compilers installed on the
user’s system</li>
<li>Source of packages installed in the R library (e.g., Bioconductor,
CRAN, GitHub)</li>
<li>Code executed in the R session leading up to the issue</li>
<li>Global options active in the session (accessible using
<code>options()</code>)</li>
</ul></div>
</section><section><h2 class="section-heading" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
<hr class="half-width"><p>[1] R. A. Amezquita, A. T. L. Lun, E. Becht, et al. “Orchestrating
single-cell analysis with Bioconductor”. In: <em>Nat Methods</em> 17.2
(2020), pp. 137-145. ISSN: 1548-7105 (Electronic) 1548-7091 (Linking).
DOI: 10.1038/s41592-019-0654-x. <a href="https://www.ncbi.nlm.nih.gov/pubmed/31792435" class="external-link uri">https://www.ncbi.nlm.nih.gov/pubmed/31792435</a>.</p>
<p>[2] R. C. Gentleman, V. J. Carey, D. M. Bates, et al. “Bioconductor:
open software development for computational biology and bioinformatics”.
In: <em>Genome Biol</em> 5.10 (2004), p. R80. ISSN: 1474-760X
(Electronic) 1474-7596 (Linking). DOI: 10.1186/gb-2004-5-10-r80. <a href="https://www.ncbi.nlm.nih.gov/pubmed/15461798" class="external-link uri">https://www.ncbi.nlm.nih.gov/pubmed/15461798</a>.</p>
<p>[3] W. Huber, V. J. Carey, R. Gentleman, et al. “Orchestrating
high-throughput genomic analysis with Bioconductor”. In: <em>Nat
Methods</em> 12.2 (2015), pp. 115-21. ISSN: 1548-7105 (Electronic)
1548-7091 (Linking). DOI: 10.1038/nmeth.3252. <a href="https://www.ncbi.nlm.nih.gov/pubmed/25633503" class="external-link uri">https://www.ncbi.nlm.nih.gov/pubmed/25633503</a>.</p>
<p>[4] S. Su, V. Carey, L. Shepherd, et al. “BiocPkgTools: Toolkit for
mining the Bioconductor package ecosystem [version 1; peer review: 2
approved, 1 approved with reservations]”. In: <em>F1000Research</em>
8.752 (2019). DOI: 10.12688/f1000research.19410.1.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>R packages are but one aspect of the Bioconductor project.</li>
<li>The Bioconductor project extends and complements the CRAN
repository.</li>
<li>Different types of packages provide not only software, but also
annotations, experimental data, and demonstrate the use of multiple
packages in integrated workflows.</li>
<li>Interoperability beteen Bioconductor packages facilitates the
writing of integrated workflows and minimizes the cognitive burden on
users.</li>
<li>Educational materials from courses and conferences are archived and
accessible on the Bioconductor website and YouTube channel.</li>
<li>Different channels of communication enable community members to
converse and help each other, both as users and package developers.</li>
<li>The Bioconductor project is governed by scientific, technical, and
advisory boards, as well as a Code of Conduct committee.</li>
<li>The BiocManager package is available from the CRAN repository.</li>
<li>
<code>BiocManager::install()</code> is used to install and update
Bioconductor packages (but also from CRAN and GitHub).</li>
<li>
<code>BiocManager::valid()</code> is used to check for available
package updates.</li>
<li>
<code>BiocManager::version()</code> reports the version of
Bioconductor currently installed.</li>
<li>
<code>BiocManager::install()</code> can also be used to update an
entire R library to a specific version of Bioconductor.</li>
<li>The <code>browseVignettes()</code> function is recommended to access
the vignette(s) installed with each package.</li>
<li>The Bioconductor main website contains general information, package
documentation, and course materials.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/01-introduction.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/03-biological-sequences.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/01-introduction.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction
        </a>
        <a class="chapter-link float-end" href="../instructor/03-biological-sequences.html" rel="next">
          Next: Biological sequences...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/episodes/02-bioconductor.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/" class="external-link">Source</a></p>
				<p><a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:siobhon.egan@murdoch.edu.au">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://siobhon-egan.github.io/2025-11-13-COMBINE-WA/instructor/02-bioconductor.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Bioconductor Introduction &amp;amp; Install",
  "creativeWorkStatus": "active",
  "url": "https://siobhon-egan.github.io/2025-11-13-COMBINE-WA/instructor/02-bioconductor.html",
  "identifier": "https://siobhon-egan.github.io/2025-11-13-COMBINE-WA/instructor/02-bioconductor.html",
  "dateCreated": "2025-11-13",
  "dateModified": "2025-11-11",
  "datePublished": "2025-11-11"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

