<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Intermediate R - biological data: Reproducible analysis</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png"><link rel="manifest" href="../favicons/incubator/site.webmanifest"><link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../11-reproducible-analysis.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Intermediate R - biological data
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Intermediate R - biological data
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Intermediate R - biological data
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 91%" class="percentage">
    91%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 91%" aria-valuenow="91" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../11-reproducible-analysis.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-setup.html">1. Introduction and setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-introduction-to-bioconductor.html">2. Introduction to Bioconductor</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-installing-bioconductor.html">3. Installing Bioconductor packages</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-getting-help.html">4. Getting help</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-s4.html">5. S4 classes in Bioconductor</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-biological-sequences.html">6. Working with biological sequences</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-genomic-ranges.html">7. Working with genomics ranges</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-annotations.html">8. Working with annotations</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-summarizedexperiment.html">9. The SummarizedExperiment class</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-application-microbiome.html">10. Microbiome application</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        11. Reproducible analysis
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#load-packages">Load packages</a></li>
<li><a href="#reproducible-data">Reproducible data</a></li>
<li><a href="#making-reproducible-and-easy-to-share-reports">Making reproducible and easy to share reports</a></li>
<li><a href="#session-info">Session info</a></li>
<li><a href="#recommendations">Recommendations</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/10-application-microbiome.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/10-application-microbiome.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Microbiome
        </a>
        <a class="chapter-link float-end" href="../instructor/index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Reproducible analysis</h1>
        <p>Last updated on 2025-11-10 |

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/episodes/11-reproducible-analysis.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 10 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<p>In this exercise we will cover some aspects relevant to generating
and most importantly <strong>re-generating</strong> analysis.</p>
<p>We will use the libraries and data we covered in the previous episode
<a href="https://siobhonlegan.com/2025-11-13-COMBINE-WA/10-application-microbiome.html" class="external-link">Application
for microbiome data</a>.</p>
<section><h2 class="section-heading" id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a></h2>
<hr class="half-width"><p>Now load in the required packages for this analysis</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">microbiome</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">knitr</span><span class="op">)</span></span>
<span><span class="co"># load the peerj32 dataset</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="va">peerj32</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="reproducible-data">Reproducible data<a class="anchor" aria-label="anchor" href="#reproducible-data"></a></h2>
<hr class="half-width"><p>Let’s say you have spent a long time getting your data formatted and
read into R.</p>
<p>Now you will want to do some visualizations on this data and it is
likely this will need to be repeated (many) times in the future.</p>
<p>You can save your data in and <code>.rda</code> or <code>.rds</code>
format so that it will allow you to quickly read it back into the work
space and pick up on more figures. It will also made it easy to read in
and generate reports in a reproducible way (which we will do
shortly).</p>
<p>Have a go at saving the <code>peerj32</code> object that is in your
environment to your local directory. If you have followed the previous
episodes in this lesson you should already have a <strong>data/</strong>
directory.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">save</span><span class="op">(</span><span class="va">peerj32</span>, file <span class="op">=</span> <span class="st">"data/peerj32.rda"</span><span class="op">)</span></span></code></pre>
</div>
<p>You can quickly load the <code>.rda</code> file back in</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">load</span><span class="op">(</span><span class="st">"data/peerj32.rda"</span><span class="op">)</span></span></code></pre>
</div>
<p>….</p>
</section><section><h2 class="section-heading" id="making-reproducible-and-easy-to-share-reports">Making reproducible and easy to share reports<a class="anchor" aria-label="anchor" href="#making-reproducible-and-easy-to-share-reports"></a></h2>
<hr class="half-width"><p>Next I have prepared a <code>.qmd</code> file that you will download
and then “knit” to reproducible a <code>.html</code> file.</p>
<div id="the-situation---sharing-results" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="the-situation---sharing-results" class="callout-inner">
<h3 class="callout-title">The Situation - Sharing Results</h3>
<div class="callout-content">
<p>Imagine you need to give your supervisors a summary of your findings
early on, instead of spending weeks neatly formatting and exporting
plots one by one in R you may like to share this html file early in your
analysis to get some general feedback and direction before find tuning
your analysis.</p>
</div>
</div>
</div>
<p>It is best practice to not re-install the packages each time when
making these reports, so ensure you have the required packages installed
prior to running.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install your CRAN packages for the qmd</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"gtsummary"</span><span class="op">)</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"reshape2"</span><span class="op">)</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"knitr"</span><span class="op">)</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"DT"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## install your Bioconductor packages</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw">require</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"microbiome"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"phyloseq"</span><span class="op">)</span></span></code></pre>
</div>
<p>Download the <code>.qmd</code> file</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span></span>
<span>  url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/siobhon-egan/2025-11-13-COMBINE-WA/refs/heads/main/episodes/files/microbiome-data-report.qmd"</span>,</span>
<span>  destfile <span class="op">=</span> <span class="st">"data/microbiome-report.qmd"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>Have a go at rendering the <code>.qmd</code> file.</p>
<div id="challenge" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Practical opening and “knitting” the <code>.qmd</code> file to html
output.</p>
</div>
</div>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: `includeHTML()` was provided a `path` that appears to be a complete HTML document.
✖ Path: files/microbiome-data-report.html
ℹ Use `tags$iframe()` to include an HTML document. You can either ensure `path` is accessible in your app or document (see e.g. `shiny::addResourcePath()`) and pass the relative path to the `src` argument. Or you can read the contents of `path` and pass the contents to `srcdoc`.</code></pre>
</div>
<!--html_preserve-->
&lt;!DOCTYPE html&gt;


<meta charset="utf-8"><meta name="generator" content="quarto-1.7.23"><meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes"><meta name="author" content="Siobhon Egan"><meta name="dcterms.date" content="2025-11-13"><title>
Microbiome data report for Supervisor meeting
</title><style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style><script src="microbiome-data-report_files/libs/clipboard/clipboard.min.js"></script><script src="microbiome-data-report_files/libs/quarto-html/quarto.js" type="module"></script><script src="microbiome-data-report_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script><script src="microbiome-data-report_files/libs/quarto-html/popper.min.js"></script><script src="microbiome-data-report_files/libs/quarto-html/tippy.umd.min.js"></script><script src="microbiome-data-report_files/libs/quarto-html/anchor.min.js"></script><link href="microbiome-data-report_files/libs/quarto-html/tippy.css" rel="stylesheet"><link href="microbiome-data-report_files/libs/quarto-html/quarto-syntax-highlighting-226bd0f977fa82dfae4534cac220d79a.css" rel="stylesheet" id="quarto-text-highlighting-styles"><script src="microbiome-data-report_files/libs/bootstrap/bootstrap.min.js"></script><p><link href="microbiome-data-report_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet"><link href="microbiome-data-report_files/libs/bootstrap/bootstrap-be9206a58a27969b3ea3001da8f26fef.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light"></p>


<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
Microbiome data report for Supervisor meeting
</h1>
</div>
<div class="quarto-title-meta">
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">SH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode sh" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="op">&lt;</span>div<span class="op">&gt;</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="op">&lt;</span>div <span class="va">class</span><span class="op">=</span><span class="st">"quarto-title-meta-heading"</span><span class="op">&gt;</span>Author<span class="op">&lt;</span>/div<span class="op">&gt;</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="op">&lt;</span>div <span class="va">class</span><span class="op">=</span><span class="st">"quarto-title-meta-contents"</span><span class="op">&gt;</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>         <span class="op">&lt;</span>p<span class="op">&gt;</span>Siobhon <span class="ex">Egan</span> <span class="op">&lt;</span>/p<span class="op">&gt;</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>      <span class="op">&lt;</span>/div<span class="op">&gt;</span></span></code></pre>
</div>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">SH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode sh" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="op">&lt;</span>div<span class="op">&gt;</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="op">&lt;</span>div <span class="va">class</span><span class="op">=</span><span class="st">"quarto-title-meta-heading"</span><span class="op">&gt;</span>Published<span class="op">&lt;</span>/div<span class="op">&gt;</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="op">&lt;</span>div <span class="va">class</span><span class="op">=</span><span class="st">"quarto-title-meta-contents"</span><span class="op">&gt;</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>  <span class="op">&lt;</span>p <span class="va">class</span><span class="op">=</span><span class="st">"date"</span><span class="op">&gt;</span>November <span class="ex">13,</span> <span class="dv">2025</span><span class="op">&lt;</span>/p<span class="op">&gt;</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="op">&lt;</span>/div<span class="op">&gt;</span></span></code></pre>
</div>
</header></main></div>
</section><p>
The following datasets and analysis have been used from tutorials
available in the <a href="">microbiome package</a>, for
<a href="https://microbiome.github.io/tutorials/Heatmap.html" class="external-link">heatmaps</a>
</p>
<section class="level2"><h2 class="anchored" data-anchor-id="information" id="information">
Information
<a class="anchor" aria-label="anchor" href="#information"></a></h2>
<p>
Prepare a summary table of the participants in your dataset
</p>
<div class="cell">
<details class="code-fold"><summary>
Show the code
</summary><div id="cb1" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span><span class="op">@</span><span class="va">sam_data</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">group</span>, <span class="va">sex</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 × 4
# Groups:   group, sex [4]
  group   sex     time     n
  &lt;fct&gt;   &lt;fct&gt;  &lt;int&gt; &lt;int&gt;
1 LGG     female     1     5
2 LGG     female     2     5
3 LGG     male       1     3
4 LGG     male       2     3
5 Placebo female     1    10
6 Placebo female     2    10
7 Placebo male       1     4
8 Placebo male       2     4</code></pre>
</div>
<details class="code-fold"><summary>
Show the code
</summary><div id="cb3" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span><span class="op">@</span><span class="va">sam_data</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">group</span>, <span class="va">sex</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 3
# Groups:   group [2]
  group   sex        n
  &lt;fct&gt;   &lt;fct&gt;  &lt;int&gt;
1 LGG     female    10
2 LGG     male       6
3 Placebo female    20
4 Placebo male       8</code></pre>
</div>
</div>
</section><section class="level2"><h2 class="anchored" data-anchor-id="alpha-diversity" id="alpha-diversity">
Alpha diversity
<a class="anchor" aria-label="anchor" href="#alpha-diversity"></a></h2>
</section><section class="level2"><h2 class="anchored" data-anchor-id="beta-diversity" id="beta-diversity">
Beta diversity
<a class="anchor" aria-label="anchor" href="#beta-diversity"></a></h2>
</section><section class="level2"><h2 class="anchored" data-anchor-id="heatmap-of-abundance" id="heatmap-of-abundance">
Heatmap of abundance
<a class="anchor" aria-label="anchor" href="#heatmap-of-abundance"></a></h2>
<p>
Abundance of microbial taxa (using the Z transformed data)
</p>
<div class="cell">
<details class="code-fold"><summary>
Show the code
</summary><div id="cb5" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the abundances heatmap, round values to 2 decimals</span></span>
<span><span class="va">dfm</span> <span class="op">&lt;-</span> <span class="fu">melt</span><span class="op">(</span><span class="fu">round</span><span class="op">(</span><span class="fu">abundances</span><span class="op">(</span><span class="va">pseqz</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">dfm</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Taxa"</span>, <span class="st">"Sample"</span>, <span class="st">"value"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">heat</span><span class="op">(</span><span class="va">dfm</span>, <span class="st">"Taxa"</span>, <span class="st">"Sample"</span>, <span class="st">"value"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>text<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>size<span class="op">=</span><span class="fl">10</span><span class="op">)</span>, </span>
<span>        axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>        legend.key.size <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fl">1.2</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output-display">
<div>
<figure class="figure"><p>
<img src="../microbiome-data-report_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img figure" width="672"></p>
</figure></div>
</div>
</div>
</section><section class="level2"><h2 class="anchored" data-anchor-id="correlation-between-microbes-and-lipids" id="correlation-between-microbes-and-lipids">
Correlation between microbes and lipids
<a class="anchor" aria-label="anchor" href="#correlation-between-microbes-and-lipids"></a></h2>
<p>
Cross-correlate columns of two data sets from related to microbiome and
blood serum lipids associations
(<a href="https://peerj.com/articles/32/" class="external-link">PeerJ 1:e32</a>).
</p>
<p>
The function returns correlations, raw p-values, and fdr estimates (not
strictly proper as the comparisons are not independent). Keep only those
elements that have at least only one significant correlation (n.signif):
</p>
<div class="cell">
<details class="code-fold"><summary>
Show the code
</summary><div id="cb6" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load example data </span></span>
<span><span class="va">otu</span> <span class="op">&lt;-</span> <span class="va">peerj32</span><span class="op">$</span><span class="va">microbes</span> </span>
<span><span class="va">lipids</span> <span class="op">&lt;-</span> <span class="va">peerj32</span><span class="op">$</span><span class="va">lipids</span> </span>
<span></span>
<span><span class="co"># Define data sets to cross-correlate</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">log10</span><span class="op">(</span><span class="va">otu</span><span class="op">)</span> <span class="co"># OTU Log10 (44 samples x 130 genera)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">lipids</span><span class="op">)</span> <span class="co"># Lipids (44 samples x 389 lipids)</span></span>
<span></span>
<span><span class="co"># Cross correlate data sets</span></span>
<span><span class="va">correlations</span> <span class="op">&lt;-</span> <span class="fu">associate</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, method <span class="op">=</span> <span class="st">"spearman"</span>, mode <span class="op">=</span> <span class="st">"matrix"</span>, p.adj.threshold <span class="op">=</span> <span class="fl">0.05</span>, n.signif <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Or, alternatively, the same output is also available in a handy table format</span></span>
<span><span class="va">correlation.table</span> <span class="op">&lt;-</span> <span class="fu">associate</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, method <span class="op">=</span> <span class="st">"spearman"</span>, mode <span class="op">=</span> <span class="st">"table"</span>, p.adj.threshold <span class="op">=</span> <span class="fl">0.05</span>, n.signif <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="fu">head</span><span class="op">(</span><span class="va">correlation.table</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output-display">
<table class="table caption-top table table-sm table-striped small"><colgroup><col style="width: 7%"><col style="width: 45%"><col style="width: 15%"><col style="width: 17%"><col style="width: 14%"></colgroup><thead><tr class="header"><th style="text-align: left;">
</th>
<th style="text-align: left;">
X1
</th>
<th style="text-align: left;">
X2
</th>
<th style="text-align: right;">
Correlation
</th>
<th style="text-align: right;">
p.adj
</th>
</tr></thead><tbody><tr class="odd"><td style="text-align: left;">
1648
</td>
<td style="text-align: left;">
Ruminococcus gnavus et rel.
</td>
<td style="text-align: left;">
TG(54:5).2
</td>
<td style="text-align: right;">
0.7164958
</td>
<td style="text-align: right;">
0.0022842
</td>
</tr><tr class="even"><td style="text-align: left;">
384
</td>
<td style="text-align: left;">
Moraxellaceae
</td>
<td style="text-align: left;">
PC(40:3e)
</td>
<td style="text-align: right;">
-0.6941863
</td>
<td style="text-align: right;">
0.0029225
</td>
</tr><tr class="odd"><td style="text-align: left;">
1829
</td>
<td style="text-align: left;">
Uncultured Bacteroidetes
</td>
<td style="text-align: left;">
TG(56:2).1
</td>
<td style="text-align: right;">
-0.6987375
</td>
<td style="text-align: right;">
0.0029225
</td>
</tr><tr class="even"><td style="text-align: left;">
349
</td>
<td style="text-align: left;">
Lactobacillus plantarum et rel.
</td>
<td style="text-align: left;">
PC(40:3)
</td>
<td style="text-align: right;">
-0.6877976
</td>
<td style="text-align: right;">
0.0031520
</td>
</tr><tr class="odd"><td style="text-align: left;">
1198
</td>
<td style="text-align: left;">
Ruminococcus gnavus et rel.
</td>
<td style="text-align: left;">
TG(52:5)
</td>
<td style="text-align: right;">
0.6806216
</td>
<td style="text-align: right;">
0.0037518
</td>
</tr><tr class="even"><td style="text-align: left;">
264
</td>
<td style="text-align: left;">
Moraxellaceae
</td>
<td style="text-align: left;">
PC(38:4).1
</td>
<td style="text-align: right;">
-0.6700504
</td>
<td style="text-align: right;">
0.0038414
</td>
</tr></tbody></table></div>
</div>
<section class="level3"><h3 class="anchored" data-anchor-id="association-heatmaps" id="association-heatmaps">
Association heatmaps
<a class="anchor" aria-label="anchor" href="#association-heatmaps"></a></h3>
<p>
Rearrange the data and plot the heatmap and mark significant
correlations with stars to reproduce microbiota-lipidome heatmap from
<a href="https://peerj.com/articles/32/" class="external-link">Lahti et al. PeerJ (2013)</a>
(the ordering of rows and columns may be different):
</p>
<div class="cell">
<details class="code-fold"><summary>
Show the code
</summary><div id="cb7" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">heat</span><span class="op">(</span><span class="va">correlation.table</span>, <span class="st">"X1"</span>, <span class="st">"X2"</span>, </span>
<span>          fill <span class="op">=</span> <span class="st">"Correlation"</span>, </span>
<span>          star <span class="op">=</span> <span class="st">"p.adj"</span>, </span>
<span>          p.adj.threshold <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> </span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>text<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>size<span class="op">=</span><span class="fl">10</span><span class="op">)</span>, </span>
<span>          axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, </span>
<span>          legend.key.size <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fl">1.3</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output-display">
<div>
<figure class="figure"><p>
<img src="../microbiome-data-report_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img figure" width="672"></p>
</figure></div>
</div>
</div>
</section><section class="level3"><h3 class="anchored" data-anchor-id="heatmap-of-significant-correlations" id="heatmap-of-significant-correlations">
Heatmap of significant correlations
<a class="anchor" aria-label="anchor" href="#heatmap-of-significant-correlations"></a></h3>
<p>
This time we are modifying the plot to only include significant
correlations with a threshold of P &lt; 0.05. The cells marked with a
white cross indicate most significant correlations with a significance
of P &lt; 0.02.
</p>
<div class="cell">
<details class="code-fold"><summary>
Show the code
</summary><div id="cb8" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Order the rows and columns with levels argument if needed:</span></span>
<span><span class="va">correlation.table</span><span class="op">$</span><span class="va">X1</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">correlation.table</span><span class="op">$</span><span class="va">X1</span>, levels <span class="op">=</span> <span class="fu">unique</span><span class="op">(</span><span class="fu">as.character</span><span class="op">(</span><span class="va">correlation.table</span><span class="op">$</span><span class="va">X1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">correlation.table</span><span class="op">$</span><span class="va">X2</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">correlation.table</span><span class="op">$</span><span class="va">X2</span>, levels <span class="op">=</span> <span class="fu">unique</span><span class="op">(</span><span class="fu">as.character</span><span class="op">(</span><span class="va">correlation.table</span><span class="op">$</span><span class="va">X2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set black-and-white theme</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="fu">theme_set</span><span class="op">(</span><span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Pick only the correlations with p.adj&lt;0.05</span></span>
<span><span class="co"># Note: this will leave other cells empty</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="va">subtable</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">correlation.table</span>, <span class="va">p.adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Arrange the figure</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">subtable</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">X1</span>, y <span class="op">=</span> <span class="va">X2</span>, fill <span class="op">=</span> <span class="va">Correlation</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">geom_tile</span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">scale_fill_gradientn</span><span class="op">(</span><span class="st">"Correlation"</span>, </span>
<span>                       breaks <span class="op">=</span> <span class="fu">seq</span><span class="op">(</span>from <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, to <span class="op">=</span> <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>, </span>
<span>                   colours <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"darkblue"</span>, <span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>, <span class="st">"darkred"</span><span class="op">)</span>, </span>
<span>                   limits <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Polish texts</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>axis.text.x<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust<span class="op">=</span><span class="fl">1</span>, face <span class="op">=</span> <span class="st">"italic"</span><span class="op">)</span>,</span>
<span>               axis.text.y<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">xlab</span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">""</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Mark the most significant cells with stars using threshold of P &lt; 0.02</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">geom_text</span><span class="op">(</span>data <span class="op">=</span> <span class="fu">subset</span><span class="op">(</span><span class="va">correlation.table</span>, <span class="va">p.adj</span> <span class="op">&lt;</span> <span class="fl">0.02</span><span class="op">)</span>, </span>
<span>               <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">X1</span>, y <span class="op">=</span> <span class="va">X2</span>, label <span class="op">=</span> <span class="st">"+"</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"white"</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output-display">
<div>
<figure class="figure"><p>
<img src="../microbiome-data-report_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img figure" width="672"></p>
</figure></div>
</div>
</div>
<div class="cell">
<details class="code-fold"><summary>
Show the code
</summary><div id="cb9" class="codewrapper sourceCode">
<h3 class="code-label">R CODE-WITH-COPY<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
</details><div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.1 (2025-06-13)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Perth
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] gtsummary_2.4.0   knitr_1.50        reshape2_1.4.4    dplyr_1.1.4
[5] microbiome_1.30.0 ggplot2_4.0.0     phyloseq_1.52.0

loaded via a namespace (and not attached):
 [1] gtable_0.3.6            xfun_0.53               htmlwidgets_1.6.4
 [4] rhdf5_2.52.1            Biobase_2.68.0          lattice_0.22-7
 [7] rhdf5filters_1.20.0     vctrs_0.6.5             tools_4.5.1
[10] generics_0.1.4          biomformat_1.36.0       stats4_4.5.1
[13] parallel_4.5.1          tibble_3.3.0            cluster_2.1.8.1
[16] pkgconfig_2.0.3         Matrix_1.7-3            data.table_1.17.8
[19] RColorBrewer_1.1-3      S7_0.2.0                S4Vectors_0.46.0
[22] lifecycle_1.0.4         GenomeInfoDbData_1.2.14 compiler_4.5.1
[25] farver_2.1.2            stringr_1.5.2           Biostrings_2.76.0
[28] codetools_0.2-20        permute_0.9-8           GenomeInfoDb_1.44.3
[31] htmltools_0.5.8.1       yaml_2.3.10             tidyr_1.3.1
[34] pillar_1.11.1           crayon_1.5.3            MASS_7.3-65
[37] vegan_2.7-2             iterators_1.0.14        foreach_1.5.2
[40] nlme_3.1-168            tidyselect_1.2.1        digest_0.6.37
[43] Rtsne_0.17              stringi_1.8.7           purrr_1.1.0
[46] splines_4.5.1           ade4_1.7-23             fastmap_1.2.0
[49] grid_4.5.1              cli_3.6.5               magrittr_2.0.4
[52] utf8_1.2.6              survival_3.8-3          ape_5.8-1
[55] withr_3.0.2             UCSC.utils_1.4.0        scales_1.4.0
[58] rmarkdown_2.30          XVector_0.48.0          httr_1.4.7
[61] igraph_2.2.0            multtest_2.64.0         evaluate_1.0.5
[64] IRanges_2.42.0          mgcv_1.9-3              rlang_1.1.6
[67] Rcpp_1.1.0              glue_1.8.0              BiocGenerics_0.54.1
[70] rstudioapi_0.17.1       jsonlite_2.0.0          R6_2.6.1
[73] Rhdf5lib_1.30.0         plyr_1.8.9             </code></pre>
</div>
</div>
</section></section><!-- /main column --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button,
          { trigger: "manual",
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config);
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined;
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              }
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            }
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><!-- /content --><!--/html_preserve--><section><h2 class="section-heading" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a></h2>
<hr class="half-width"><p>When sharing R Markdown (.Rmd) or Quarto (.qmd) notebooks, it’s good
practice to include information about your R session at the end of the
document. This helps others (and future you!) understand:</p>
<ul><li>Which version of R was used</li>
<li>What packages were loaded</li>
<li>What platform the code was run on</li>
</ul><div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.5.2 (2025-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] knitr_1.50        microbiome_1.32.0 ggplot2_4.0.0     phyloseq_1.54.0

loaded via a namespace (and not attached):
 [1] gtable_0.3.6        xfun_0.54           rhdf5_2.54.0
 [4] Biobase_2.70.0      lattice_0.22-5      rhdf5filters_1.22.0
 [7] vctrs_0.6.5         tools_4.5.2         generics_0.1.4
[10] biomformat_1.38.0   stats4_4.5.2        parallel_4.5.2
[13] tibble_3.3.0        cluster_2.1.8.1     pkgconfig_2.0.3
[16] Matrix_1.7-4        data.table_1.17.8   RColorBrewer_1.1-3
[19] S7_0.2.0            S4Vectors_0.48.0    lifecycle_1.0.4
[22] compiler_4.5.2      farver_2.1.2        stringr_1.6.0
[25] Biostrings_2.78.0   Seqinfo_1.0.0       codetools_0.2-19
[28] permute_0.9-8       htmltools_0.5.8.1   yaml_2.3.10
[31] pillar_1.11.1       crayon_1.5.3        tidyr_1.3.1
[34] MASS_7.3-65         vegan_2.7-2         iterators_1.0.14
[37] foreach_1.5.2       nlme_3.1-168        tidyselect_1.2.1
[40] digest_0.6.37       Rtsne_0.17          stringi_1.8.7
[43] purrr_1.2.0         dplyr_1.1.4         reshape2_1.4.4
[46] splines_4.5.2       fastmap_1.2.0       ade4_1.7-23
[49] grid_4.5.2          cli_3.6.5           magrittr_2.0.4
[52] survival_3.8-3      ape_5.8-1           withr_3.0.2
[55] scales_1.4.0        XVector_0.50.0      igraph_2.2.1
[58] multtest_2.66.0     evaluate_1.0.5      IRanges_2.44.0
[61] mgcv_1.9-1          rlang_1.1.6         Rcpp_1.1.0
[64] glue_1.8.0          BiocManager_1.30.26 renv_1.1.5
[67] BiocGenerics_0.56.0 jsonlite_2.0.0      R6_2.6.1
[70] Rhdf5lib_1.32.0     plyr_1.8.9         </code></pre>
</div>
<p>Some other similar handy functions include:</p>
<ul><li>
<code>devtools::session_info()</code> - more detailed, especially
for package dependencies, and identifies where packages were installed
from.</li>
<li>
<code>renv::snapshot()</code> – for managing project-specific
environments.</li>
</ul></section><section><h2 class="section-heading" id="recommendations">Recommendations<a class="anchor" aria-label="anchor" href="#recommendations"></a></h2>
<hr class="half-width"><p>Some recommendations for rendering your <code>.qmd</code>:</p>
<ul><li>Hide warnings and messages from your R chunks</li>
<li>Keeping your code in the output can be good to see what data was
used for analysis and if any filtering or transformations were used,
however it is not always neccessary and can make it difficult for other
“non-coders” to read. In the “YAML” at the top you will see the option
<code>code-fold: true</code> is used to allow us to toggle the code on
and off.</li>
<li>Include the <code>sessionInfo()</code> at the bottom of your file to
keep track of package versions used!</li>
</ul><hr><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>Saving your data in <code>.rda</code> format can make it easier to
load and repeat data analysis.</li>
<li>A notebook in <code>.rmd</code> or <code>.qmd</code> format and
knitted to html can be an efficient way to store and share data
workflows and analysis</li>
<li>Use <code>sessionInfo()</code> when using R to generate reports as a
handy way to track package versions used to generate analysis and
figures.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/10-application-microbiome.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/10-application-microbiome.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Microbiome
        </a>
        <a class="chapter-link float-end" href="../instructor/index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/edit/main/episodes/11-reproducible-analysis.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/" class="external-link">Source</a></p>
				<p><a href="https://github.com/siobhon-egan/2025-11-13-COMBINE-WA/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:siobhon.egan@murdoch.edu.au">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://siobhon-egan.github.io/2025-11-13-COMBINE-WA/instructor/11-reproducible-analysis.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Reproducible analysis",
  "creativeWorkStatus": "active",
  "url": "https://siobhon-egan.github.io/2025-11-13-COMBINE-WA/instructor/11-reproducible-analysis.html",
  "identifier": "https://siobhon-egan.github.io/2025-11-13-COMBINE-WA/instructor/11-reproducible-analysis.html",
  "dateCreated": "2025-11-13",
  "dateModified": "2025-11-10",
  "datePublished": "2025-11-10"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

