---
title: "Microbiome data report for Supervisor meeting"
author: "Siobhon Egan"
date: "2025-11-13"
execute:
  message: false
  warning: false
format:
  html:
    code-fold: true
    code-summary: "Show the code"
---

```{r setup}
#| message: false
#| warning: false
#| include: false
#Load packages and data
library(microbiome) # Load libraries
library(phyloseq)
library(dplyr)
library(reshape2)
library(knitr)
library(DT)
library(ggplot2)
library(ggpubr)
```

Load in the `peerj32` data from [Lahti et al. PeerJ 1:e32, 2013](https://doi.org/10.7717/peerj.32) made available by [*Microbiome*](https://microbiome.github.io/tutorials/Data.html) package. This characterizes associations between human intestinal microbiota and blood serum lipids. Note that this data set contains an additional data matrix of lipid species. 

```{r}
# Load in data - apply any sub-setting and transformations
data(peerj32)
pseq <- peerj32$phyloseq    # Rename data

# Transform data to compsoition instead of count
pseq.comp <- microbiome::transform(pseq, "compositional")

# Pick data subset (group LGG and taxa from Phylum Bacteroidetes)
pseq2 <- pseq %>%
         subset_taxa(Phylum == "Bacteroidetes") %>%
         subset_samples(group == "LGG")
# Z transformed abundance data
pseq2z <- microbiome::transform(pseq2, "Z")
```

```{r themes, include=FALSE}
theme_set(theme_bw())
```

## Information

Prepare a summary table of the participants in your dataset
```{r sampledata}
pseq@sam_data %>%
  group_by(group, sex, time) %>%
  summarise(n = n())
pseq@sam_data %>%
  group_by(group, sex) %>%
  summarise(n = n())
```

### Sequencing depth

This will give you an overview of the number of reads per sample and per OTU. Important to know the ‘depth’ of sequencing. Generally for amplicon 16S microbiome you want many 10’s of thousands of (good reads) per sample. The more complex the sample the more reads you need (but there is a very large variation in studies and not set rule).

```{r readsum}
readsumsdf <- data.frame(
  nreads = sort(taxa_sums(pseq), TRUE),
  sorted = 1:ntaxa(pseq),
  type = "OTUs"
)
readsumsdf <- rbind(readsumsdf,
  data.frame(
    nreads = sort(sample_sums(pseq), TRUE),
    sorted = 1:nsamples(pseq),
    type = "Samples"
  ))
title <- "Total number of reads"
nreads <- ggplot(readsumsdf, aes(x = sorted, y = nreads)) + geom_bar(stat = "identity")
nreads <- nreads + ggtitle(title) + scale_y_log10() + facet_wrap( ~ type, 1, scales = "free")
nreads
```


## Alpha diversity

Generate alpha diversity measures across all indexes

```{r alphadiv}
tab <- microbiome::alpha(pseq, index = "all")
kable(head(tab))
```
View raw results as a table
```{r alphadivDT, message=FALSE, echo=F}
# display the linelist data as a table
DT::datatable(
  head(tab, 50),
  rownames = FALSE,
  filter = "top",
  options = list(pageLength = 10, scrollX = T),
  class = 'white-space: nowrap'
)
```

Shannon alpha diversity measure by sex
```{r shannondiv}
p.shannon <- boxplot_alpha(pseq, 
                           index = "shannon",
                           x_var = "sex",
                           fill.colors = c(female="cyan4", male="deeppink4"))
# Improve the figure text
p.shannon <- p.shannon + theme_minimal() + 
  labs(x="Sex", y="Shannon diversity") +
  theme(axis.text = element_text(size=12),
        axis.title = element_text(size=16),
        legend.text = element_text(size=12),
        legend.title = element_text(size=16))
p.shannon
```


**Testing differences in alpha diversity** 

Using the non-parametric Kolmogorov-Smirnov test for two-group comparisons when there are no relevant covariates.

```{r alphadiv-stat}
# Construct the data
d <- meta(pseq)
d$diversity <- microbiome::diversity(pseq, "shannon")$shannon
# Split the values by group
spl <- split(d$diversity, d$sex)
# Kolmogorov-Smironv test
pv <- ks.test(spl$female, spl$male)$p.value
# Adjust the p-value
padj <- p.adjust(pv)
padj
```

Identified no difference so reproduce using the grouping variable and do not separate out by sex

Shannon alpha diversity measure by **group**
```{r shannondiv2}
# use a different name so we can merge the two figures back together to create a panel
p.shannon2 <- boxplot_alpha(pseq, 
                           index = "shannon",
                           x_var = "group",
                           fill.colors = c(Placebo="#4575b4", LGG="#d73027"))
# Improve the figure text
p.shannon2 <- p.shannon2 + theme_minimal() + 
  labs(x="Group", y="Shannon diversity") +
  theme(axis.text = element_text(size=12),
        axis.title = element_text(size=16),
        legend.text = element_text(size=12),
        legend.title = element_text(size=16))
p.shannon2
```


**Testing differences in alpha diversity** 

Using the non-parametric Kolmogorov-Smirnov test for two-group comparisons when there are no relevant covariates.

```{r alphadiv-stat2}
# Construct the data
d <- meta(pseq)
d$diversity <- microbiome::diversity(pseq, "shannon")$shannon
# Split the values by group
spl <- split(d$diversity, d$group)
# Kolmogorov-Smironv test
pv <- ks.test(spl$Placebo, spl$LGG)$p.value
# Adjust the p-value
padj <- p.adjust(pv)
padj
```

> Still no significant difference in alpha diversity (shannon index)

Make a panel figure for alpha diversity measuring showing no difference between sex or the grouping.
```{r alpha-merged}
#ggpubr package for formatting multi panel figures
mergealpha <- ggarrange(p.shannon, p.shannon2 + rremove("y.title"), # removes the duplicate y axis title
          labels = c("A", "B"),
          ncol = 2, nrow = 1)
mergealpha
ggsave(mergealpha, file = "alphadiv.png", width = 7, height = 3)
```


## Plot taxa prevalence

This function allows you to have an overview of OTU prevalences alongwith their taxonomic affiliations. This will aid in checking if you filter OTUs based on prevalence, then what taxonomic affliations will be lost.

```{r comp-phylym}
# Define detection and prevalence thresholds to filter out rare taxa
pseqcore <- core(pseq, detection = 0.1/100, prevalence = 1/100)

# For the available taxonomic levels
pcomcore <- plot_taxa_prevalence(pseqcore, "Phylum", detection = 0.1/100)
pcomcore
```

```{r comp-aggregated}
# get relative abudance
pseq.comp.fam <- aggregate_rare(
  pseq.comp,
  level = "Family",
  detection = 0.005,
  prevalence = 0.5
)

pcomp.fam <- plot_composition(pseq.comp.fam, average_by = "group") +
  guides(fill = guide_legend(ncol = 1)) +
  labs(
    x = "group",
    y = "Relative abundance",
    title = "Relative abundance data",
  )
pcomp.fam <- pcomp.fam + scale_fill_brewer("Family", palette = "Paired")
pcomp.fam
```

Make a panel figure for these two displays of taxonomic composition
```{r compmerge}
#ggpubr package for formatting multi panel figures
mergecomp <- ggarrange(pcomcore, pcomp.fam,
          labels = c("A", "B"),
          ncol = 1, nrow = 2)
mergecomp
ggsave(mergecomp, file = "comp.png", width = 7, height = 12)
```

## Heatmap of abundance

Abundance of microbial taxa (using the Z transformed data)

```{r heatmap-abundance}
# Plot the abundances heatmap, round values to 2 decimals
dfm <- melt(round(abundances(pseq2z), 1))

colnames(dfm) <- c("Taxa", "Sample", "value")

heat(dfm, "Taxa", "Sample", "value") +
  theme(text=element_text(size=10), 
        axis.text.x = element_text(angle = 90, hjust = 1),
        legend.key.size = unit(1.2, "cm"))
```

## Correlation between microbes and lipids

Cross-correlate columns of two data sets from related to microbiome and blood serum lipids associations ([PeerJ 1:e32](https://peerj.com/articles/32/)).

The function returns correlations, raw p-values, and fdr estimates (not strictly proper as the comparisons are not independent). Keep only those elements that have at least only one significant correlation (n.signif):

```{r correlation}
# Load example data 
otu <- peerj32$microbes 
lipids <- peerj32$lipids 

# Define data sets to cross-correlate
x <- log10(otu) # OTU Log10 (44 samples x 130 genera)
y <- as.matrix(lipids) # Lipids (44 samples x 389 lipids)

# Cross correlate data sets
correlations <- associate(x, y, method = "spearman", mode = "matrix", p.adj.threshold = 0.05, n.signif = 1)

# Or, alternatively, the same output is also available in a handy table format
correlation.table <- associate(x, y, method = "spearman", mode = "table", p.adj.threshold = 0.05, n.signif = 1)

kable(head(correlation.table))
```

### Association heatmaps

Rearrange the data and plot the heatmap and mark significant correlations with stars to reproduce microbiota-lipidome heatmap from [Lahti et al. PeerJ (2013)](https://peerj.com/articles/32/) (the ordering of rows and columns may be different):

```{r associations}
heatmap <- heat(correlation.table, "X1", "X2", 
          fill = "Correlation", 
          star = "p.adj", 
          p.adj.threshold = 0.05) 
heatmap + theme(text=element_text(size=10), 
          axis.text.x = element_text(angle = 90, hjust = 1), 
          legend.key.size = unit(1.3, "cm"))
```

### Heatmap of significant correlations

This time we are modifying the plot to only include significant correlations with a threshold of P < 0.05. The cells marked with a white cross indicate most significant correlations with a significance of P < 0.02. 

```{r correlation-sig}
# Order the rows and columns with levels argument if needed:
correlation.table$X1 <- factor(correlation.table$X1, levels = unique(as.character(correlation.table$X1)))
correlation.table$X2 <- factor(correlation.table$X2, levels = unique(as.character(correlation.table$X2)))

# Pick only the correlations with p.adj<0.05
# Note: this will leave other cells empty
subtable <- filter(correlation.table, p.adj < 0.05)

# Arrange the figure
heatmap2 <- ggplot(subtable, aes(x = X1, y = X2, fill = Correlation))
heatmap2 <- heatmap2 + geom_tile() 
heatmap2 <- heatmap2 + scale_fill_gradientn("Correlation", 
                       breaks = seq(from = -1, to = 1, by = 0.2), 
                   colours = c("darkblue", "blue", "white", "red", "darkred"), 
                   limits = c(-1,1)) 

# Polish texts
heatmap2 <- heatmap2 + theme(axis.text.x=element_text(angle = 90, hjust=1, face = "italic"),
               axis.text.y=element_text(size = 8))
heatmap2 <- heatmap2 + xlab("") + ylab("")

# Mark the most significant cells with stars using threshold of P < 0.02
heatmap2 <- heatmap2 + geom_text(data = subset(correlation.table, p.adj < 0.02), 
               aes(x = X1, y = X2, label = "+"), col = "white", size = 5)

# Plot
print(heatmap2)
```



---

```{r session-info}
sessionInfo()
```
